// solapi/dist/index.js
'use strict';

var effect = require('effect');
var crypto = require('crypto');
var dateFns = require('date-fns');
var Effect10 = require('effect/Effect');
var fs = require('fs');
var url = require('url');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var Effect10__namespace = /*#__PURE__*/_interopNamespace(Effect10);

// src/lib/effectErrorHandler.ts
var InvalidDateError = class extends effect.Data.TaggedError("InvalidDateError") {
  toString() {
    return `InvalidDateError: ${this.message}${this.originalValue ? ` (value: ${this.originalValue})` : ""}`;
  }
}, ApiKeyError = class extends effect.Data.TaggedError("ApiKeyError") {
  toString() {
    return `ApiKeyError: ${this.message}`;
  }
}, DefaultError = class extends effect.Data.TaggedError("DefaultError") {
  get message() {
    return `${this.errorCode}: ${this.errorMessage}`;
  }
  toString() {
    return this.message;
  }
}, MessageNotReceivedError = class extends effect.Data.TaggedError(
  "MessageNotReceivedError"
) {
  get message() {
    return `${this.totalCount}\uAC1C\uC758 \uBA54\uC2DC\uC9C0\uAC00 \uC811\uC218\uB418\uC9C0 \uBABB\uD588\uC2B5\uB2C8\uB2E4. \uC790\uC138\uD55C \uC5D0\uB7EC \uBA54\uC2DC\uC9C0\uB294 \uD574\uB2F9 \uC5D0\uB7EC \uB0B4 failedMessageList\uB97C \uD655\uC778\uD574\uC8FC\uC138\uC694.`;
  }
  toString() {
    return `MessageNotReceivedError: ${this.message}
Failed messages: ${this.failedMessageList.length}`;
  }
}, BadRequestError = class extends effect.Data.TaggedError("BadRequestError") {
  toString() {
    return `BadRequestError: ${this.message}${this.validationErrors ? `
Validation errors: ${this.validationErrors.join(", ")}` : ""}`;
  }
}, NetworkError = class extends effect.Data.TaggedError("NetworkError") {
  get message() {
    return `${this.method} ${this.url} \uC694\uCCAD \uC2E4\uD328 - ${this.cause}`;
  }
  toString() {
    return `NetworkError: ${this.message}`;
  }
}, ClientError = class extends effect.Data.TaggedError("ClientError") {
  get message() {
    return `${this.errorCode}: ${this.errorMessage}`;
  }
  toString() {
    return this.message;
  }
}, UnexpectedDefectError = class extends effect.Data.TaggedError(
  "UnexpectedDefectError"
) {
  toString() {
    return `UnexpectedDefectError: ${this.message}`;
  }
}, UnhandledExitError = class extends effect.Data.TaggedError("UnhandledExitError") {
  toString() {
    return `UnhandledExitError: ${this.message}`;
  }
}, ResponseSchemaMismatchError = class extends effect.Data.TaggedError(
  "ResponseSchemaMismatchError"
) {
  toString() {
    let header = `ResponseSchemaMismatchError: ${this.message}`, url = this.url ? `
URL: ${this.url}` : "", issues = this.validationErrors.length > 0 ? `
Issues:
- ${this.validationErrors.join(`
- `)}` : "", env = "production"?.trim().toLowerCase(), body = (env === "development" || env === "test") && this.responseBody ? `
Response: ${this.responseBody.substring(0, 500)}` : "";
    return `${header}${url}${issues}${body}`;
  }
}, ServerError = class extends effect.Data.TaggedError("ServerError") {
  get message() {
    return `${this.errorCode} - ${this.errorMessage}`;
  }
  toString() {
    return `ServerError(${this.httpStatus}): ${this.message}`;
  }
}, isErrorResponse = (value) => value == null || typeof value != "object" || !("errorCode" in value) || !("errorMessage" in value) ? false : typeof value.errorCode == "string" && value.errorCode !== "" && typeof value.errorMessage == "string" && value.errorMessage !== "";

// src/lib/effectErrorHandler.ts
var isTaggedDefect = (value) => value !== null && typeof value == "object" && "_tag" in value && typeof value._tag == "string", extractDefectInfo = (defect) => {
  if (isTaggedDefect(defect)) {
    let tag = defect._tag, message = defect.message != null ? String(defect.message) : "";
    return {
      summary: `${tag}${message ? `: ${message}` : ""}`,
      details: `Tagged Error [${tag}]: ${JSON.stringify(defect, null, 2)}`
    };
  }
  if (defect !== null && typeof defect == "object") {
    let keys = Object.keys(defect);
    return {
      summary: keys.length > 0 ? `Object with keys: ${keys.slice(0, 3).join(", ")}${keys.length > 3 ? "..." : ""}` : "Empty object",
      details: JSON.stringify(defect, null, 2)
    };
  }
  return {
    summary: String(defect),
    details: `Value (${typeof defect}): ${String(defect)}`
  };
}, unwrapCause = (cause) => {
  let failure = effect.Cause.failureOption(cause);
  if (failure._tag === "Some")
    return failure.value;
  let defects = effect.Cause.defects(cause);
  if (defects.length > 0) {
    let firstDefect = effect.Chunk.unsafeGet(defects, 0);
    if (firstDefect instanceof Error)
      return firstDefect;
    let defectInfo = extractDefectInfo(firstDefect), message2 = `Unexpected error: ${defectInfo.summary}` ;
    return new UnexpectedDefectError({ message: message2 });
  }
  let message = "Effect execution failed unexpectedly" ;
  return new UnhandledExitError({ message });
}, runSafePromise = (effect$1) => effect.Effect.runPromiseExit(effect$1).then(
  effect.Exit.match({
    onFailure: (cause) => Promise.reject(unwrapCause(cause)),
    onSuccess: (value) => Promise.resolve(value)
  })
);
var countSchema = effect.Schema.Struct({
  total: effect.Schema.Number,
  sentTotal: effect.Schema.Number,
  sentFailed: effect.Schema.Number,
  sentSuccess: effect.Schema.Number,
  sentPending: effect.Schema.Number,
  sentReplacement: effect.Schema.Number,
  refund: effect.Schema.Number,
  registeredFailed: effect.Schema.Number,
  registeredSuccess: effect.Schema.Number
}), countryChargeStatusSchema = effect.Schema.Record({
  key: effect.Schema.String,
  value: effect.Schema.Number
}), countForChargeSchema = effect.Schema.Struct({
  sms: countryChargeStatusSchema,
  lms: countryChargeStatusSchema,
  mms: countryChargeStatusSchema,
  ata: countryChargeStatusSchema,
  cta: countryChargeStatusSchema,
  cti: countryChargeStatusSchema,
  nsa: countryChargeStatusSchema,
  rcs_sms: countryChargeStatusSchema,
  rcs_lms: countryChargeStatusSchema,
  rcs_mms: countryChargeStatusSchema,
  rcs_tpl: countryChargeStatusSchema,
  rcs_itpl: effect.Schema.optional(countryChargeStatusSchema),
  rcs_ltpl: effect.Schema.optional(countryChargeStatusSchema),
  fax: effect.Schema.optional(countryChargeStatusSchema),
  voice: effect.Schema.optional(countryChargeStatusSchema),
  bms_text: effect.Schema.optional(countryChargeStatusSchema),
  bms_image: effect.Schema.optional(countryChargeStatusSchema),
  bms_wide: effect.Schema.optional(countryChargeStatusSchema),
  bms_wide_item_list: effect.Schema.optional(countryChargeStatusSchema),
  bms_carousel_feed: effect.Schema.optional(countryChargeStatusSchema),
  bms_premium_video: effect.Schema.optional(countryChargeStatusSchema),
  bms_commerce: effect.Schema.optional(countryChargeStatusSchema),
  bms_carousel_commerce: effect.Schema.optional(countryChargeStatusSchema),
  bms_free: effect.Schema.optional(countryChargeStatusSchema)
}), commonCashResponseSchema = effect.Schema.Struct({
  requested: effect.Schema.Number,
  replacement: effect.Schema.Number,
  refund: effect.Schema.Number,
  sum: effect.Schema.Number
}), messageTypeRecordSchema = effect.Schema.Struct({
  sms: effect.Schema.Number,
  lms: effect.Schema.Number,
  mms: effect.Schema.Number,
  ata: effect.Schema.Number,
  cta: effect.Schema.Number,
  cti: effect.Schema.Number,
  nsa: effect.Schema.Number,
  rcs_sms: effect.Schema.Number,
  rcs_lms: effect.Schema.Number,
  rcs_mms: effect.Schema.Number,
  rcs_tpl: effect.Schema.Number,
  rcs_itpl: effect.Schema.optional(effect.Schema.Number),
  rcs_ltpl: effect.Schema.optional(effect.Schema.Number),
  fax: effect.Schema.optional(effect.Schema.Number),
  voice: effect.Schema.optional(effect.Schema.Number),
  bms_text: effect.Schema.optional(effect.Schema.Number),
  bms_image: effect.Schema.optional(effect.Schema.Number),
  bms_wide: effect.Schema.optional(effect.Schema.Number),
  bms_wide_item_list: effect.Schema.optional(effect.Schema.Number),
  bms_carousel_feed: effect.Schema.optional(effect.Schema.Number),
  bms_premium_video: effect.Schema.optional(effect.Schema.Number),
  bms_commerce: effect.Schema.optional(effect.Schema.Number),
  bms_carousel_commerce: effect.Schema.optional(effect.Schema.Number),
  bms_free: effect.Schema.optional(effect.Schema.Number)
}), partialMessageTypeRecordSchema = effect.Schema.partial(
  messageTypeRecordSchema
), appSchema = effect.Schema.Struct({
  profit: messageTypeRecordSchema,
  appId: effect.Schema.NullishOr(effect.Schema.String)
}), logSchema = effect.Schema.Array(
  effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.Unknown })
), groupIdSchema = effect.Schema.String, groupSchema = effect.Schema.Struct({
  count: countSchema,
  balance: commonCashResponseSchema,
  point: commonCashResponseSchema,
  app: appSchema,
  sdkVersion: effect.Schema.NullishOr(effect.Schema.String),
  osPlatform: effect.Schema.NullishOr(effect.Schema.String),
  log: logSchema,
  status: effect.Schema.String,
  scheduledDate: effect.Schema.NullishOr(effect.Schema.String),
  dateSent: effect.Schema.NullishOr(effect.Schema.String),
  dateCompleted: effect.Schema.NullishOr(effect.Schema.String),
  isRefunded: effect.Schema.Boolean,
  groupId: groupIdSchema,
  accountId: effect.Schema.String,
  countForCharge: countForChargeSchema,
  dateCreated: effect.Schema.String,
  dateUpdated: effect.Schema.String
}), handleKeySchema = effect.Schema.String, blackSchema = effect.Schema.Struct({
  handleKey: handleKeySchema,
  type: effect.Schema.Literal("DENIAL"),
  senderNumber: effect.Schema.String,
  recipientNumber: effect.Schema.String,
  dateCreated: effect.Schema.String,
  dateUpdated: effect.Schema.String
}), blockGroupSchema = effect.Schema.Struct({
  blockGroupId: effect.Schema.String,
  accountId: effect.Schema.String,
  status: effect.Schema.Literal("INACTIVE", "ACTIVE"),
  name: effect.Schema.NullishOr(effect.Schema.String),
  useAll: effect.Schema.Boolean,
  senderNumbers: effect.Schema.Array(effect.Schema.String),
  dateCreated: effect.Schema.String,
  dateUpdated: effect.Schema.String
}), blockNumberSchema = effect.Schema.Struct({
  blockNumberId: effect.Schema.String,
  accountId: effect.Schema.String,
  memo: effect.Schema.String,
  phoneNumber: effect.Schema.String,
  blockGroupIds: effect.Schema.Array(effect.Schema.String),
  dateCreated: effect.Schema.String,
  dateUpdated: effect.Schema.String
});
var bmsButtonLinkTypeSchema = effect.Schema.Literal(
  "AC",
  "WL",
  "AL",
  "BK",
  "MD",
  "BC",
  "BT",
  "BF"
), bmsWebButtonSchema = effect.Schema.Struct({
  name: effect.Schema.String,
  linkType: effect.Schema.Literal("WL"),
  linkMobile: effect.Schema.String,
  linkPc: effect.Schema.optional(effect.Schema.String),
  targetOut: effect.Schema.optional(effect.Schema.Boolean)
}), bmsAppButtonSchema = effect.Schema.Struct({
  name: effect.Schema.String,
  linkType: effect.Schema.Literal("AL"),
  linkMobile: effect.Schema.optional(effect.Schema.String),
  linkAndroid: effect.Schema.optional(effect.Schema.String),
  linkIos: effect.Schema.optional(effect.Schema.String),
  targetOut: effect.Schema.optional(effect.Schema.Boolean)
}).pipe(
  effect.Schema.filter((button) => button.linkMobile !== void 0 && button.linkMobile.trim() !== "" || button.linkAndroid !== void 0 && button.linkAndroid.trim() !== "" || button.linkIos !== void 0 && button.linkIos.trim() !== "" ? true : "AL \uD0C0\uC785 \uBC84\uD2BC\uC740 linkMobile, linkAndroid, linkIos \uC911 \uD558\uB098 \uC774\uC0C1 \uD544\uC218\uC785\uB2C8\uB2E4.")
), bmsChannelAddButtonSchema = effect.Schema.Struct({
  name: effect.Schema.optional(effect.Schema.String),
  linkType: effect.Schema.Literal("AC")
}), bmsBotKeywordButtonSchema = effect.Schema.Struct({
  name: effect.Schema.String,
  linkType: effect.Schema.Literal("BK"),
  chatExtra: effect.Schema.optional(effect.Schema.String)
}), bmsMessageDeliveryButtonSchema = effect.Schema.Struct({
  name: effect.Schema.String,
  linkType: effect.Schema.Literal("MD"),
  chatExtra: effect.Schema.optional(effect.Schema.String)
}), bmsConsultButtonSchema = effect.Schema.Struct({
  name: effect.Schema.String,
  linkType: effect.Schema.Literal("BC"),
  chatExtra: effect.Schema.optional(effect.Schema.String)
}), bmsBotTransferButtonSchema = effect.Schema.Struct({
  name: effect.Schema.String,
  linkType: effect.Schema.Literal("BT"),
  chatExtra: effect.Schema.optional(effect.Schema.String)
}), bmsBusinessFormButtonSchema = effect.Schema.Struct({
  name: effect.Schema.String,
  linkType: effect.Schema.Literal("BF")
}), bmsButtonSchema = effect.Schema.Union(
  bmsWebButtonSchema,
  bmsAppButtonSchema,
  bmsChannelAddButtonSchema,
  bmsBotKeywordButtonSchema,
  bmsMessageDeliveryButtonSchema,
  bmsConsultButtonSchema,
  bmsBotTransferButtonSchema,
  bmsBusinessFormButtonSchema
), bmsLinkButtonSchema = effect.Schema.Union(
  bmsWebButtonSchema,
  bmsAppButtonSchema
);
var NumberOrNumericString = effect.Schema.transformOrFail(
  effect.Schema.Union(effect.Schema.Number, effect.Schema.String),
  effect.Schema.Number,
  {
    strict: true,
    decode: (input, _, ast) => {
      if (typeof input == "number")
        return effect.ParseResult.succeed(input);
      if (input.trim() === "")
        return effect.ParseResult.fail(
          new effect.ParseResult.Type(ast, input, "\uC720\uD6A8\uD55C \uC22B\uC790 \uD615\uC2DD\uC774 \uC544\uB2D9\uB2C8\uB2E4.")
        );
      let parsed = parseFloat(input);
      return Number.isNaN(parsed) ? effect.ParseResult.fail(
        new effect.ParseResult.Type(ast, input, "\uC720\uD6A8\uD55C \uC22B\uC790 \uD615\uC2DD\uC774 \uC544\uB2D9\uB2C8\uB2E4.")
      ) : effect.ParseResult.succeed(parsed);
    },
    encode: (n) => effect.ParseResult.succeed(n)
  }
), COMMERCE_PRICE_MAX = 99999999, DISCOUNT_RATE_MAX = 100, numberInRange = (fieldName, min, max) => NumberOrNumericString.pipe(
  effect.Schema.filter((n) => n >= min && n <= max, {
    message: () => `${fieldName} \uAC12\uC774 \uC798\uBABB\uB418\uC5C8\uC2B5\uB2C8\uB2E4. ${min} \uC774\uC0C1 ${max} \uC774\uD558\uC758 \uC22B\uC790\uC5EC\uC57C \uD569\uB2C8\uB2E4.`
  })
), validateCommercePricingCombination = (commerce) => {
  let hasDiscountPrice = commerce.discountPrice !== void 0, hasDiscountRate = commerce.discountRate !== void 0, hasDiscountFixed = commerce.discountFixed !== void 0;
  return !hasDiscountPrice && !hasDiscountRate && !hasDiscountFixed || hasDiscountPrice && hasDiscountRate && !hasDiscountFixed || hasDiscountPrice && hasDiscountFixed && !hasDiscountRate ? true : hasDiscountRate && hasDiscountFixed ? "discountRate\uC640 discountFixed\uB294 \uB3D9\uC2DC\uC5D0 \uC0AC\uC6A9\uD560 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4. \uD560\uC778\uC728(discountRate) \uB610\uB294 \uC815\uC561\uD560\uC778(discountFixed) \uC911 \uD558\uB098\uB9CC \uC120\uD0DD\uD558\uC138\uC694." : !hasDiscountPrice && (hasDiscountRate || hasDiscountFixed) ? "discountRate \uB610\uB294 discountFixed\uB97C \uC0AC\uC6A9\uD558\uB824\uBA74 discountPrice(\uD560\uC778\uAC00)\uB3C4 \uD568\uAED8 \uC9C0\uC815\uD574\uC57C \uD569\uB2C8\uB2E4." : hasDiscountPrice && !hasDiscountRate && !hasDiscountFixed ? "discountPrice\uB97C \uC0AC\uC6A9\uD558\uB824\uBA74 discountRate(\uD560\uC778\uC728) \uB610\uB294 discountFixed(\uC815\uC561\uD560\uC778) \uC911 \uD558\uB098\uB97C \uD568\uAED8 \uC9C0\uC815\uD574\uC57C \uD569\uB2C8\uB2E4." : "\uC54C \uC218 \uC5C6\uB294 \uAC00\uACA9 \uC870\uD569\uC785\uB2C8\uB2E4. regularPrice\uB9CC \uC0AC\uC6A9\uD558\uAC70\uB098, regularPrice + discountPrice + discountRate/discountFixed \uC870\uD569\uC744 \uC0AC\uC6A9\uD558\uC138\uC694.";
}, bmsCommerceSchema = effect.Schema.Struct({
  title: effect.Schema.String,
  regularPrice: numberInRange("regularPrice", 0, COMMERCE_PRICE_MAX),
  discountPrice: effect.Schema.optional(
    numberInRange("discountPrice", 0, COMMERCE_PRICE_MAX)
  ),
  discountRate: effect.Schema.optional(
    numberInRange("discountRate", 0, DISCOUNT_RATE_MAX)
  ),
  discountFixed: effect.Schema.optional(
    numberInRange("discountFixed", 0, COMMERCE_PRICE_MAX)
  )
}).pipe(effect.Schema.filter(validateCommercePricingCombination));
var wonDiscountPattern = /^([1-9]\d{0,7})원 할인 쿠폰$/, percentDiscountPattern = /^([1-9]\d?|100)% 할인 쿠폰$/, freeCouponPattern = /^\S{1,7} 무료 쿠폰$/, upCouponPattern = /^\S{1,7} UP 쿠폰$/, isValidCouponTitle = (title) => {
  if (title === "\uBC30\uC1A1\uBE44 \uD560\uC778 \uCFE0\uD3F0") return true;
  let wonMatch = title.match(wonDiscountPattern);
  if (wonMatch) {
    let num = parseInt(wonMatch[1], 10);
    return num >= 1 && num <= 99999999;
  }
  return percentDiscountPattern.test(title) || freeCouponPattern.test(title) ? true : upCouponPattern.test(title);
}, bmsCouponTitleSchema = effect.Schema.String.pipe(
  effect.Schema.filter(isValidCouponTitle, {
    message: () => '\uCFE0\uD3F0 \uC81C\uBAA9\uC740 \uB2E4\uC74C \uD615\uC2DD \uC911 \uD558\uB098\uC5EC\uC57C \uD569\uB2C8\uB2E4: "N\uC6D0 \uD560\uC778 \uCFE0\uD3F0" (1~99999999), "N% \uD560\uC778 \uCFE0\uD3F0" (1~100), "\uBC30\uC1A1\uBE44 \uD560\uC778 \uCFE0\uD3F0", "OOO \uBB34\uB8CC \uCFE0\uD3F0" (7\uC790 \uC774\uB0B4), "OOO UP \uCFE0\uD3F0" (7\uC790 \uC774\uB0B4)'
  })
), bmsCouponSchema = effect.Schema.Struct({
  title: bmsCouponTitleSchema,
  description: effect.Schema.String,
  linkMobile: effect.Schema.optional(effect.Schema.String),
  linkPc: effect.Schema.optional(effect.Schema.String),
  linkAndroid: effect.Schema.optional(effect.Schema.String),
  linkIos: effect.Schema.optional(effect.Schema.String)
});

// src/models/base/kakao/bms/bmsCarousel.ts
var bmsCarouselHeadSchema = effect.Schema.Struct({
  header: effect.Schema.String,
  content: effect.Schema.String,
  imageId: effect.Schema.String,
  linkMobile: effect.Schema.optional(effect.Schema.String),
  linkPc: effect.Schema.optional(effect.Schema.String),
  linkAndroid: effect.Schema.optional(effect.Schema.String),
  linkIos: effect.Schema.optional(effect.Schema.String)
}).pipe(
  effect.Schema.filter((head) => (head.linkPc !== void 0 || head.linkAndroid !== void 0 || head.linkIos !== void 0) && head.linkMobile === void 0 ? "linkPc, linkAndroid, linkIos \uC911 \uD558\uB098\uB77C\uB3C4 \uC788\uC73C\uBA74 linkMobile \uAC12\uC774 \uD544\uC218\uC785\uB2C8\uB2E4." : true)
), bmsCarouselTailSchema = effect.Schema.Struct({
  linkMobile: effect.Schema.String,
  linkPc: effect.Schema.optional(effect.Schema.String),
  linkAndroid: effect.Schema.optional(effect.Schema.String),
  linkIos: effect.Schema.optional(effect.Schema.String)
}), bmsCarouselFeedItemSchema = effect.Schema.Struct({
  header: effect.Schema.String,
  content: effect.Schema.String,
  imageId: effect.Schema.String,
  imageLink: effect.Schema.optional(effect.Schema.String),
  buttons: effect.Schema.Array(bmsLinkButtonSchema),
  coupon: effect.Schema.optional(bmsCouponSchema)
}), bmsCarouselCommerceItemSchema = effect.Schema.Struct({
  commerce: bmsCommerceSchema,
  imageId: effect.Schema.String,
  imageLink: effect.Schema.optional(effect.Schema.String),
  buttons: effect.Schema.Array(bmsLinkButtonSchema),
  additionalContent: effect.Schema.optional(effect.Schema.String),
  coupon: effect.Schema.optional(bmsCouponSchema)
}), bmsCarouselFeedSchema = effect.Schema.Struct({
  list: effect.Schema.Array(bmsCarouselFeedItemSchema),
  tail: effect.Schema.optional(bmsCarouselTailSchema)
}), bmsCarouselCommerceSchema = effect.Schema.Struct({
  head: effect.Schema.optional(bmsCarouselHeadSchema),
  list: effect.Schema.Array(bmsCarouselCommerceItemSchema),
  tail: effect.Schema.optional(bmsCarouselTailSchema)
});
var bmsChatBubbleTypeSchema = effect.Schema.Literal(
  "TEXT",
  "IMAGE",
  "WIDE",
  "WIDE_ITEM_LIST",
  "COMMERCE",
  "CAROUSEL_FEED",
  "CAROUSEL_COMMERCE",
  "PREMIUM_VIDEO"
);

// src/models/base/kakao/bms/bmsConstraints.ts
var BMS_COUPON_DESCRIPTION_MAX = {
  TEXT: 12,
  IMAGE: 12,
  WIDE: 18,
  WIDE_ITEM_LIST: 18,
  COMMERCE: 12,
  CAROUSEL_FEED: 12,
  CAROUSEL_COMMERCE: 12,
  PREMIUM_VIDEO: 12
}, BMS_BUTTON_COUNT = {
  TEXT: { max: 5, maxWithCoupon: 4 },
  IMAGE: { max: 5, maxWithCoupon: 4 },
  WIDE: { max: 2 },
  WIDE_ITEM_LIST: { max: 2 },
  COMMERCE: { max: 2, min: 1 },
  CAROUSEL_FEED: { max: 2, min: 1 },
  CAROUSEL_COMMERCE: { max: 2, min: 1 },
  PREMIUM_VIDEO: { max: 1 }
}, BMS_BUTTON_NAME_MAX = {
  TEXT: 14,
  IMAGE: 14,
  WIDE: 8,
  WIDE_ITEM_LIST: 8,
  COMMERCE: 8,
  CAROUSEL_FEED: 8,
  CAROUSEL_COMMERCE: 8,
  PREMIUM_VIDEO: 8
}, BMS_ALLOWED_LINK_TYPES = {
  COMMERCE: ["WL", "AL"],
  CAROUSEL_FEED: ["WL", "AL"],
  CAROUSEL_COMMERCE: ["WL", "AL"]
}, BMS_HEADER_MAX = 20, BMS_ADDITIONAL_CONTENT_MAX = 34, BMS_MAIN_WIDE_ITEM_TITLE_MAX = 25, BMS_SUB_WIDE_ITEM_TITLE_MAX = 30, BMS_CAROUSEL_HEAD_CONTENT_MAX = 50, BMS_CAROUSEL_FEED_ITEM_CONTENT_MAX = 180, BMS_CAROUSEL_COMMERCE_ADDITIONAL_CONTENT_MAX = 34, BMS_CONTENT_MAX = {
  TEXT: 1300,
  IMAGE: 1300,
  WIDE: 76,
  PREMIUM_VIDEO: 76
}, BMS_CAROUSEL_LIST_RANGE = {
  withHead: { min: 1, max: 5 },
  withoutHead: { min: 2, max: 6 }
}, hasCoupon = (bms) => bms.coupon != null, validateCouponDescription = (bms) => {
  let limit = BMS_COUPON_DESCRIPTION_MAX[bms.chatBubbleType];
  if (bms.coupon && bms.coupon.description.length > limit)
    return `\uCFE0\uD3F0 \uC124\uBA85\uC740 \uCD5C\uB300 ${limit}\uC790 \uC774\uD558\uB85C \uC785\uB825\uD574\uC8FC\uC138\uC694.`;
  for (let item of bms.carousel?.list ?? [])
    if (item.coupon && item.coupon.description.length > limit)
      return `\uCFE0\uD3F0 \uC124\uBA85\uC740 \uCD5C\uB300 ${limit}\uC790 \uC774\uD558\uB85C \uC785\uB825\uD574\uC8FC\uC138\uC694.`;
  return true;
}, validateSingleButtonGroup = (chatBubbleType, buttons, withCoupon) => {
  let range = BMS_BUTTON_COUNT[chatBubbleType], max = withCoupon && range.maxWithCoupon != null ? range.maxWithCoupon : range.max;
  return buttons.length > max ? withCoupon && range.maxWithCoupon != null ? `${chatBubbleType} \uD0C0\uC785\uC5D0\uC11C\uB294 \uCFE0\uD3F0 \uC0AC\uC6A9 \uC2DC \uCD5C\uB300 ${max}\uAC1C\uC758 \uBC84\uD2BC\uB9CC \uC0AC\uC6A9\uD560 \uC218 \uC788\uC2B5\uB2C8\uB2E4.` : `${chatBubbleType} \uD0C0\uC785\uC5D0\uC11C\uB294 \uCD5C\uB300 ${max}\uAC1C\uC758 \uBC84\uD2BC\uB9CC \uC0AC\uC6A9\uD560 \uC218 \uC788\uC2B5\uB2C8\uB2E4.` : range.min != null && buttons.length < range.min ? `${chatBubbleType} \uD0C0\uC785\uC5D0\uC11C\uB294 \uCD5C\uC18C ${range.min}\uAC1C\uC758 \uBC84\uD2BC\uC774 \uD544\uC694\uD569\uB2C8\uB2E4.` : true;
}, validateButtonCount = (bms) => {
  let type = bms.chatBubbleType;
  if (type === "CAROUSEL_FEED" || type === "CAROUSEL_COMMERCE") {
    for (let item of bms.carousel?.list ?? []) {
      let result = validateSingleButtonGroup(
        type,
        item.buttons ?? [],
        hasCoupon(bms) || item.coupon != null
      );
      if (result !== true) return result;
    }
    return true;
  }
  return bms.buttons ? validateSingleButtonGroup(type, bms.buttons, hasCoupon(bms)) : true;
}, validateButtonNames = (bms) => {
  let type = bms.chatBubbleType, limit = BMS_BUTTON_NAME_MAX[type], checkList = (buttons) => {
    for (let button of buttons)
      if (button.linkType !== "AC" && button.name != null && button.name.length > limit)
        return `${type} \uD0C0\uC785 button.name\uC740 \uCD5C\uB300 ${limit}\uC790 \uC774\uD558\uB85C \uC785\uB825\uD574\uC8FC\uC138\uC694.`;
    return true;
  };
  if (bms.buttons) {
    let result = checkList(bms.buttons);
    if (result !== true) return result;
  }
  for (let item of bms.carousel?.list ?? []) {
    let result = checkList(item.buttons ?? []);
    if (result !== true) return result;
  }
  return true;
}, validateAllowedLinkTypes = (bms) => {
  let type = bms.chatBubbleType, allowed = BMS_ALLOWED_LINK_TYPES[type];
  if (!allowed) return true;
  let checkList = (buttons) => {
    for (let button of buttons)
      if (!allowed.includes(button.linkType))
        return `${type} \uD0C0\uC785\uC5D0\uC11C\uB294 ${allowed.join(", ")} \uD0C0\uC785\uC758 \uBC84\uD2BC\uB9CC \uC0AC\uC6A9\uD560 \uC218 \uC788\uC2B5\uB2C8\uB2E4.`;
    return true;
  };
  if (bms.buttons) {
    let result = checkList(bms.buttons);
    if (result !== true) return result;
  }
  for (let item of bms.carousel?.list ?? []) {
    let result = checkList(item.buttons ?? []);
    if (result !== true) return result;
  }
  return true;
}, lengthError = (chatBubbleType, fieldName, max) => `${chatBubbleType} \uD0C0\uC785 ${fieldName}\uC740 \uCD5C\uB300 ${max}\uC790 \uC774\uD558\uB85C \uC785\uB825\uD574\uC8FC\uC138\uC694.`, validateTextLengths = (bms) => {
  let type = bms.chatBubbleType;
  if (bms.header != null && bms.header.length > BMS_HEADER_MAX)
    return lengthError(type, "header", BMS_HEADER_MAX);
  let contentLimit = BMS_CONTENT_MAX[type];
  if (contentLimit != null && bms.content != null && bms.content.length > contentLimit)
    return lengthError(type, "content", contentLimit);
  if (bms.additionalContent != null && bms.additionalContent.length > BMS_ADDITIONAL_CONTENT_MAX)
    return lengthError(type, "additionalContent", BMS_ADDITIONAL_CONTENT_MAX);
  if (bms.mainWideItem?.title != null && bms.mainWideItem.title.length > BMS_MAIN_WIDE_ITEM_TITLE_MAX)
    return lengthError(
      type,
      "mainWideItem.title",
      BMS_MAIN_WIDE_ITEM_TITLE_MAX
    );
  for (let sub of bms.subWideItemList ?? [])
    if (sub.title.length > BMS_SUB_WIDE_ITEM_TITLE_MAX)
      return lengthError(
        type,
        "subWideItem.title",
        BMS_SUB_WIDE_ITEM_TITLE_MAX
      );
  if (bms.carousel?.head) {
    let { header, content } = bms.carousel.head;
    if (header != null && header.length > BMS_HEADER_MAX)
      return lengthError(type, "carousel.head.header", BMS_HEADER_MAX);
    if (content != null && content.length > BMS_CAROUSEL_HEAD_CONTENT_MAX)
      return lengthError(
        type,
        "carousel.head.content",
        BMS_CAROUSEL_HEAD_CONTENT_MAX
      );
  }
  if (type === "CAROUSEL_FEED")
    for (let item of bms.carousel?.list ?? []) {
      if (item.header != null && item.header.length > BMS_HEADER_MAX)
        return lengthError(type, "carousel.list.header", BMS_HEADER_MAX);
      if (item.content != null && item.content.length > BMS_CAROUSEL_FEED_ITEM_CONTENT_MAX)
        return lengthError(
          type,
          "carousel.list.content",
          BMS_CAROUSEL_FEED_ITEM_CONTENT_MAX
        );
    }
  if (type === "CAROUSEL_COMMERCE") {
    for (let item of bms.carousel?.list ?? [])
      if (item.additionalContent != null && item.additionalContent.length > BMS_CAROUSEL_COMMERCE_ADDITIONAL_CONTENT_MAX)
        return lengthError(
          type,
          "carousel.list.additionalContent",
          BMS_CAROUSEL_COMMERCE_ADDITIONAL_CONTENT_MAX
        );
  }
  return true;
}, validateCarouselListCount = (bms) => {
  let type = bms.chatBubbleType;
  if (type !== "CAROUSEL_FEED" && type !== "CAROUSEL_COMMERCE") return true;
  let list = bms.carousel?.list ?? [], hasHead = bms.carousel?.head != null, range = hasHead ? BMS_CAROUSEL_LIST_RANGE.withHead : BMS_CAROUSEL_LIST_RANGE.withoutHead;
  return list.length < range.min || list.length > range.max ? `${hasHead ? "\uCE90\uB7EC\uC140 \uC778\uD2B8\uB85C\uAC00 \uC788\uB294 \uACBD\uC6B0 \uCE90\uB7EC\uC140 \uB9AC\uC2A4\uD2B8\uB294" : "\uCE90\uB7EC\uC140 \uB9AC\uC2A4\uD2B8\uB294"} \uCD5C\uC18C ${range.min}\uAC1C, \uCD5C\uB300 ${range.max}\uAC1C\uAE4C\uC9C0 \uAC00\uB2A5\uD569\uB2C8\uB2E4.` : true;
}, countNewlines = (text) => {
  let matches = text.match(/\r\n|\r|\n/g);
  return matches ? matches.length : 0;
}, newlineError = (chatBubbleType, fieldName, max) => `${chatBubbleType} \uD0C0\uC785 ${fieldName}\uC740 \uC904\uBC14\uAFC8 \uCD5C\uB300 ${max}\uAC1C \uAE4C\uC9C0 \uAC00\uB2A5\uD569\uB2C8\uB2E4.`, contentNewlineLimit = (type) => type === "TEXT" || type === "IMAGE" ? 99 : type === "WIDE" || type === "PREMIUM_VIDEO" ? 1 : null, validateNewlineLimits = (bms) => {
  let type = bms.chatBubbleType;
  if (bms.header != null && countNewlines(bms.header) > 0)
    return newlineError(type, "header", 0);
  let contentNL = contentNewlineLimit(type);
  if (contentNL != null && bms.content != null && countNewlines(bms.content) > contentNL)
    return newlineError(type, "content", contentNL);
  if (bms.mainWideItem?.title != null && countNewlines(bms.mainWideItem.title) > 1)
    return newlineError(type, "mainWideItem.title", 1);
  for (let sub of bms.subWideItemList ?? [])
    if (countNewlines(sub.title) > 1)
      return newlineError(type, "subWideItem.title", 1);
  if (bms.carousel?.head) {
    let { header, content } = bms.carousel.head;
    if (header != null && countNewlines(header) > 0)
      return newlineError(type, "carousel.head.header", 0);
    if (content != null && countNewlines(content) > 2)
      return newlineError(type, "carousel.head.content", 2);
  }
  for (let item of bms.carousel?.list ?? []) {
    if (item.header != null && countNewlines(item.header) > 0)
      return newlineError(type, "carousel.list.header", 0);
    if (type === "CAROUSEL_FEED" && item.content != null && countNewlines(item.content) > 2)
      return newlineError(type, "carousel.list.content", 2);
    if (type === "CAROUSEL_COMMERCE" && item.additionalContent != null && countNewlines(item.additionalContent) > 1)
      return newlineError(type, "carousel.list.additionalContent", 1);
  }
  return true;
}, HTTP_URL_PATTERN = /^https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{1,64}\.[a-z0-9]{1,64}\b([-a-zA-Z0-9@:%~_~+.~#?&//=]*)$/i, VARIABLE_PLACEHOLDER_PATTERN = /#\{[^{}]+\}/, VARIABLE_PLACEHOLDER_GLOBAL = /#\{[^{}]+\}/g, WHITESPACE_IN_LINK_PATTERN = /\s/, isValidLink = (link) => {
  if (WHITESPACE_IN_LINK_PATTERN.test(link)) return false;
  if (HTTP_URL_PATTERN.test(link)) return true;
  if (!VARIABLE_PLACEHOLDER_PATTERN.test(link)) return false;
  let normalized = link.replace(VARIABLE_PLACEHOLDER_GLOBAL, "x");
  return HTTP_URL_PATTERN.test(normalized);
}, checkLinkPair = (obj) => obj ? obj.linkPc !== void 0 && !isValidLink(obj.linkPc) ? "linkPc \uAC12\uC774 \uC798\uBABB\uB418\uC5C8\uC2B5\uB2C8\uB2E4. \uC62C\uBC14\uB978 \uD615\uC2DD\uC740 \uC6F9 \uB9C1\uD06C \uD615\uC2DD\uC774\uC5B4\uC57C \uD569\uB2C8\uB2E4." : obj.linkMobile !== void 0 && !isValidLink(obj.linkMobile) ? "linkMobile \uAC12\uC774 \uC798\uBABB\uB418\uC5C8\uC2B5\uB2C8\uB2E4. \uC62C\uBC14\uB978 \uD615\uC2DD\uC740 \uC6F9 \uB9C1\uD06C \uD615\uC2DD\uC774\uC5B4\uC57C \uD569\uB2C8\uB2E4." : true : true, IMAGE_LINK_ERROR = "imageLink \uAC12\uC774 \uC798\uBABB\uB418\uC5C8\uC2B5\uB2C8\uB2E4. http:// \uB610\uB294 https:// \uB85C \uC2DC\uC791\uD558\uB294 \uC815\uC0C1\uC801\uC778 \uC8FC\uC18C\uB97C \uC62C\uB824\uC8FC\uC138\uC694.", checkImageLink = (imageLink) => imageLink === void 0 || isValidLink(imageLink) ? true : IMAGE_LINK_ERROR, validateLinks = (bms) => {
  let imageLinkResult = checkImageLink(bms.imageLink);
  if (imageLinkResult !== true) return imageLinkResult;
  let couponResult = checkLinkPair(bms.coupon);
  if (couponResult !== true) return couponResult;
  for (let button of bms.buttons ?? []) {
    let r = checkLinkPair(button);
    if (r !== true) return r;
  }
  let mainResult = checkLinkPair(bms.mainWideItem);
  if (mainResult !== true) return mainResult;
  for (let sub of bms.subWideItemList ?? []) {
    let r = checkLinkPair(sub);
    if (r !== true) return r;
  }
  let headResult = checkLinkPair(bms.carousel?.head);
  if (headResult !== true) return headResult;
  let tailResult = checkLinkPair(bms.carousel?.tail);
  if (tailResult !== true) return tailResult;
  for (let item of bms.carousel?.list ?? []) {
    let imageResult = checkImageLink(item.imageLink);
    if (imageResult !== true) return imageResult;
    for (let button of item.buttons ?? []) {
      let r = checkLinkPair(button);
      if (r !== true) return r;
    }
    let itemCouponResult = checkLinkPair(item.coupon);
    if (itemCouponResult !== true) return itemCouponResult;
  }
  return true;
}, FORBIDDEN_VARIABLE_PATTERN = /#\{((?!(#{|})).)+\}/, validateForbiddenVariables = (bms) => {
  let checkButtons = (buttons) => {
    for (let button of buttons)
      if (button.name && FORBIDDEN_VARIABLE_PATTERN.test(button.name))
        return "button.name\uC5D0\uB294 \uBCC0\uC218\uB97C \uC0AC\uC6A9\uD560 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4.";
    return true;
  };
  if (bms.buttons) {
    let r = checkButtons(bms.buttons);
    if (r !== true) return r;
  }
  for (let item of bms.carousel?.list ?? []) {
    let r = checkButtons(item.buttons ?? []);
    if (r !== true) return r;
  }
  let tail = bms.carousel?.tail;
  if (tail) {
    let linkKeys = [
      "linkMobile",
      "linkPc",
      "linkAndroid",
      "linkIos"
    ];
    for (let key of linkKeys) {
      let value = tail[key];
      if (value && FORBIDDEN_VARIABLE_PATTERN.test(value))
        return `${key}\uC5D0\uB294 \uBCC0\uC218\uB97C \uC0AC\uC6A9\uD560 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4.`;
    }
  }
  return true;
}, BMS_ACCEPTABLE_FIELDS = {
  TEXT: ["adult", "content", "buttons", "coupon"],
  IMAGE: ["adult", "content", "imageId", "imageLink", "buttons", "coupon"],
  WIDE: ["adult", "content", "imageId", "buttons", "coupon"],
  WIDE_ITEM_LIST: [
    "adult",
    "header",
    "mainWideItem",
    "subWideItemList",
    "buttons",
    "coupon"
  ],
  COMMERCE: [
    "adult",
    "additionalContent",
    "imageId",
    "commerce",
    "buttons",
    "coupon"
  ],
  CAROUSEL_FEED: ["adult", "carousel"],
  CAROUSEL_COMMERCE: ["adult", "additionalContent", "carousel"],
  PREMIUM_VIDEO: ["adult", "header", "content", "video", "buttons", "coupon"]
}, RESERVED_FIELDS = ["targeting", "chatBubbleType"], validateAcceptableFields = (bms) => {
  let type = bms.chatBubbleType, acceptable = BMS_ACCEPTABLE_FIELDS[type], allowed = /* @__PURE__ */ new Set([...acceptable, ...RESERVED_FIELDS]);
  for (let key of Object.keys(bms))
    if (getField(bms, key) != null && !allowed.has(key))
      return `${type}\uD0C0\uC785 \uC5D0\uC11C\uB294 ${acceptable.join(", ")} \uAC12\uB9CC \uC0AC\uC6A9\uC774 \uAC00\uB2A5\uD569\uB2C8\uB2E4.`;
  return true;
}, getField = (obj, key) => obj[key], BMS_IMAGE_ID_MAX = 32, validateImageIdLength = (bms) => bms.chatBubbleType !== "IMAGE" || !bms.imageId ? true : bms.imageId.replace(/\s/g, "").length > BMS_IMAGE_ID_MAX ? `IMAGE \uD0C0\uC785 imageId\uC740 \uCD5C\uB300 ${BMS_IMAGE_ID_MAX}\uC790 \uC774\uD558\uB85C \uC785\uB825\uD574\uC8FC\uC138\uC694.` : true;
var KAKAO_TV_URL_PREFIX = "https://tv.kakao.com/", isKakaoTvUrl = (url) => url.startsWith(KAKAO_TV_URL_PREFIX), bmsVideoSchema = effect.Schema.Struct({
  videoUrl: effect.Schema.String.pipe(
    effect.Schema.filter(isKakaoTvUrl, {
      message: () => `videoUrl\uC740 '${KAKAO_TV_URL_PREFIX}'\uC73C\uB85C \uC2DC\uC791\uD558\uB294 \uCE74\uCE74\uC624TV \uB3D9\uC601\uC0C1 \uB9C1\uD06C\uC5EC\uC57C \uD569\uB2C8\uB2E4.`
    })
  ),
  imageId: effect.Schema.optional(effect.Schema.String),
  imageLink: effect.Schema.optional(effect.Schema.String)
});
var bmsMainWideItemSchema = effect.Schema.Struct({
  title: effect.Schema.optional(effect.Schema.String),
  imageId: effect.Schema.String,
  linkMobile: effect.Schema.String,
  linkPc: effect.Schema.optional(effect.Schema.String),
  linkAndroid: effect.Schema.optional(effect.Schema.String),
  linkIos: effect.Schema.optional(effect.Schema.String)
}), bmsSubWideItemSchema = effect.Schema.Struct({
  title: effect.Schema.String,
  imageId: effect.Schema.String,
  linkMobile: effect.Schema.String,
  linkPc: effect.Schema.optional(effect.Schema.String),
  linkAndroid: effect.Schema.optional(effect.Schema.String),
  linkIos: effect.Schema.optional(effect.Schema.String)
});
var kakaoWebButtonSchema = effect.Schema.Struct({
  buttonName: effect.Schema.String,
  buttonType: effect.Schema.Literal("WL"),
  linkMo: effect.Schema.String,
  linkPc: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String))
}), kakaoAppButtonSchema = effect.Schema.Struct({
  buttonName: effect.Schema.String,
  buttonType: effect.Schema.Literal("AL"),
  linkAnd: effect.Schema.String,
  linkIos: effect.Schema.String
}), kakaoDefaultButtonSchema = effect.Schema.Struct({
  buttonName: effect.Schema.String,
  buttonType: effect.Schema.Literal("BK", "MD", "DS", "BC", "BT", "AC")
}), kakaoButtonSchema = effect.Schema.Union(
  kakaoWebButtonSchema,
  kakaoAppButtonSchema,
  kakaoDefaultButtonSchema
);

// src/models/base/kakao/kakaoOption.ts
var VariableValidationError = class extends effect.Data.TaggedError(
  "VariableValidationError"
) {
  get message() {
    return `\uBCC0\uC218\uBA85 ${this.invalidVariables.map((v) => `\`${v}\``).join(", ")}\uC5D0 \uC810(.)\uC744 \uD3EC\uD568\uD560 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4. \uC5B8\uB354\uC2A4\uCF54\uC5B4(_)\uB098 \uB2E4\uB978 \uBB38\uC790\uB97C \uC0AC\uC6A9\uD574\uC8FC\uC138\uC694.`;
  }
  toString() {
    return `VariableValidationError: ${this.message}`;
  }
}, BMS_REQUIRED_FIELDS = {
  TEXT: [],
  IMAGE: ["imageId"],
  WIDE: ["imageId"],
  WIDE_ITEM_LIST: ["header", "mainWideItem", "subWideItemList"],
  COMMERCE: ["imageId", "commerce", "buttons"],
  CAROUSEL_FEED: ["carousel"],
  CAROUSEL_COMMERCE: ["carousel"],
  PREMIUM_VIDEO: ["video"]
}, bmsCarouselSchema = effect.Schema.Union(
  bmsCarouselFeedSchema,
  bmsCarouselCommerceSchema
), baseBmsSchema = effect.Schema.Struct({
  // 필수 필드
  targeting: effect.Schema.Literal("I", "M", "N"),
  chatBubbleType: bmsChatBubbleTypeSchema,
  // 선택 필드
  adult: effect.Schema.optional(effect.Schema.Boolean),
  header: effect.Schema.optional(effect.Schema.String),
  imageId: effect.Schema.optional(effect.Schema.String),
  imageLink: effect.Schema.optional(effect.Schema.String),
  additionalContent: effect.Schema.optional(effect.Schema.String),
  content: effect.Schema.optional(effect.Schema.String),
  // 복합 타입 필드
  carousel: effect.Schema.optional(bmsCarouselSchema),
  mainWideItem: effect.Schema.optional(bmsMainWideItemSchema),
  subWideItemList: effect.Schema.optional(effect.Schema.Array(bmsSubWideItemSchema)),
  buttons: effect.Schema.optional(effect.Schema.Array(bmsButtonSchema)),
  coupon: effect.Schema.optional(bmsCouponSchema),
  commerce: effect.Schema.optional(bmsCommerceSchema),
  video: effect.Schema.optional(bmsVideoSchema)
}), WIDE_ITEM_LIST_MIN_SUB_ITEMS = 3, validateBmsRequiredFields = (bms) => {
  let chatBubbleType = bms.chatBubbleType, missingFields = (BMS_REQUIRED_FIELDS[chatBubbleType] ?? []).filter(
    (field) => bms[field] === void 0 || bms[field] === null
  );
  if (missingFields.length > 0)
    return `BMS ${chatBubbleType} \uD0C0\uC785\uC5D0 \uD544\uC218 \uD544\uB4DC\uAC00 \uB204\uB77D\uB418\uC5C8\uC2B5\uB2C8\uB2E4: ${missingFields.join(", ")}`;
  if (chatBubbleType === "WIDE_ITEM_LIST") {
    let subWideItemList = bms.subWideItemList;
    if (!subWideItemList || subWideItemList.length < WIDE_ITEM_LIST_MIN_SUB_ITEMS)
      return `WIDE_ITEM_LIST \uD0C0\uC785\uC758 subWideItemList\uB294 \uCD5C\uC18C ${WIDE_ITEM_LIST_MIN_SUB_ITEMS}\uAC1C \uC774\uC0C1\uC774\uC5B4\uC57C \uD569\uB2C8\uB2E4. \uD604\uC7AC: ${subWideItemList?.length ?? 0}\uAC1C`;
  }
  return true;
}, validateBmsConstraints = (bms) => {
  let input = bms, validators = [
    validateAcceptableFields,
    validateCarouselListCount,
    validateCouponDescription,
    validateAllowedLinkTypes,
    validateButtonCount,
    validateButtonNames,
    validateImageIdLength,
    validateTextLengths,
    validateNewlineLimits,
    validateForbiddenVariables,
    validateLinks
  ], requiredResult = validateBmsRequiredFields(bms);
  if (requiredResult !== true) return requiredResult;
  for (let validator of validators) {
    let result = validator(input);
    if (result !== true) return result;
  }
  return true;
}, kakaoOptionBmsSchema = baseBmsSchema.pipe(
  effect.Schema.filter(validateBmsConstraints)
), VARIABLE_KEY_PATTERN = /^#\{.+}$/, DOT_PATTERN = /\./, extractVariableName = (key) => VARIABLE_KEY_PATTERN.test(key) ? key.slice(2, -1) : key, formatVariableKey = (key) => VARIABLE_KEY_PATTERN.test(key) ? key : `#{${key}}`, validateVariableNames = (variables) => effect.pipe(
  Object.keys(variables),
  effect.Array.map(extractVariableName),
  effect.Array.filter((variableName) => DOT_PATTERN.test(variableName)),
  (invalidVariables) => invalidVariables.length > 0 ? effect.Effect.fail(new VariableValidationError({ invalidVariables })) : effect.Effect.succeed(variables)
), transformVariables = (variables) => effect.pipe(
  validateVariableNames(variables),
  effect.Effect.map(
    (validVariables) => effect.pipe(
      Object.entries(validVariables),
      effect.Array.map(
        ([key, value]) => [formatVariableKey(key), value]
      ),
      (entries) => Object.fromEntries(entries)
    )
  )
), baseKakaoOptionSchema = effect.Schema.Struct({
  pfId: effect.Schema.String,
  templateId: effect.Schema.optional(effect.Schema.String),
  variables: effect.Schema.optional(
    effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.String }).pipe(
      effect.Schema.transformOrFail(
        effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.String }),
        {
          decode: (fromU, _, ast) => transformVariables(fromU).pipe(
            effect.Effect.mapError(
              (err) => new effect.ParseResult.Type(ast, fromU, err.message)
            )
          ),
          encode: (toI) => effect.ParseResult.succeed(toI)
        }
      )
    )
  ),
  disableSms: effect.Schema.optional(effect.Schema.Boolean),
  adFlag: effect.Schema.optional(effect.Schema.Boolean),
  imageId: effect.Schema.optional(effect.Schema.String),
  buttons: effect.Schema.optional(effect.Schema.Array(kakaoButtonSchema)),
  bms: effect.Schema.optional(kakaoOptionBmsSchema)
});
var naverOptionButtonSchema = effect.Schema.Struct({
  buttonName: effect.Schema.String,
  buttonType: effect.Schema.String,
  linkMo: effect.Schema.optional(effect.Schema.String),
  linkPc: effect.Schema.optional(effect.Schema.String),
  linkAnd: effect.Schema.optional(effect.Schema.String),
  linkIos: effect.Schema.optional(effect.Schema.String)
}), naverOptionSchema = effect.Schema.Struct({
  talkId: effect.Schema.String,
  templateId: effect.Schema.String,
  disableSms: effect.Schema.optional(effect.Schema.Boolean),
  variables: effect.Schema.optional(
    effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.String })
  ),
  buttons: effect.Schema.optional(effect.Schema.Array(naverOptionButtonSchema))
});
effect.Schema.Literal(
  "WL",
  "ML",
  "MQ",
  "MR",
  "CA",
  "CL",
  "DL",
  "MS"
); var rcsWebButtonSchema = effect.Schema.Struct({
  buttonName: effect.Schema.String,
  buttonType: effect.Schema.Literal("WL"),
  link: effect.Schema.String
}), rcsMapButtonSchema = effect.Schema.Struct({
  buttonName: effect.Schema.String,
  buttonType: effect.Schema.Literal("ML"),
  latitude: effect.Schema.String,
  longitude: effect.Schema.String
}), rcsDefaultButtonSchema = effect.Schema.Struct({
  buttonName: effect.Schema.String,
  buttonType: effect.Schema.Literal("MQ", "MR", "CA", "CL", "DL", "MS"),
  link: effect.Schema.String
}), rcsButtonSchema = effect.Schema.Union(
  rcsWebButtonSchema,
  rcsMapButtonSchema,
  rcsDefaultButtonSchema
);

// src/models/base/rcs/rcsOption.ts
var additionalBodySchema = effect.Schema.Struct({
  title: effect.Schema.String,
  description: effect.Schema.String,
  imaggeId: effect.Schema.optional(effect.Schema.String),
  buttons: effect.Schema.optional(effect.Schema.Array(rcsButtonSchema))
}), rcsOptionRequestSchema = effect.Schema.Struct({
  brandId: effect.Schema.String,
  templateId: effect.Schema.optional(effect.Schema.String),
  copyAllowed: effect.Schema.optional(effect.Schema.Boolean),
  variables: effect.Schema.optional(
    effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.String })
  ),
  mmsType: effect.Schema.optional(
    effect.Schema.Literal("M3", "S3", "M4", "S4", "M5", "S5", "M6", "S6")
  ),
  commercialType: effect.Schema.optional(effect.Schema.Boolean),
  disableSms: effect.Schema.optional(effect.Schema.Boolean),
  additionalBody: effect.Schema.optional(additionalBodySchema),
  buttons: effect.Schema.optional(effect.Schema.Array(rcsButtonSchema))
}), rcsOptionSchema = rcsOptionRequestSchema;
var voiceOptionSchema = effect.Schema.Struct({
  voiceType: effect.Schema.Literal("FEMALE", "MALE"),
  headerMessage: effect.Schema.optional(effect.Schema.String),
  tailMessage: effect.Schema.optional(effect.Schema.String),
  replyRange: effect.Schema.optional(effect.Schema.Literal(1, 2, 3, 4, 5, 6, 7, 8, 9)),
  counselorNumber: effect.Schema.optional(effect.Schema.String)
});
var messageTypeSchema = effect.Schema.Literal(
  "SMS",
  "LMS",
  "MMS",
  "ATA",
  "CTA",
  "CTI",
  "NSA",
  "RCS_SMS",
  "RCS_LMS",
  "RCS_MMS",
  "RCS_TPL",
  "RCS_ITPL",
  "RCS_LTPL",
  "FAX",
  "VOICE",
  "BMS_TEXT",
  "BMS_IMAGE",
  "BMS_WIDE",
  "BMS_WIDE_ITEM_LIST",
  "BMS_CAROUSEL_FEED",
  "BMS_PREMIUM_VIDEO",
  "BMS_COMMERCE",
  "BMS_CAROUSEL_COMMERCE",
  "BMS_FREE"
), messageSchema = effect.Schema.Struct({
  to: effect.Schema.Union(effect.Schema.String, effect.Schema.Array(effect.Schema.String)),
  from: effect.Schema.optional(effect.Schema.String),
  text: effect.Schema.optional(effect.Schema.String),
  imageId: effect.Schema.optional(effect.Schema.String),
  type: effect.Schema.optional(messageTypeSchema),
  subject: effect.Schema.optional(effect.Schema.String),
  autoTypeDetect: effect.Schema.optional(effect.Schema.Boolean),
  kakaoOptions: effect.Schema.optional(baseKakaoOptionSchema),
  rcsOptions: effect.Schema.optional(rcsOptionSchema),
  country: effect.Schema.optional(effect.Schema.String),
  replacements: effect.Schema.optional(effect.Schema.Array(effect.Schema.Struct({}))),
  customFields: effect.Schema.optional(
    effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.String })
  ),
  naverOptions: effect.Schema.optional(naverOptionSchema),
  faxOptions: effect.Schema.optional(
    effect.Schema.Struct({ fileIds: effect.Schema.Array(effect.Schema.String) })
  ),
  voiceOptions: effect.Schema.optional(voiceOptionSchema)
});

// src/models/base/messages/storedMessage.ts
var booleanOrZeroOne = effect.Schema.transformOrFail(
  effect.Schema.Union(effect.Schema.Boolean, effect.Schema.Number),
  effect.Schema.Boolean,
  {
    decode: (value, _opts, ast) => typeof value == "boolean" ? effect.ParseResult.succeed(value) : value === 0 ? effect.ParseResult.succeed(false) : value === 1 ? effect.ParseResult.succeed(true) : effect.ParseResult.fail(
      new effect.ParseResult.Type(
        ast,
        value,
        `Expected boolean, 0, or 1 but received ${String(value)}`
      )
    ),
    encode: (value) => effect.ParseResult.succeed(value),
    strict: true
  }
), storedMessageSchema = effect.Schema.Struct({
  messageId: effect.Schema.optional(effect.Schema.String),
  type: effect.Schema.NullishOr(messageTypeSchema),
  to: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.Array(effect.Schema.String))),
  from: effect.Schema.NullishOr(effect.Schema.String),
  text: effect.Schema.NullishOr(effect.Schema.String),
  imageId: effect.Schema.NullishOr(effect.Schema.String),
  subject: effect.Schema.NullishOr(effect.Schema.String),
  country: effect.Schema.NullishOr(effect.Schema.String),
  accountId: effect.Schema.optional(effect.Schema.String),
  groupId: effect.Schema.optional(effect.Schema.String),
  status: effect.Schema.NullishOr(effect.Schema.String),
  statusCode: effect.Schema.NullishOr(effect.Schema.String),
  reason: effect.Schema.NullishOr(effect.Schema.String),
  networkName: effect.Schema.NullishOr(effect.Schema.String),
  networkCode: effect.Schema.NullishOr(effect.Schema.String),
  customFields: effect.Schema.optional(
    effect.Schema.NullishOr(effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.String }))
  ),
  autoTypeDetect: effect.Schema.optional(booleanOrZeroOne),
  replacement: effect.Schema.optional(booleanOrZeroOne),
  resendCount: effect.Schema.optional(effect.Schema.Number),
  dateCreated: effect.Schema.optional(effect.Schema.String),
  dateUpdated: effect.Schema.optional(effect.Schema.String),
  dateProcessed: effect.Schema.NullishOr(effect.Schema.String),
  dateReceived: effect.Schema.NullishOr(effect.Schema.String),
  dateReported: effect.Schema.NullishOr(effect.Schema.String),
  // 옵션 객체는 서버 정규화 포맷(저장 형태)으로 발송 요청용 스키마와 필드가 다르다.
  // 상세 타이핑을 확정하려면 각 옵션별 별도 조회 스키마 정의가 필요하지만 본 PR 범위를
  // 벗어나므로, 최소한 "object"임을 보장해 원시 값이 섞이는 drift를 감지할 수 있게 한다.
  kakaoOptions: effect.Schema.optional(
    effect.Schema.NullishOr(
      effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.Unknown })
    )
  ),
  rcsOptions: effect.Schema.optional(
    effect.Schema.NullishOr(
      effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.Unknown })
    )
  ),
  naverOptions: effect.Schema.optional(
    effect.Schema.NullishOr(
      effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.Unknown })
    )
  ),
  faxOptions: effect.Schema.optional(
    effect.Schema.NullishOr(
      effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.Unknown })
    )
  ),
  voiceOptions: effect.Schema.optional(
    effect.Schema.NullishOr(
      effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.Unknown })
    )
  ),
  replacements: effect.Schema.optional(effect.Schema.NullishOr(effect.Schema.Array(effect.Schema.Unknown))),
  log: effect.Schema.optional(effect.Schema.NullishOr(effect.Schema.Array(effect.Schema.Unknown))),
  queues: effect.Schema.optional(effect.Schema.NullishOr(effect.Schema.Array(effect.Schema.Unknown))),
  currentQueue: effect.Schema.optional(effect.Schema.NullishOr(effect.Schema.Unknown)),
  clusterKey: effect.Schema.NullishOr(effect.Schema.String),
  unavailableSenderNumber: effect.Schema.optional(effect.Schema.NullishOr(booleanOrZeroOne)),
  faxPageCount: effect.Schema.optional(effect.Schema.NullishOr(effect.Schema.Number)),
  voiceDuration: effect.Schema.optional(effect.Schema.NullishOr(effect.Schema.Number)),
  voiceReplied: effect.Schema.optional(effect.Schema.NullishOr(booleanOrZeroOne)),
  _id: effect.Schema.optional(effect.Schema.String)
});

// src/models/responses/messageResponses.ts
var groupMessageResponseSchema = effect.Schema.Struct({
  count: countSchema,
  countForCharge: countForChargeSchema,
  balance: commonCashResponseSchema,
  point: commonCashResponseSchema,
  app: appSchema,
  log: logSchema,
  status: effect.Schema.String,
  allowDuplicates: effect.Schema.Boolean,
  isRefunded: effect.Schema.Boolean,
  accountId: effect.Schema.String,
  masterAccountId: effect.Schema.NullOr(effect.Schema.String),
  apiVersion: effect.Schema.String,
  groupId: effect.Schema.String,
  price: effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.Unknown }),
  dateCreated: effect.Schema.String,
  dateUpdated: effect.Schema.String,
  scheduledDate: effect.Schema.NullishOr(effect.Schema.String),
  dateSent: effect.Schema.NullishOr(effect.Schema.String),
  dateCompleted: effect.Schema.NullishOr(effect.Schema.String)
}), addMessageResultSchema = effect.Schema.Struct({
  to: effect.Schema.String,
  from: effect.Schema.String,
  type: effect.Schema.String,
  country: effect.Schema.String,
  messageId: effect.Schema.String,
  statusCode: effect.Schema.String,
  statusMessage: effect.Schema.String,
  accountId: effect.Schema.String,
  customFields: effect.Schema.optional(
    effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.String })
  )
}), addMessageResponseSchema = effect.Schema.Struct({
  errorCount: effect.Schema.String,
  resultList: effect.Schema.Array(addMessageResultSchema)
}), getMessagesResponseSchema = effect.Schema.Struct({
  startKey: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  nextKey: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  limit: effect.Schema.Number,
  messageList: effect.Schema.Record({ key: effect.Schema.String, value: storedMessageSchema })
}), removeGroupMessagesResponseSchema = effect.Schema.Struct({
  groupId: groupIdSchema,
  errorCount: effect.Schema.Number,
  resultList: effect.Schema.Array(
    effect.Schema.Struct({
      messageId: effect.Schema.String,
      resultCode: effect.Schema.String
    })
  )
}), getGroupsResponseSchema = effect.Schema.Struct({
  startKey: effect.Schema.NullishOr(effect.Schema.String),
  limit: effect.Schema.Number,
  nextKey: effect.Schema.NullishOr(effect.Schema.String),
  groupList: effect.Schema.Record({ key: groupIdSchema, value: groupSchema })
}), statisticsPeriodResultSchema = effect.Schema.Struct({
  total: effect.Schema.Number,
  sms: effect.Schema.Number,
  lms: effect.Schema.Number,
  mms: effect.Schema.Number,
  ata: effect.Schema.Number,
  cta: effect.Schema.Number,
  cti: effect.Schema.Number,
  nsa: effect.Schema.Number,
  rcs_sms: effect.Schema.Number,
  rcs_lms: effect.Schema.Number,
  rcs_mms: effect.Schema.Number,
  rcs_tpl: effect.Schema.Number,
  rcs_itpl: effect.Schema.optional(effect.Schema.Number),
  rcs_ltpl: effect.Schema.optional(effect.Schema.Number),
  fax: effect.Schema.optional(effect.Schema.Number),
  voice: effect.Schema.optional(effect.Schema.Number),
  bms_text: effect.Schema.optional(effect.Schema.Number),
  bms_image: effect.Schema.optional(effect.Schema.Number),
  bms_wide: effect.Schema.optional(effect.Schema.Number),
  bms_wide_item_list: effect.Schema.optional(effect.Schema.Number),
  bms_carousel_feed: effect.Schema.optional(effect.Schema.Number),
  bms_premium_video: effect.Schema.optional(effect.Schema.Number),
  bms_commerce: effect.Schema.optional(effect.Schema.Number),
  bms_carousel_commerce: effect.Schema.optional(effect.Schema.Number),
  bms_free: effect.Schema.optional(effect.Schema.Number)
}), refundSchema = effect.Schema.Struct({
  balance: effect.Schema.Number,
  point: effect.Schema.Number,
  deposit: effect.Schema.optional(effect.Schema.Number)
}), dayPeriodSchema = effect.Schema.Struct({
  _id: effect.Schema.String,
  month: effect.Schema.String,
  date: effect.Schema.optional(effect.Schema.String),
  balance: effect.Schema.Number,
  point: effect.Schema.Number,
  deposit: effect.Schema.optional(effect.Schema.Number),
  statusCode: effect.Schema.Record({
    key: effect.Schema.String,
    value: partialMessageTypeRecordSchema
  }),
  refund: refundSchema,
  total: statisticsPeriodResultSchema,
  successed: statisticsPeriodResultSchema,
  failed: statisticsPeriodResultSchema
}), monthPeriodRefundSchema = effect.Schema.Struct({
  balance: effect.Schema.Number,
  balanceAvg: effect.Schema.Number,
  point: effect.Schema.Number,
  pointAvg: effect.Schema.Number,
  deposit: effect.Schema.optional(effect.Schema.Number),
  depositAvg: effect.Schema.optional(effect.Schema.Number)
}), monthPeriodSchema = effect.Schema.Struct({
  date: effect.Schema.String,
  balance: effect.Schema.Number,
  balanceAvg: effect.Schema.Number,
  point: effect.Schema.Number,
  pointAvg: effect.Schema.Number,
  deposit: effect.Schema.optional(effect.Schema.Number),
  depositAvg: effect.Schema.optional(effect.Schema.Number),
  dayPeriod: effect.Schema.Array(dayPeriodSchema),
  refund: effect.Schema.optional(monthPeriodRefundSchema),
  total: statisticsPeriodResultSchema,
  successed: statisticsPeriodResultSchema,
  failed: statisticsPeriodResultSchema
}), getStatisticsResponseSchema = effect.Schema.Struct({
  balance: effect.Schema.Number,
  point: effect.Schema.Number,
  deposit: effect.Schema.optional(effect.Schema.Number),
  monthlyBalanceAvg: effect.Schema.Number,
  monthlyPointAvg: effect.Schema.Number,
  monthlyDepositAvg: effect.Schema.optional(effect.Schema.Number),
  monthPeriod: effect.Schema.Array(monthPeriodSchema),
  total: statisticsPeriodResultSchema,
  successed: statisticsPeriodResultSchema,
  failed: statisticsPeriodResultSchema,
  dailyBalanceAvg: effect.Schema.optional(effect.Schema.Number),
  dailyPointAvg: effect.Schema.optional(effect.Schema.Number),
  dailyTotalCountAvg: effect.Schema.optional(effect.Schema.Number),
  dailyFailedCountAvg: effect.Schema.optional(effect.Schema.Number),
  dailySuccessedCountAvg: effect.Schema.optional(effect.Schema.Number)
}), lowBalanceAlertSchema = effect.Schema.Struct({
  notificationBalance: effect.Schema.String,
  currentBalance: effect.Schema.String,
  balances: effect.Schema.Array(effect.Schema.Number),
  channels: effect.Schema.Array(effect.Schema.String),
  enabled: effect.Schema.Boolean
}), getBalanceResponseSchema = effect.Schema.Struct({
  lowBalanceAlert: effect.Schema.optional(lowBalanceAlertSchema),
  point: effect.Schema.Number,
  minimumCash: effect.Schema.optional(effect.Schema.Number),
  rechargeTo: effect.Schema.optional(effect.Schema.Number),
  rechargeTryCount: effect.Schema.optional(effect.Schema.Number),
  autoRecharge: effect.Schema.optional(effect.Schema.Number),
  accountId: effect.Schema.optional(effect.Schema.String),
  balance: effect.Schema.Number,
  deposit: effect.Schema.optional(effect.Schema.Number),
  balanceOnly: effect.Schema.optional(effect.Schema.Number)
}), fileUploadResponseSchema = effect.Schema.Struct({
  fileId: effect.Schema.String,
  type: effect.Schema.String,
  link: effect.Schema.NullishOr(effect.Schema.String)
}), requestKakaoChannelTokenResponseSchema = effect.Schema.Struct({
  success: effect.Schema.Boolean
}), createKakaoChannelResponseSchema = effect.Schema.Struct({
  accountId: effect.Schema.String,
  phoneNumber: effect.Schema.String,
  searchId: effect.Schema.String,
  dateCreated: effect.Schema.String,
  dateUpdated: effect.Schema.String,
  channelId: effect.Schema.String
});
function genCustomText(alphabet, size) {
  let bytes = crypto.randomBytes(size), genRandomText = "";
  for (let i = 0; i < size; i++)
    genRandomText += alphabet[bytes[i] % alphabet.length];
  return genRandomText;
}
function getAuthInfo(authenticationParameter, authType = 0 /* API_KEY */) {
  let { apiKey, apiSecret } = authenticationParameter;
  return !apiKey || !apiSecret ? effect.Effect.fail(
    new ApiKeyError({ message: "API Key\uC640 API Secret\uC740 \uD544\uC218\uC785\uB2C8\uB2E4." })
  ) : effect.Effect.try({
    try: () => {
      let salt = genCustomText(
        "1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ",
        32
      ), date = dateFns.formatISO(/* @__PURE__ */ new Date()), hmacData = date + salt, genHmac = crypto.createHmac("sha256", apiSecret);
      genHmac.update(hmacData);
      let signature = genHmac.digest("hex");
      return `HMAC-SHA256 apiKey=${apiKey}, date=${date}, salt=${salt}, signature=${signature}`;
    },
    catch: (e) => new DefaultError({
      errorCode: "AuthError",
      errorMessage: e instanceof Error ? e.message : String(e)
    })
  });
}

// src/lib/defaultFetcher.ts
var ERROR_MESSAGE_PREVIEW_LENGTH = 200, RETRYABLE_ERROR_KEYWORDS = [
  "aborted",
  "refused",
  "reset",
  "econn"
], RetryableError = class extends effect.Data.TaggedError("RetryableError") {
}, toMessage = (e) => e instanceof Error ? e.message : String(e), isRetryableNetworkError = (error) => {
  let cause = error.cause, causeCode = cause && typeof cause == "object" && "code" in cause ? String(cause.code) : "", message = `${error.message} ${causeCode}`.toLowerCase();
  return RETRYABLE_ERROR_KEYWORDS.some((keyword) => message.includes(keyword));
}, makeParseError = (res, message) => new DefaultError({
  errorCode: "ParseError",
  errorMessage: message,
  context: { responseStatus: res.status, responseUrl: res.url }
}), handleOkResponse = (res) => effect.pipe(
  effect.Effect.tryPromise({
    try: () => res.text(),
    catch: (e) => makeParseError(res, toMessage(e))
  }),
  effect.Effect.flatMap((responseText) => responseText ? effect.Effect.try({
    try: () => JSON.parse(responseText),
    catch: (e) => makeParseError(res, toMessage(e))
  }) : res.status === 204 ? effect.Effect.succeed({}) : effect.Effect.fail(
    makeParseError(res, "API returned empty response body")
  ))
), handleClientErrorResponse = (res) => effect.pipe(
  effect.Effect.tryPromise({
    try: () => res.text(),
    catch: (e) => makeParseError(res, toMessage(e))
  }),
  effect.Effect.flatMap((text) => {
    let genericError = new ClientError({
      errorCode: `HTTP_${res.status}`,
      errorMessage: text.substring(0, ERROR_MESSAGE_PREVIEW_LENGTH) || "Client error occurred",
      httpStatus: res.status,
      url: res.url
    });
    return effect.Effect.flatMap(
      effect.Effect.try({
        try: () => JSON.parse(text),
        catch: (e) => e instanceof SyntaxError ? genericError : new ClientError({
          errorCode: "ResponseParseError",
          errorMessage: toMessage(e),
          httpStatus: res.status,
          url: res.url
        })
      }),
      (json) => effect.Effect.fail(
        isErrorResponse(json) ? new ClientError({
          errorCode: json.errorCode,
          errorMessage: json.errorMessage,
          httpStatus: res.status,
          url: res.url
        }) : genericError
      )
    );
  })
);
function parseServerErrorBody(text, genericError, makeError) {
  return effect.Effect.flatMap(
    effect.Effect.try({
      try: () => JSON.parse(text),
      catch: (e) => e instanceof SyntaxError ? genericError : makeError("ResponseParseError", toMessage(e))
    }),
    (json) => effect.Effect.fail(
      isErrorResponse(json) ? makeError(json.errorCode, json.errorMessage) : genericError
    )
  );
}
var handleServerErrorResponse = (res) => effect.pipe(
  effect.Effect.tryPromise({
    try: () => res.text(),
    catch: (e) => new DefaultError({
      errorCode: "ResponseReadError",
      errorMessage: toMessage(e),
      context: { responseStatus: res.status, responseUrl: res.url }
    })
  }),
  effect.Effect.flatMap((text) => {
    let makeError = (errorCode, errorMessage) => new ServerError({
      errorCode,
      errorMessage,
      httpStatus: res.status,
      url: res.url,
      responseBody: void 0
    }), genericError = makeError(
      `HTTP_${res.status}`,
      text.substring(0, ERROR_MESSAGE_PREVIEW_LENGTH) || "Server error occurred"
    );
    return parseServerErrorBody(text, genericError, makeError);
  })
);
function defaultFetcherEffect(authParameter, request, data) {
  let effect$1 = effect.Effect.gen(function* () {
    let authorizationHeaderData = yield* getAuthInfo(authParameter), body = yield* effect.Effect.try({
      try: () => data ? JSON.stringify(data) : void 0,
      catch: (e) => new DefaultError({
        errorCode: "JSONStringifyError",
        errorMessage: toMessage(e),
        context: { data }
      })
    }), response = yield* effect.Effect.tryPromise({
      try: () => fetch(request.url, {
        headers: {
          Authorization: authorizationHeaderData,
          "Content-Type": "application/json"
        },
        body,
        method: request.method
      }),
      catch: (error) => error instanceof Error ? isRetryableNetworkError(error) ? new RetryableError({ error }) : new NetworkError({
        url: request.url,
        method: request.method,
        cause: error.message,
        isRetryable: false
      }) : new NetworkError({
        url: request.url,
        method: request.method,
        cause: String(error),
        isRetryable: false
      })
    });
    return yield* effect.pipe(
      effect.Match.value(response),
      effect.Match.when(
        (res) => res.status === 503,
        () => effect.Effect.fail(new RetryableError({ error: "Service Unavailable" }))
      ),
      effect.Match.when(
        (res) => res.status >= 400 && res.status < 500,
        handleClientErrorResponse
      ),
      effect.Match.when((res) => !res.ok, handleServerErrorResponse),
      effect.Match.orElse(handleOkResponse)
    );
  }), retryCount = 3, policy = effect.pipe(
    effect.Schedule.recurs(retryCount),
    effect.Schedule.whileInput(
      (e) => e instanceof RetryableError
    )
  );
  return effect.pipe(
    effect$1,
    effect.Effect.retry(policy),
    effect.Effect.catchTag(
      "RetryableError",
      () => effect.Effect.fail(
        new DefaultError({
          errorCode: "RequestFailedAfterRetryError",
          errorMessage: `Request failed after retry(count: ${retryCount})`,
          context: {
            url: request.url,
            method: request.method,
            retryCount
          }
        })
      )
    )
  );
}
function formatWithTransfer(value) {
  return dateFns.formatISO(stringDateTransfer(value));
}
function stringDateTransfer(value) {
  if (typeof value == "string") {
    let originalString = value;
    value = dateFns.parseISO(originalString);
    let invalidDateText = "Invalid Date";
    if (value.toString() === invalidDateText)
      throw new InvalidDateError({
        message: invalidDateText,
        originalValue: originalString
      });
  }
  return value;
}

// src/lib/schemaUtils.ts
var decodeWithBadRequest = (schema, data) => Effect10__namespace.mapError(
  effect.Schema.decodeUnknown(schema)(data),
  (error) => new BadRequestError({
    message: effect.ParseResult.TreeFormatter.formatErrorSync(error),
    validationErrors: effect.ParseResult.ArrayFormatter.formatErrorSync(error).map(
      (issue) => `${issue.path.length > 0 ? issue.path.join(".") : "(root)"}: ${issue.message}`
    )
  })
), safeDateTransfer = (value) => value != null ? Effect10__namespace.try({
  try: () => stringDateTransfer(value),
  catch: (error) => error instanceof InvalidDateError ? error : new InvalidDateError({
    message: error instanceof Error ? error.message : String(error)
  })
}) : Effect10__namespace.succeed(void 0), safeFormatWithTransfer = (value) => Effect10__namespace.try({
  try: () => formatWithTransfer(value),
  catch: (error) => error instanceof InvalidDateError ? error : new InvalidDateError({
    message: error instanceof Error ? error.message : String(error)
  })
}), safeFinalize = (fn) => Effect10__namespace.try({
  try: fn,
  catch: (error) => error instanceof InvalidDateError ? error : new BadRequestError({
    message: error instanceof Error ? error.message : String(error)
  })
}), stringifyResponseBody = (data) => {
  if (data !== void 0) {
    if (typeof data == "string") return data;
    try {
      return JSON.stringify(data);
    } catch (err) {
      return `[unserializable: ${err instanceof Error ? err.message : String(err)}] ${Object.prototype.toString.call(data)}`;
    }
  }
}, redactUrlForProduction = (url) => {
  if (!url) return url;
  try {
    let parsed = new URL(url), hadQuery = parsed.search.length > 0;
    return parsed.search = hadQuery ? "?[redacted]" : "", parsed.hash = "", parsed.username = "", parsed.password = "", parsed.toString();
  } catch {
    let cut = url.search(/[?#;]/);
    return cut === -1 ? url : `${url.slice(0, cut)}?[redacted]`;
  }
}, shouldRedactSensitive = () => {
  let env = "production"?.trim().toLowerCase();
  return env !== "development" && env !== "test";
}, decodeServerResponse = (schema, data, context) => (
  // onExcessProperty: 'preserve' — 서버가 추가로 내려준 미선언 필드를 strip 하지 않는다.
  // 부분 스키마로 검증하는 조회 엔드포인트에서 필드 조용히 사라지는 silent data loss를 방지.
  Effect10__namespace.mapError(
    effect.Schema.decodeUnknown(schema, { onExcessProperty: "preserve" })(data),
    (err) => {
      let redact = shouldRedactSensitive(), issues = effect.ParseResult.ArrayFormatter.formatErrorSync(err);
      return new ResponseSchemaMismatchError({
        message: redact ? `Response schema mismatch on ${issues.length} field(s)` : effect.ParseResult.TreeFormatter.formatErrorSync(err),
        validationErrors: issues.map((issue) => {
          let path = issue.path.length > 0 ? issue.path.join(".") : "(root)";
          return redact ? `${path}: [${issue._tag}]` : `${path}: ${issue.message}`;
        }),
        url: redact ? redactUrlForProduction(context?.url) : context?.url,
        responseBody: redact ? void 0 : stringifyResponseBody(data)
      });
    }
  )
);

// src/lib/stringifyQuery.ts
function stringifyQuery(obj, options = {}) {
  if (!obj || typeof obj != "object" || Object.keys(obj).length === 0)
    return "";
  let processValue = (key, value) => {
    if (Array.isArray(value))
      return options.indices === false ? value.map(
        (item) => `${encodeURIComponent(key)}=${encodeURIComponent(String(item))}`
      ) : value.map(
        (item, index) => `${encodeURIComponent(key)}[${index}]=${encodeURIComponent(String(item))}`
      );
    if (value == null)
      return [];
    if (typeof value == "object") {
      let nested = [];
      for (let [subKey, subValue] of Object.entries(value))
        nested.push(...processValue(`${key}[${subKey}]`, subValue));
      return nested;
    }
    return [`${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`];
  }, pairs = [];
  for (let [key, value] of Object.entries(obj))
    pairs.push(...processValue(key, value));
  let queryString = pairs.join("&");
  return queryString ? options.addQueryPrefix === false ? queryString : `?${queryString}` : "";
}
var DefaultService = class {
  baseUrl = "https://api.solapi.com";
  authInfo;
  constructor(apiKey, apiSecret) {
    this.authInfo = {
      apiKey,
      apiSecret
    };
  }
  requestEffect(parameter) {
    let { httpMethod, url, body, responseSchema } = parameter, requestConfig = {
      method: httpMethod,
      url: `${this.baseUrl}/${url}`
    };
    return responseSchema ? Effect10__namespace.flatMap(
      defaultFetcherEffect(this.authInfo, requestConfig, body),
      (data) => decodeServerResponse(responseSchema, data, { url: requestConfig.url })
    ) : defaultFetcherEffect(this.authInfo, requestConfig, body);
  }
  async request(parameter) {
    return runSafePromise(this.requestEffect(parameter));
  }
  getWithQuery(config) {
    let reqEffect = this.requestEffect.bind(this);
    return Effect10__namespace.gen(function* () {
      let validated = config.data ? yield* decodeWithBadRequest(config.schema, config.data) : void 0, payload = yield* safeFinalize(() => config.finalize(validated)), parameter = stringifyQuery(payload, {
        indices: false,
        addQueryPrefix: true
      });
      return yield* reqEffect({
        httpMethod: "GET",
        url: `${config.url}${parameter}`,
        responseSchema: config.responseSchema
      });
    });
  }
};

// src/services/cash/cashService.ts
var CashService = class extends DefaultService {
  /**
   * 잔액조회
   * @returns GetBalanceResponse
   */
  async getBalance() {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "GET",
        url: "cash/v1/balance",
        responseSchema: getBalanceResponseSchema
      })
    );
  }
};
var getBlacksRequestSchema = effect.Schema.Struct({
  senderNumber: effect.Schema.optional(effect.Schema.String),
  startKey: effect.Schema.optional(effect.Schema.String),
  limit: effect.Schema.optional(effect.Schema.Number),
  startDate: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)),
  endDate: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf))
});
function finalizeGetBlacksRequest(data) {
  if (!data) return { type: "DENIAL" };
  let payload = { type: "DENIAL" };
  return payload.senderNumber = data.senderNumber, payload.startKey = data.startKey, payload.limit = data.limit, data.startDate != null && (payload.dateCreated = Object.assign(payload.dateCreated ?? {}, {
    gte: formatWithTransfer(data.startDate)
  })), data.endDate != null && (payload.dateCreated = Object.assign(payload.dateCreated ?? {}, {
    lte: formatWithTransfer(data.endDate)
  })), payload;
}
var likeLiteralSchema = effect.Schema.Struct({ like: effect.Schema.String }), datePayloadSchema = effect.Schema.Struct({
  eq: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)),
  gte: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)),
  lte: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)),
  gt: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)),
  lt: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf))
});

// src/models/requests/iam/getBlockGroupsRequest.ts
var getBlockGroupsRequestSchema = effect.Schema.Struct({
  blockGroupId: effect.Schema.optional(effect.Schema.String),
  useAll: effect.Schema.optional(effect.Schema.Boolean),
  senderNumber: effect.Schema.optional(effect.Schema.String),
  name: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, likeLiteralSchema)),
  status: effect.Schema.optional(effect.Schema.Literal("ACTIVE", "INACTIVE")),
  startKey: effect.Schema.optional(effect.Schema.String),
  limit: effect.Schema.optional(effect.Schema.Number)
});
function finalizeGetBlockGroupsRequest(data) {
  if (!data) return {};
  let payload = {
    blockGroupId: data.blockGroupId,
    useAll: data.useAll,
    senderNumber: data.senderNumber,
    status: data.status,
    startKey: data.startKey,
    limit: data.limit
  };
  return data.name != null && (payload.name = typeof data.name == "string" ? { like: data.name } : data.name), payload;
}
var getBlockNumbersRequestSchema = effect.Schema.Struct({
  blockNumberId: effect.Schema.optional(effect.Schema.String),
  phoneNumber: effect.Schema.optional(effect.Schema.String),
  blockGroupId: effect.Schema.optional(effect.Schema.String),
  memo: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, likeLiteralSchema)),
  startKey: effect.Schema.optional(effect.Schema.String),
  limit: effect.Schema.optional(effect.Schema.Number)
});
function finalizeGetBlockNumbersRequest(data) {
  if (!data) return {};
  let payload = {
    blockNumberId: data.blockNumberId,
    phoneNumber: data.phoneNumber,
    blockGroupId: data.blockGroupId,
    startKey: data.startKey,
    limit: data.limit
  };
  return data.memo != null && (payload.memo = typeof data.memo == "string" ? { like: data.memo } : data.memo), payload;
}
var getBlacksResponseSchema = effect.Schema.Struct({
  startKey: effect.Schema.NullishOr(effect.Schema.String),
  limit: effect.Schema.Number,
  nextKey: effect.Schema.NullishOr(effect.Schema.String),
  blackList: effect.Schema.Array(blackSchema)
});
var getBlockGroupsResponseSchema = effect.Schema.Struct({
  startKey: effect.Schema.NullishOr(effect.Schema.String),
  limit: effect.Schema.Number,
  nextKey: effect.Schema.NullishOr(effect.Schema.String),
  blockGroups: effect.Schema.Array(blockGroupSchema)
});
var getBlockNumbersResponseSchema = effect.Schema.Struct({
  startKey: effect.Schema.NullishOr(effect.Schema.String),
  limit: effect.Schema.Number,
  nextKey: effect.Schema.NullishOr(effect.Schema.String),
  blockNumbers: effect.Schema.Array(blockNumberSchema)
});

// src/services/iam/iamService.ts
var IamService = class extends DefaultService {
  /**
   * 080 수신 거부 조회
   * @param data 080 수신 거부 상세 조회용 request 데이터
   * @returns GetBlacksResponse
   */
  async getBlacks(data) {
    return runSafePromise(
      this.getWithQuery({
        schema: getBlacksRequestSchema,
        finalize: finalizeGetBlacksRequest,
        url: "iam/v1/black",
        data,
        responseSchema: getBlacksResponseSchema
      })
    );
  }
  /**
   * 수신 거부 그룹 조회
   * @param data 수신 거부 그룹 조회용 request 데이터
   * @returns GetBlockGroupsResponse
   */
  async getBlockGroups(data) {
    return runSafePromise(
      this.getWithQuery({
        schema: getBlockGroupsRequestSchema,
        finalize: finalizeGetBlockGroupsRequest,
        url: "iam/v1/block/groups",
        data,
        responseSchema: getBlockGroupsResponseSchema
      })
    );
  }
  /**
   * 수신 거부 번호 조회
   * @param data 수신 거부 번호 상세 조회용 request 데이터
   * @returns GetBlockNumbersResponse
   */
  async getBlockNumbers(data) {
    return runSafePromise(
      this.getWithQuery({
        schema: getBlockNumbersRequestSchema,
        finalize: finalizeGetBlockNumbersRequest,
        url: "iam/v1/block/numbers",
        data,
        responseSchema: getBlockNumbersResponseSchema
      })
    );
  }
};
var kakaoChannelCategorySchema = effect.Schema.Struct({
  code: effect.Schema.String,
  name: effect.Schema.String
}), kakaoChannelSchema = effect.Schema.Struct({
  channelId: effect.Schema.String,
  searchId: effect.Schema.String,
  accountId: effect.Schema.String,
  phoneNumber: effect.Schema.optional(effect.Schema.String),
  sharedAccountIds: effect.Schema.Array(effect.Schema.String),
  dateCreated: effect.Schema.optional(
    effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)
  ),
  dateUpdated: effect.Schema.optional(
    effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)
  )
});
function decodeKakaoChannel(data) {
  return Effect10__namespace.gen(function* () {
    let dateCreated = yield* safeDateTransfer(data.dateCreated), dateUpdated = yield* safeDateTransfer(data.dateUpdated);
    return {
      channelId: data.channelId,
      searchId: data.searchId,
      accountId: data.accountId,
      phoneNumber: data.phoneNumber,
      sharedAccountIds: data.sharedAccountIds,
      dateCreated,
      dateUpdated
    };
  });
}
var getKakaoChannelsRequestSchema = effect.Schema.Struct({
  channelId: effect.Schema.optional(effect.Schema.String),
  searchId: effect.Schema.optional(effect.Schema.String),
  phoneNumber: effect.Schema.optional(effect.Schema.String),
  categoryCode: effect.Schema.optional(effect.Schema.String),
  startKey: effect.Schema.optional(effect.Schema.String),
  limit: effect.Schema.optional(effect.Schema.Number),
  isMine: effect.Schema.optional(effect.Schema.Boolean),
  startDate: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)),
  endDate: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf))
});
function finalizeGetKakaoChannelsRequest(data) {
  if (!data) return {};
  let payload = {
    channelId: data.channelId,
    searchId: data.searchId,
    phoneNumber: data.phoneNumber,
    categoryCode: data.categoryCode,
    startKey: data.startKey,
    limit: data.limit,
    isMine: data.isMine
  };
  return data.startDate != null && (payload.dateCreated = Object.assign(payload.dateCreated ?? {}, {
    gte: formatWithTransfer(data.startDate)
  })), data.endDate != null && (payload.dateCreated = Object.assign(payload.dateCreated ?? {}, {
    lte: formatWithTransfer(data.endDate)
  })), payload;
}
var getKakaoChannelsResponseSchema = effect.Schema.Struct({
  limit: effect.Schema.Number,
  startKey: effect.Schema.NullishOr(effect.Schema.String),
  nextKey: effect.Schema.NullishOr(effect.Schema.String),
  channelList: effect.Schema.Array(kakaoChannelSchema)
});
var kakaoChannelCategoryListSchema = effect.Schema.Array(kakaoChannelCategorySchema), KakaoChannelService = class extends DefaultService {
  async getKakaoChannelCategories() {
    return runSafePromise(
      Effect10__namespace.map(
        this.requestEffect({
          httpMethod: "GET",
          url: "kakao/v2/channels/categories",
          responseSchema: kakaoChannelCategoryListSchema
        }),
        (list) => [...list]
      )
    );
  }
  async getKakaoChannels(data) {
    let reqEffect = this.requestEffect.bind(this);
    return runSafePromise(
      Effect10__namespace.gen(function* () {
        let validated = data ? yield* decodeWithBadRequest(getKakaoChannelsRequestSchema, data) : void 0, payload = yield* safeFinalize(
          () => finalizeGetKakaoChannelsRequest(validated)
        ), parameter = stringifyQuery(payload, {
          indices: false,
          addQueryPrefix: true
        }), response = yield* reqEffect({
          httpMethod: "GET",
          url: `kakao/v2/channels${parameter}`,
          responseSchema: getKakaoChannelsResponseSchema
        });
        return {
          limit: response.limit,
          nextKey: response.nextKey,
          startKey: response.startKey,
          channelList: yield* Effect10__namespace.all(
            response.channelList.map(decodeKakaoChannel)
          )
        };
      })
    );
  }
  async getKakaoChannel(channelId) {
    return runSafePromise(
      Effect10__namespace.flatMap(
        this.requestEffect({
          httpMethod: "GET",
          url: `kakao/v2/channels/${channelId}`,
          responseSchema: kakaoChannelSchema
        }),
        decodeKakaoChannel
      )
    );
  }
  async requestKakaoChannelToken(data) {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "POST",
        url: "kakao/v2/channels/token",
        body: data
      })
    );
  }
  async createKakaoChannel(data) {
    return runSafePromise(
      this.requestEffect(
        {
          httpMethod: "POST",
          url: "kakao/v2/channels",
          body: data
        }
      )
    );
  }
  async removeKakaoChannel(channelId) {
    return runSafePromise(
      Effect10__namespace.flatMap(
        this.requestEffect({
          httpMethod: "DELETE",
          url: `kakao/v2/channels/${channelId}`
        }),
        decodeKakaoChannel
      )
    );
  }
};
var kakaoAlimtalkTemplateQuickReplyWebLinkSchema = effect.Schema.Struct({
  name: effect.Schema.String,
  linkType: effect.Schema.Literal("WL"),
  linkMo: effect.Schema.String,
  linkPc: effect.Schema.optional(effect.Schema.String)
}), kakaoAlimtalkTemplateQuickReplyAppLinkSchema = effect.Schema.Struct({
  name: effect.Schema.String,
  linkType: effect.Schema.Literal("AL"),
  linkAnd: effect.Schema.String,
  linkIos: effect.Schema.String
}), kakaoAlimtalkTemplateQuickReplyDefaultSchema = effect.Schema.Struct({
  name: effect.Schema.String,
  linkType: effect.Schema.Literal("BK", "BT", "BC")
}), kakaoAlimtalkTemplateQuickReplySchema = effect.Schema.Union(
  kakaoAlimtalkTemplateQuickReplyWebLinkSchema,
  kakaoAlimtalkTemplateQuickReplyAppLinkSchema,
  kakaoAlimtalkTemplateQuickReplyDefaultSchema
);

// src/models/base/kakao/kakaoAlimtalkTemplate.ts
var kakaoAlimtalkTemplateMessageTypeSchema = effect.Schema.Literal(
  "BA",
  "EX",
  "AD",
  "MI"
), kakaoAlimtalkTemplateEmphasizeTypeSchema = effect.Schema.Literal(
  "NONE",
  "TEXT",
  "IMAGE",
  "ITEM_LIST"
), kakaoAlimtalkTemplateAssignTypeSchema = effect.Schema.Literal(
  "CHANNEL",
  "GROUP"
), kakaoAlimtalkTemplateStatusSchema = effect.Schema.Literal(
  "PENDING",
  "INSPECTING",
  "APPROVED",
  "REJECTED"
), kakaoAlimtalkTemplateCommentTypeSchema = effect.Schema.Struct({
  isAdmin: effect.Schema.Boolean,
  memberId: effect.Schema.String,
  content: effect.Schema.NullOr(effect.Schema.String),
  dateCreated: effect.Schema.String
}), kakaoAlimtalkTemplateHighlightTypeSchema = effect.Schema.Struct({
  title: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  description: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  imageId: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String))
}), kakaoAlimtalkTemplateItemTypeSchema = effect.Schema.Struct({
  list: effect.Schema.Array(
    effect.Schema.Struct({
      title: effect.Schema.String,
      description: effect.Schema.String
    })
  ),
  summary: effect.Schema.Struct({
    title: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
    description: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String))
  })
}), kakaoAlimtalkTemplateSchema = effect.Schema.Struct({
  name: effect.Schema.String,
  channelId: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  channelGroupId: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  content: effect.Schema.optional(effect.Schema.String),
  isHidden: effect.Schema.optional(effect.Schema.Boolean),
  messageType: kakaoAlimtalkTemplateMessageTypeSchema,
  emphasizeType: kakaoAlimtalkTemplateEmphasizeTypeSchema,
  extra: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  ad: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  emphasizeTitle: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  emphasizeSubtitle: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  securityFlag: effect.Schema.Boolean,
  imageId: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  assignType: effect.Schema.optional(kakaoAlimtalkTemplateAssignTypeSchema),
  buttons: effect.Schema.optional(effect.Schema.Array(kakaoButtonSchema)),
  comments: effect.Schema.optional(
    effect.Schema.Array(kakaoAlimtalkTemplateCommentTypeSchema)
  ),
  commentable: effect.Schema.optional(effect.Schema.Boolean),
  quickReplies: effect.Schema.optional(
    effect.Schema.Array(kakaoAlimtalkTemplateQuickReplySchema)
  ),
  header: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  highlight: effect.Schema.optional(
    effect.Schema.NullOr(kakaoAlimtalkTemplateHighlightTypeSchema)
  ),
  item: effect.Schema.optional(effect.Schema.NullOr(kakaoAlimtalkTemplateItemTypeSchema)),
  templateId: effect.Schema.String,
  code: effect.Schema.optional(effect.Schema.NullOr(effect.Schema.String)),
  status: kakaoAlimtalkTemplateStatusSchema,
  variables: effect.Schema.optional(
    effect.Schema.Array(
      effect.Schema.Struct({
        name: effect.Schema.String
      })
    )
  ),
  dateCreated: effect.Schema.optional(
    effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)
  ),
  dateUpdated: effect.Schema.optional(
    effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)
  )
});
function decodeKakaoAlimtalkTemplate(data) {
  return Effect10__namespace.gen(function* () {
    let dateCreated = yield* safeDateTransfer(data.dateCreated), dateUpdated = yield* safeDateTransfer(data.dateUpdated);
    return {
      ...data,
      dateCreated,
      dateUpdated
    };
  });
}
var alimtalkTemplatesNameTypeSchema = effect.Schema.Union(
  effect.Schema.String,
  effect.Schema.Struct({ like: effect.Schema.String }),
  effect.Schema.Struct({
    eq: effect.Schema.optional(effect.Schema.String),
    ne: effect.Schema.optional(effect.Schema.String)
  })
), getKakaoAlimtalkTemplatesRequestSchema = effect.Schema.Struct({
  name: effect.Schema.optional(alimtalkTemplatesNameTypeSchema),
  channelId: effect.Schema.optional(effect.Schema.String),
  templateId: effect.Schema.optional(effect.Schema.String),
  isHidden: effect.Schema.optional(effect.Schema.Boolean),
  status: effect.Schema.optional(kakaoAlimtalkTemplateStatusSchema),
  startKey: effect.Schema.optional(effect.Schema.String),
  limit: effect.Schema.optional(effect.Schema.Number),
  startDate: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)),
  endDate: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf))
});
function finalizeGetKakaoAlimtalkTemplatesRequest(data) {
  if (!data) return {};
  let payload = {
    channelId: data.channelId,
    isHidden: data.isHidden,
    templateId: data.templateId,
    startKey: data.startKey,
    status: data.status,
    limit: data.limit
  };
  return data.name != null && (payload.name = typeof data.name == "string" ? { like: data.name } : data.name), data.startDate != null && (payload.dateCreated = Object.assign(payload.dateCreated ?? {}, {
    gte: formatWithTransfer(data.startDate)
  })), data.endDate != null && (payload.dateCreated = Object.assign(payload.dateCreated ?? {}, {
    lte: formatWithTransfer(data.endDate)
  })), payload;
}
var getKakaoTemplateResponseSchema = kakaoAlimtalkTemplateSchema.pipe(
  effect.Schema.omit("assignType", "commentable", "dateCreated", "dateUpdated"),
  effect.Schema.extend(
    effect.Schema.Struct({
      assignType: kakaoAlimtalkTemplateAssignTypeSchema,
      accountId: effect.Schema.NullishOr(effect.Schema.String),
      commentable: effect.Schema.Boolean,
      dateCreated: effect.Schema.String,
      dateUpdated: effect.Schema.String
    })
  )
);

// src/models/responses/kakao/getKakaoAlimtalkTemplatesResponse.ts
var getKakaoAlimtalkTemplatesResponseSchema = effect.Schema.Struct({
  limit: effect.Schema.Number,
  templateList: effect.Schema.Array(getKakaoTemplateResponseSchema),
  startKey: effect.Schema.NullishOr(effect.Schema.String),
  nextKey: effect.Schema.NullishOr(effect.Schema.String)
});
var kakaoAlimtalkTemplateCategoryListSchema = effect.Schema.Array(
  kakaoChannelCategorySchema
), KakaoTemplateService = class extends DefaultService {
  /**
   * 카카오 템플릿 카테고리 조회
   */
  async getKakaoAlimtalkTemplateCategories() {
    return runSafePromise(
      Effect10__namespace.map(
        this.requestEffect({
          httpMethod: "GET",
          url: "kakao/v2/templates/categories",
          responseSchema: kakaoAlimtalkTemplateCategoryListSchema
        }),
        (list) => [...list]
      )
    );
  }
  /**
   * @description 카카오 알림톡 템플릿 생성
   */
  async createKakaoAlimtalkTemplate(data) {
    return runSafePromise(
      Effect10__namespace.flatMap(
        this.requestEffect({
          httpMethod: "POST",
          url: "kakao/v2/templates",
          body: data
        }),
        decodeKakaoAlimtalkTemplate
      )
    );
  }
  /**
   * 카카오 템플릿 목록 조회
   */
  async getKakaoAlimtalkTemplates(data) {
    let reqEffect = this.requestEffect.bind(this);
    return runSafePromise(
      Effect10__namespace.gen(function* () {
        let validated = data ? yield* decodeWithBadRequest(
          getKakaoAlimtalkTemplatesRequestSchema,
          data
        ) : void 0, payload = yield* safeFinalize(
          () => finalizeGetKakaoAlimtalkTemplatesRequest(validated)
        ), parameter = stringifyQuery(payload, {
          indices: false,
          addQueryPrefix: true
        }), response = yield* reqEffect({
          httpMethod: "GET",
          url: `kakao/v2/templates${parameter}`,
          responseSchema: getKakaoAlimtalkTemplatesResponseSchema
        }), templateList = yield* Effect10__namespace.all(
          response.templateList.map(
            (item) => Effect10__namespace.flatMap(
              decodeWithBadRequest(kakaoAlimtalkTemplateSchema, item),
              decodeKakaoAlimtalkTemplate
            )
          )
        );
        return {
          limit: response.limit,
          nextKey: response.nextKey,
          startKey: response.startKey,
          templateList
        };
      })
    );
  }
  /**
   * 카카오 템플릿 상세 조회
   */
  async getKakaoAlimtalkTemplate(templateId) {
    return runSafePromise(
      Effect10__namespace.flatMap(
        this.requestEffect({
          httpMethod: "GET",
          url: `kakao/v2/templates/${templateId}`,
          responseSchema: getKakaoTemplateResponseSchema
        }),
        decodeKakaoAlimtalkTemplate
      )
    );
  }
  /**
   * 카카오 알림톡 템플릿 검수 취소 요청
   */
  async cancelInspectionKakaoAlimtalkTemplate(templateId) {
    return runSafePromise(
      Effect10__namespace.flatMap(
        this.requestEffect({
          httpMethod: "PUT",
          url: `kakao/v2/templates/${templateId}/inspection/cancel`
        }),
        decodeKakaoAlimtalkTemplate
      )
    );
  }
  /**
   * 카카오 알림톡 템플릿 수정(검수 X)
   */
  async updateKakaoAlimtalkTemplate(templateId, data) {
    return runSafePromise(
      Effect10__namespace.flatMap(
        this.requestEffect({
          httpMethod: "PUT",
          url: `kakao/v2/templates/${templateId}`,
          body: data
        }),
        decodeKakaoAlimtalkTemplate
      )
    );
  }
  /**
   * 카카오 알림톡 템플릿 이름 수정(검수 상태 상관없이 변경가능)
   */
  async updateKakaoAlimtalkTemplateName(templateId, name) {
    return runSafePromise(
      Effect10__namespace.flatMap(
        this.requestEffect({
          httpMethod: "PUT",
          url: `kakao/v2/templates/${templateId}/name`,
          body: { name }
        }),
        decodeKakaoAlimtalkTemplate
      )
    );
  }
  /**
   * 카카오 알림톡 템플릿 삭제(대기, 반려 상태일 때만 삭제가능)
   */
  async removeKakaoAlimtalkTemplate(templateId) {
    return runSafePromise(
      Effect10__namespace.flatMap(
        this.requestEffect({
          httpMethod: "DELETE",
          url: `kakao/v2/templates/${templateId}`
        }),
        decodeKakaoAlimtalkTemplate
      )
    );
  }
};
var getGroupsRequestSchema = effect.Schema.Struct({
  groupId: effect.Schema.optional(effect.Schema.String),
  startKey: effect.Schema.optional(effect.Schema.String),
  limit: effect.Schema.optional(effect.Schema.Number),
  startDate: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)),
  endDate: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf))
});
function finalizeGetGroupsRequest(data) {
  if (!data) return {};
  let payload = {
    startKey: data.startKey,
    limit: data.limit
  };
  return data.groupId && (payload.criteria = "groupId", payload.cond = "eq", payload.value = data.groupId), data.startDate != null && (payload.startDate = formatWithTransfer(data.startDate)), data.endDate != null && (payload.endDate = formatWithTransfer(data.endDate)), payload;
}

// package.json
var package_default = {
  version: "6.0.1"};

// src/models/requests/messages/requestConfig.ts
var osPlatform = `${process.platform} | ${process.version}`, sdkVersion = `nodejs/${package_default.version}`, defaultAgentTypeSchema = effect.Schema.Struct({
  sdkVersion: effect.Schema.optional(effect.Schema.String).pipe(
    effect.Schema.withDecodingDefault(() => sdkVersion),
    effect.Schema.withConstructorDefault(() => sdkVersion)
  ),
  osPlatform: effect.Schema.optional(effect.Schema.String).pipe(
    effect.Schema.withDecodingDefault(() => osPlatform),
    effect.Schema.withConstructorDefault(() => osPlatform)
  ),
  appId: effect.Schema.optional(effect.Schema.String)
}), sendRequestConfigSchema = effect.Schema.Struct({
  scheduledDate: effect.Schema.optional(
    effect.Schema.Union(effect.Schema.DateFromSelf, effect.Schema.DateFromString).pipe(
      effect.Schema.transformOrFail(effect.Schema.String, {
        decode: (fromA, _, ast) => safeFormatWithTransfer(fromA).pipe(
          effect.Effect.mapError(
            (err) => new effect.ParseResult.Type(ast, fromA, err.message)
          )
        ),
        encode: (toI) => effect.ParseResult.succeed(new Date(toI))
      })
    )
  ),
  allowDuplicates: effect.Schema.optional(effect.Schema.Boolean),
  appId: effect.Schema.optional(effect.Schema.String),
  showMessageList: effect.Schema.optional(effect.Schema.Boolean)
}), defaultMessageRequestSchema = effect.Schema.Struct({
  allowDuplicates: effect.Schema.optional(effect.Schema.Boolean),
  agent: effect.Schema.optional(defaultAgentTypeSchema)
});
var removeHyphens = (s) => s.replace(/-/g, ""), phoneNumberSchema = effect.Schema.String.pipe(
  effect.Schema.transform(effect.Schema.String, {
    decode: removeHyphens,
    encode: (s) => s
  }),
  effect.Schema.filter((s) => s.trim().length > 0, {
    message: () => "\uC804\uD654\uBC88\uD638\uB294 \uBE48 \uBB38\uC790\uC5F4\uC77C \uC218 \uC5C6\uC2B5\uB2C8\uB2E4."
  }),
  effect.Schema.filter((s) => /^[0-9]+$/.test(s), {
    message: () => "\uC804\uD654\uBC88\uD638\uB294 \uC22B\uC790 \uBC0F \uD2B9\uC218\uBB38\uC790 - \uC678 \uBB38\uC790\uB97C \uD3EC\uD568\uD560 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4."
  })
), nonEmptyArrayFilter = (schema) => effect.Schema.Array(schema).pipe(
  effect.Schema.filter((arr) => arr.length > 0, {
    message: () => "\uB370\uC774\uD130\uAC00 \uBC18\uB4DC\uC2DC 1\uAC74 \uC774\uC0C1 \uAE30\uC785\uB418\uC5B4 \uC788\uC5B4\uC57C \uD569\uB2C8\uB2E4."
  })
), requestSendOneMessageSchema = messageSchema.pipe(
  effect.Schema.omit("to", "from"),
  effect.Schema.extend(
    effect.Schema.Struct({
      to: effect.Schema.Union(phoneNumberSchema, effect.Schema.Array(phoneNumberSchema)),
      from: effect.Schema.optional(phoneNumberSchema)
    })
  )
), requestSendMessageSchema = effect.Schema.Union(
  requestSendOneMessageSchema,
  nonEmptyArrayFilter(requestSendOneMessageSchema)
), defaultAgentValue = effect.Schema.decodeSync(defaultAgentTypeSchema)({}), agentWithDefaultSchema = effect.Schema.optional(defaultAgentTypeSchema).pipe(
  effect.Schema.withDecodingDefault(() => defaultAgentValue),
  effect.Schema.withConstructorDefault(() => defaultAgentValue)
), multipleMessageSendingRequestSchema = effect.Schema.Struct({
  allowDuplicates: effect.Schema.optional(effect.Schema.Boolean),
  agent: agentWithDefaultSchema,
  messages: nonEmptyArrayFilter(requestSendOneMessageSchema),
  scheduledDate: effect.Schema.optional(
    effect.Schema.Union(effect.Schema.DateFromSelf, effect.Schema.DateFromString)
  ),
  showMessageList: effect.Schema.optional(effect.Schema.Boolean)
});
var GroupService = class extends DefaultService {
  /**
   * 그룹 생성
   * @param allowDuplicates 생성할 그룹이 중복 수신번호를 허용하는지 여부를 확인합니다.
   * @param appId 생성할 그룹에 특정 appId를 넣을 수 있습니다.
   * @param customFields 생성할 그룹에 사용자 정의 데이터를 Record 형태로 삽입할 수 있습니다.
   */
  async createGroup(allowDuplicates = false, appId, customFields) {
    return runSafePromise(
      Effect10__namespace.map(
        this.requestEffect({
          httpMethod: "POST",
          url: "messages/v4/groups",
          body: {
            sdkVersion,
            osPlatform,
            allowDuplicates,
            appId,
            customFields
          }
        }),
        (response) => response.groupId
      )
    );
  }
  /**
   * 그룹 메시지 추가
   * 한번 요청으로 최대 10,000건의 메시지를 추가할 수 있습니다.
   * 추가 가능한 최대 메시지 건 수는 1,000,000건 입니다.
   * @param groupId 생성 된 Group ID
   * @param messages 여러 메시지(문자, 알림톡 등)
   */
  async addMessagesToGroup(groupId, messages) {
    let reqEffect = this.requestEffect.bind(this);
    return runSafePromise(
      Effect10__namespace.flatMap(
        decodeWithBadRequest(requestSendMessageSchema, messages),
        (validatedMessages) => reqEffect({
          httpMethod: "PUT",
          url: `messages/v4/groups/${groupId}/messages`,
          body: {
            messages: Array.isArray(validatedMessages) ? validatedMessages : [validatedMessages]
          }
        })
      )
    );
  }
  /**
   * 그룹 메시지 전송 요청
   * @param groupId 생성 된 Group ID
   */
  async sendGroup(groupId) {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "POST",
        url: `messages/v4/groups/${groupId}/send`
      })
    );
  }
  /**
   * 그룹 예약 발송 설정
   * @param groupId 생성 된 Group ID
   * @param scheduledDate 예약발송 할 날짜
   */
  async reserveGroup(groupId, scheduledDate) {
    let reqEffect = this.requestEffect.bind(this);
    return runSafePromise(
      Effect10__namespace.flatMap(
        safeFormatWithTransfer(scheduledDate),
        (formattedScheduledDate) => reqEffect({
          httpMethod: "POST",
          url: `messages/v4/groups/${groupId}/schedule`,
          body: { scheduledDate: formattedScheduledDate }
        })
      )
    );
  }
  /**
   * 그룹 내 예약 발송 취소(메시지 실패 전체 처리 됨)
   * @param groupId 생성 된 Group ID
   */
  async removeReservationToGroup(groupId) {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "DELETE",
        url: `messages/v4/groups/${groupId}/schedule`
      })
    );
  }
  /**
   * 그룹 목록 정보 조회
   * @param data 그룹 정보 상세 조회용 request 데이터
   */
  async getGroups(data) {
    return runSafePromise(
      this.getWithQuery({
        schema: getGroupsRequestSchema,
        finalize: finalizeGetGroupsRequest,
        url: "messages/v4/groups",
        data,
        responseSchema: getGroupsResponseSchema
      })
    );
  }
  /**
   * 단일 그룹정보 조회
   * @param groupId 그룹 ID
   */
  async getGroup(groupId) {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "GET",
        url: `messages/v4/groups/${groupId}`,
        responseSchema: groupMessageResponseSchema
      })
    );
  }
  /**
   * 그룹 내 메시지 목록 조회
   * @param groupId 생성 된 Group ID
   * @param data startkey, limit 등 쿼리 조건 파라미터
   */
  async getGroupMessages(groupId, data) {
    let parameter = stringifyQuery(data, {
      indices: false,
      addQueryPrefix: true
    });
    return runSafePromise(
      this.requestEffect({
        httpMethod: "GET",
        url: `messages/v4/groups/${groupId}/messages${parameter}`,
        responseSchema: getMessagesResponseSchema
      })
    );
  }
  /**
   * 그룹 내 특정 메시지 삭제
   * @param groupId 생성 된 Group Id
   * @param messageIds 생성 된 메시지 ID 목록
   */
  async removeGroupMessages(groupId, messageIds) {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "DELETE",
        url: `messages/v4/groups/${groupId}/messages`,
        body: { messageIds }
      })
    );
  }
  /**
   * 그룹 삭제
   * @param groupId
   */
  async removeGroup(groupId) {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "DELETE",
        url: `messages/v4/groups/${groupId}`
      })
    );
  }
};
var dateTypeSchema = effect.Schema.Literal("CREATED", "UPDATED"), baseGetMessagesRequestSchema = effect.Schema.Struct({
  startKey: effect.Schema.optional(effect.Schema.String),
  limit: effect.Schema.optional(effect.Schema.Number),
  messageId: effect.Schema.optional(effect.Schema.String),
  messageIds: effect.Schema.optional(effect.Schema.Array(effect.Schema.String)),
  groupId: effect.Schema.optional(effect.Schema.String),
  to: effect.Schema.optional(effect.Schema.String),
  from: effect.Schema.optional(effect.Schema.String),
  type: effect.Schema.optional(messageTypeSchema),
  statusCode: effect.Schema.optional(effect.Schema.String),
  dateType: effect.Schema.optional(dateTypeSchema),
  startDate: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)),
  endDate: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf))
}), getMessagesRequestSchema = baseGetMessagesRequestSchema.pipe(
  effect.Schema.filter((data) => {
    let hasDate = data.startDate != null || data.endDate != null;
    return data.dateType != null && !hasDate ? "dateType\uC740 startDate \uB610\uB294 endDate\uC640 \uD568\uAED8 \uC0AC\uC6A9\uD574\uC57C \uD569\uB2C8\uB2E4." : true;
  })
);
function finalizeGetMessagesRequest(data) {
  if (!data) return {};
  let payload = {
    startKey: data.startKey,
    limit: data.limit,
    dateType: data.startDate != null || data.endDate != null ? data.dateType ?? "CREATED" : data.dateType,
    messageId: data.messageId,
    messageIds: data.messageIds,
    groupId: data.groupId,
    to: data.to,
    from: data.from,
    type: data.type,
    statusCode: data.statusCode
  };
  return data.startDate != null && (payload.startDate = formatWithTransfer(data.startDate)), data.endDate != null && (payload.endDate = formatWithTransfer(data.endDate)), payload;
}
var getStatisticsRequestSchema = effect.Schema.Struct({
  masterAccountId: effect.Schema.optional(effect.Schema.String),
  startDate: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf)),
  endDate: effect.Schema.optional(effect.Schema.Union(effect.Schema.String, effect.Schema.DateFromSelf))
});
function finalizeGetStatisticsRequest(data) {
  if (!data) return {};
  let payload = {
    masterAccountId: data.masterAccountId
  };
  return data.startDate != null && (payload.startDate = formatWithTransfer(data.startDate)), data.endDate != null && (payload.endDate = formatWithTransfer(data.endDate)), payload;
}
var MessageService = class extends DefaultService {
  /**
   * 메시지 발송 기능, sendMany 함수보다 개선된 오류 표시 기능등을 제공합니다.
   * 한번의 요청으로 최대 10,000건까지 발송할 수 있습니다.
   * @param messages 발송 요청할 메시지 파라미터(문자, 알림톡 등)
   * @param requestConfigParameter request시 필요한 파라미터 오브젝트
   * @throws MessageNotReceivedError 모든 메시지 접수건이 실패건으로 진행되는 경우 반환되는 에러
   * @throws BadRequestError 잘못된 파라미터를 기입했거나, 데이터가 아예 없는 경우 반환되는 에러
   */
  async send(messages, requestConfigParameter) {
    let reqEffect = this.requestEffect.bind(this);
    return runSafePromise(
      Effect10__namespace.gen(function* () {
        let messageSchema2 = yield* decodeWithBadRequest(
          requestSendMessageSchema,
          messages
        ), messageParameters = Array.isArray(messageSchema2) ? messageSchema2 : [messageSchema2];
        if (messageParameters.length === 0)
          return yield* new BadRequestError({
            message: "\uB370\uC774\uD130\uAC00 \uBC18\uB4DC\uC2DC 1\uAC74 \uC774\uC0C1 \uAE30\uC785\uB418\uC5B4 \uC788\uC5B4\uC57C \uD569\uB2C8\uB2E4."
          });
        let decodedConfig = yield* decodeWithBadRequest(
          sendRequestConfigSchema,
          requestConfigParameter ?? {}
        ), parameterObject = {
          messages: messageParameters,
          allowDuplicates: decodedConfig.allowDuplicates,
          ...decodedConfig.appId ? { agent: { appId: decodedConfig.appId } } : {},
          scheduledDate: decodedConfig.scheduledDate,
          showMessageList: decodedConfig.showMessageList
        }, parameter = yield* decodeWithBadRequest(
          multipleMessageSendingRequestSchema,
          parameterObject
        ), response = yield* reqEffect({
          httpMethod: "POST",
          url: "messages/v4/send-many/detail",
          body: parameter
        }), { count } = response.groupInfo;
        return response.failedMessageList.length > 0 && count.total === count.registeredFailed ? yield* new MessageNotReceivedError({
          failedMessageList: response.failedMessageList,
          totalCount: response.failedMessageList.length
        }) : response;
      })
    );
  }
  /**
   * 메시지 목록 조회
   * @param data 목록 조회 상세조건 파라미터
   */
  async getMessages(data) {
    return runSafePromise(
      this.getWithQuery({
        schema: getMessagesRequestSchema,
        finalize: finalizeGetMessagesRequest,
        url: "messages/v4/list",
        data,
        responseSchema: getMessagesResponseSchema
      })
    );
  }
  /**
   * 통계 조회
   * @param data 통계 상세 조건 파라미터
   * @returns GetStatisticsResponse 통계 결과
   */
  async getStatistics(data) {
    return runSafePromise(
      this.getWithQuery({
        schema: getStatisticsRequestSchema,
        finalize: finalizeGetStatisticsRequest,
        url: "messages/v4/statistics",
        data,
        responseSchema: getStatisticsResponseSchema
      })
    );
  }
};
var isHttpUrl = (value) => {
  try {
    let url$1 = new url.URL(value);
    return url$1.protocol === "http:" || url$1.protocol === "https:";
  } catch {
    return false;
  }
}, fromUrl = (url) => Effect10__namespace.flatMap(
  Effect10__namespace.tryPromise({
    try: () => fetch(url),
    catch: (error) => new DefaultError({
      errorCode: "FileUrlFetchError",
      errorMessage: `\uB124\uD2B8\uC6CC\uD06C \uC624\uB958\uB85C URL(${url})\uC744(\uB97C) \uAC00\uC838\uC624\uC9C0 \uBABB\uD588\uC2B5\uB2C8\uB2E4.`,
      context: { url, cause: String(error) }
    })
  }),
  (response) => response.ok ? Effect10__namespace.tryPromise({
    try: () => response.arrayBuffer(),
    catch: (error) => new DefaultError({
      errorCode: "FileReadError",
      errorMessage: "\uC751\uB2F5 body \uCC98\uB9AC \uC911 \uC624\uB958\uAC00 \uBC1C\uC0DD\uD588\uC2B5\uB2C8\uB2E4.",
      context: { url, cause: String(error) }
    })
  }) : Effect10__namespace.fail(
    new DefaultError({
      errorCode: "FileUrlFetchError",
      errorMessage: `URL(${url}) \uC694\uCCAD \uC2E4\uD328 \u2013 \uC0C1\uD0DC \uCF54\uB4DC: ${response.status} ${response.statusText}`,
      context: { url, status: response.status }
    })
  )
).pipe(
  Effect10__namespace.map((arrayBuffer) => Buffer.from(arrayBuffer).toString("base64"))
), fromPath = (path) => Effect10__namespace.tryPromise({
  try: () => fs.promises.readFile(path),
  catch: (error) => new DefaultError({
    errorCode: "FileReadError",
    errorMessage: `\uD30C\uC77C\uC744 \uC77D\uC744 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4: ${path}`,
    context: { path, cause: String(error) }
  })
}).pipe(Effect10__namespace.map((buffer) => buffer.toString("base64")));
function fileToBase64Effect(path) {
  return isHttpUrl(path) ? fromUrl(path) : fromPath(path);
}
var StorageService = class extends DefaultService {
  /**
   * 파일(이미지) 업로드
   * 카카오 친구톡 이미지는 500kb, MMS는 200kb, 발신번호 서류 인증용 파일은 2mb의 제한이 있음
   * @param filePath 해당 파일의 경로 또는 접근 가능한 이미지 URL
   * @param fileType 저장할 파일의 유형, 예) 카카오 친구톡 용 이미지 -> KAKAO, MMS용 사진 -> MMS, 발신번호 서류 인증에 쓰이는 문서 등 -> DOCUMENT, RCS 이미지 -> RCS
   * @param name 파일 이름
   * @param link 파일 링크, 친구톡의 경우 필수 값
   */
  async uploadFile(filePath, fileType, name, link) {
    let reqEffect = this.requestEffect.bind(this);
    return runSafePromise(
      Effect10__namespace.flatMap(
        fileToBase64Effect(filePath),
        (encodedFile) => reqEffect({
          httpMethod: "POST",
          url: "storage/v1/files",
          body: { file: encodedFile, type: fileType, name, link }
        })
      )
    );
  }
};
var baseKakaoAlimtalkTemplateRequestSchema = effect.Schema.Struct({
  name: effect.Schema.String,
  content: effect.Schema.String,
  categoryCode: effect.Schema.String,
  buttons: effect.Schema.optional(effect.Schema.Array(kakaoButtonSchema)),
  quickReplies: effect.Schema.optional(
    effect.Schema.Array(kakaoAlimtalkTemplateQuickReplySchema)
  ),
  messageType: effect.Schema.optional(kakaoAlimtalkTemplateMessageTypeSchema),
  emphasizeType: effect.Schema.optional(kakaoAlimtalkTemplateEmphasizeTypeSchema),
  header: effect.Schema.optional(effect.Schema.String),
  highlight: effect.Schema.optional(kakaoAlimtalkTemplateHighlightTypeSchema),
  item: effect.Schema.optional(kakaoAlimtalkTemplateItemTypeSchema),
  extra: effect.Schema.optional(effect.Schema.String),
  emphasizeTitle: effect.Schema.optional(effect.Schema.String),
  emphasizeSubTitle: effect.Schema.optional(effect.Schema.String),
  securityFlag: effect.Schema.optional(effect.Schema.Boolean),
  imageId: effect.Schema.optional(effect.Schema.String)
}), createKakaoChannelAlimtalkTemplateRequestSchema = effect.Schema.extend(
  baseKakaoAlimtalkTemplateRequestSchema,
  effect.Schema.Struct({ channelId: effect.Schema.String })
), createKakaoChannelGroupAlimtalkTemplateRequestSchema = effect.Schema.extend(
  baseKakaoAlimtalkTemplateRequestSchema,
  effect.Schema.Struct({ channelGroupId: effect.Schema.String })
), createKakaoAlimtalkTemplateRequestSchema = effect.Schema.Union(
  createKakaoChannelAlimtalkTemplateRequestSchema,
  createKakaoChannelGroupAlimtalkTemplateRequestSchema
);
var createKakaoChannelTokenRequestSchema = effect.Schema.Struct({
  searchId: effect.Schema.String,
  phoneNumber: effect.Schema.String
}), createKakaoChannelRequestSchema = effect.Schema.Struct({
  searchId: effect.Schema.String,
  phoneNumber: effect.Schema.String,
  categoryCode: effect.Schema.String,
  token: effect.Schema.String
});
var updateKakaoAlimtalkTemplateRequestSchema = effect.Schema.Struct({
  name: effect.Schema.optional(effect.Schema.String),
  content: effect.Schema.optional(effect.Schema.String),
  categoryCode: effect.Schema.optional(effect.Schema.String),
  buttons: effect.Schema.optional(effect.Schema.Array(kakaoButtonSchema)),
  quickReplies: effect.Schema.optional(
    effect.Schema.Array(kakaoAlimtalkTemplateQuickReplySchema)
  ),
  messageType: effect.Schema.optional(kakaoAlimtalkTemplateMessageTypeSchema),
  emphasizeType: effect.Schema.optional(kakaoAlimtalkTemplateEmphasizeTypeSchema),
  header: effect.Schema.optional(effect.Schema.String),
  highlight: effect.Schema.optional(kakaoAlimtalkTemplateHighlightTypeSchema),
  item: effect.Schema.optional(kakaoAlimtalkTemplateItemTypeSchema),
  extra: effect.Schema.optional(effect.Schema.String),
  emphasizeTitle: effect.Schema.optional(effect.Schema.String),
  emphasizeSubTitle: effect.Schema.optional(effect.Schema.String),
  securityFlag: effect.Schema.optional(effect.Schema.Boolean),
  imageId: effect.Schema.optional(effect.Schema.String)
});
var groupMessageAddRequestSchema = effect.Schema.Struct({
  messages: effect.Schema.Array(messageSchema)
}), scheduledDateSendingRequestSchema = effect.Schema.Struct({
  scheduledDate: effect.Schema.String
}), removeMessageIdsToGroupRequestSchema = effect.Schema.Struct({
  messageIds: effect.Schema.Array(effect.Schema.String)
}), getGroupMessagesRequestSchema = effect.Schema.Struct({
  startKey: effect.Schema.optional(effect.Schema.String),
  limit: effect.Schema.optional(effect.Schema.Number)
}), fileIdsSchema = effect.Schema.Struct({
  fileIds: effect.Schema.Array(effect.Schema.String)
}), fileTypeSchema = effect.Schema.Literal(
  "KAKAO",
  "MMS",
  "DOCUMENT",
  "RCS",
  "FAX",
  "BMS",
  "BMS_WIDE",
  "BMS_WIDE_MAIN_ITEM_LIST",
  "BMS_WIDE_SUB_ITEM_LIST",
  "BMS_CAROUSEL_FEED_LIST",
  "BMS_CAROUSEL_COMMERCE_LIST"
), fileUploadRequestSchema = effect.Schema.Struct({
  file: effect.Schema.String,
  type: fileTypeSchema,
  name: effect.Schema.optional(effect.Schema.String),
  link: effect.Schema.optional(effect.Schema.String)
}), createGroupRequestSchema = effect.Schema.extend(
  defaultAgentTypeSchema,
  effect.Schema.Struct({
    allowDuplicates: effect.Schema.Boolean,
    customFields: effect.Schema.optional(
      effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.String })
    )
  })
);
var failedMessageSchema = effect.Schema.Struct({
  to: effect.Schema.String,
  from: effect.Schema.String,
  type: effect.Schema.String,
  statusMessage: effect.Schema.String,
  country: effect.Schema.String,
  messageId: effect.Schema.String,
  statusCode: effect.Schema.String,
  accountId: effect.Schema.String,
  customFields: effect.Schema.optional(
    effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.String })
  )
}), messageResponseItemSchema = effect.Schema.Struct({
  messageId: effect.Schema.String,
  statusCode: effect.Schema.String,
  customFields: effect.Schema.optional(
    effect.Schema.Record({ key: effect.Schema.String, value: effect.Schema.String })
  ),
  statusMessage: effect.Schema.String
}), detailGroupMessageResponseSchema = effect.Schema.Struct({
  /**
   * 메시지 발송 접수에 실패한 메시지 요청 목록들
   */
  failedMessageList: effect.Schema.Array(failedMessageSchema),
  /**
   * 발송 정보(성공, 실패 등) 응답 데이터
   */
  groupInfo: groupMessageResponseSchema,
  /**
   * Send 메소드 호출 당시 showMessageList 값이 true로 되어있을 때 표시되는 메시지 목록
   */
  messageList: effect.Schema.optional(effect.Schema.Array(messageResponseItemSchema))
});

// src/index.ts
var SolapiMessageService = class {
  /**
   * 잔액조회
   * @returns GetBalanceResponse
   */
  getBalance;
  /**
   * 080 수신 거부 조회
   * @param data 080 수신 거부 상세 조회용 request 데이터
   * @returns GetBlacksResponse
   */
  getBlacks;
  /**
   * 수신 거부 그룹 조회
   * @param data 수신 거부 그룹 조회용 request 데이터
   * @returns GetBlockGroupsResponse
   */
  getBlockGroups;
  /**
   * 수신 거부 번호 조회
   * @param data 수신 거부 번호 상세 조회용 request 데이터
   * @returns GetBlockNumbersResponse
   */
  getBlockNumbers;
  /**
   * 카카오 채널 카테고리 조회
   */
  getKakaoChannelCategories;
  /**
   * 카카오 채널 목록 조회
   * @param data 카카오 채널 목록을 더 자세하게 조회할 때 필요한 파라미터
   */
  getKakaoChannels;
  /**
   * @description 카카오 채널 조회
   * @param channelId 카카오 채널 ID(구 pfId)
   */
  getKakaoChannel;
  /**
   * @description 카카오 채널 연동을 위한 인증 토큰 요청
   */
  requestKakaoChannelToken;
  /**
   * @description 카카오 채널 연동 메소드
   * getKakaoChannelCategories, requestKakaoChannelToken 메소드를 선행적으로 호출해야 합니다!
   */
  createKakaoChannel;
  /**
   * @description 카카오 채널 삭제, 채널이 삭제 될 경우 해당 채널의 템플릿이 모두 삭제됩니다!
   * @param channelId 카카오 채널 ID
   */
  removeKakaoChannel;
  /**
   * 카카오 템플릿 카테고리 조회
   */
  getKakaoAlimtalkTemplateCategories;
  /**
   * @description 카카오 알림톡 템플릿 생성
   * 반드시 getKakaoAlimtalkTemplateCategories를 먼저 호출하여 카테고리 값을 확인해야 합니다!
   * @param data 알림톡 템플릿 생성을 위한 파라미터
   */
  createKakaoAlimtalkTemplate;
  /**
   * 카카오 템플릿 목록 조회
   * @param data 카카오 템플릿 목록을 더 자세하게 조회할 때 필요한 파라미터
   */
  getKakaoAlimtalkTemplates;
  /**
   * 카카오 템플릿 상세 조회
   * @param templateId 카카오 알림톡 템플릿 ID
   */
  getKakaoAlimtalkTemplate;
  /**
   * 카카오 알림톡 템플릿 검수 취소 요청
   * @param templateId 카카오 알림톡 템플릿 ID
   */
  cancelInspectionKakaoAlimtalkTemplate;
  /**
   * 카카오 알림톡 템플릿 수정(검수 X)
   * @param templateId 카카오 알림톡 템플릿 ID
   * @param data 카카오 알림톡 템플릿 수정을 위한 파라미터
   */
  updateKakaoAlimtalkTemplate;
  /**
   * 카카오 알림톡 템플릿 이름 수정(검수 상태 상관없이 변경가능)
   * @param templateId 카카오 알림톡 템플릿 ID
   * @param name 카카오 알림톡 템플릿 이름 변경을 위한 파라미터
   */
  updateKakaoAlimtalkTemplateName;
  /**
   * 카카오 알림톡 템플릿 삭제(대기, 반려 상태일 때만 삭제가능)
   * @param templateId 카카오 알림톡 템플릿 ID
   */
  removeKakaoAlimtalkTemplate;
  /**
   * 그룹 생성
   * @param allowDuplicates 생성할 그룹이 중복 수신번호를 허용하는지 여부를 확인합니다.
   * @param appId 생성할 그룹에 특정 appId를 넣을 수 있습니다.
   * @param customFields 생성할 그룹에 사용자 정의 데이터를 Record 형태로 삽입할 수 있습니다.
   */
  createGroup;
  /**
   * 그룹 메시지 추가
   * 한번 요청으로 최대 10,000건의 메시지를 추가할 수 있습니다.
   * 추가 가능한 최대 메시지 건 수는 1,000,000건 입니다.
   * @param groupId 생성 된 Group ID
   * @param messages 여러 메시지(문자, 알림톡 등)
   */
  addMessagesToGroup;
  /**
   * 그룹 메시지 전송 요청
   * @param groupId 생성 된 Group ID
   */
  sendGroup;
  /**
   * 그룹 예약 발송 설정
   * @param groupId 생성 된 Group ID
   * @param scheduledDate 예약발송 할 날짜
   */
  reserveGroup;
  /**
   * 그룹 내 예약 발송 취소(메시지 실패 전체 처리 됨)
   * @param groupId 생성 된 Group ID
   */
  removeReservationToGroup;
  /**
   * 그룹 목록 정보 조회
   * @param data 그룹 정보 상세 조회용 request 데이터
   */
  getGroups;
  /**
   * 단일 그룹정보 조회
   * @param groupId 그룹 ID
   */
  getGroup;
  /**
   * 그룹 내 메시지 목록 조회
   * @param groupId 생성 된 Group ID
   * @param data startkey, limit 등 쿼리 조건 파라미터
   */
  getGroupMessages;
  /**
   * 그룹 내 특정 메시지 삭제
   * @param groupId 생성 된 Group Id
   * @param messageIds 생성 된 메시지 ID 목록
   */
  removeGroupMessages;
  /**
   * 그룹 삭제
   * @param groupId
   */
  removeGroup;
  /**
   * 메시지 발송 기능, sendMany 함수보다 개선된 오류 표시 기능등을 제공합니다.
   * 한번의 요청으로 최대 10,000건까지 발송할 수 있습니다.
   * @param messages 발송 요청할 메시지 파라미터(문자, 알림톡 등)
   * @param requestConfigParameter request시 필요한 파라미터 오브젝트
   * @throws MessageNotReceivedError 모든 메시지 접수건이 실패건으로 진행되는 경우 반환되는 에러
   * @throws BadRequestError 잘못된 파라미터를 기입했거나, 데이터가 아예 없는 경우 반환되는 에러
   */
  send;
  /**
   * 메시지 목록 조회
   * @param data 목록 조회 상세조건 파라미터
   */
  getMessages;
  /**
   * 통계 조회
   * @param data 통계 상세 조건 파라미터
   * @returns GetStatisticsResponse 통계 결과
   */
  getStatistics;
  /**
   * 파일(이미지) 업로드
   * 카카오 친구톡 이미지는 500kb, MMS는 200kb, 발신번호 서류 인증용 파일은 2mb의 제한이 있음
   * @param filePath 해당 파일의 경로 또는 접근 가능한 이미지 URL
   * @param fileType 저장할 파일의 유형, 예) 카카오 친구톡 용 이미지 -> KAKAO, MMS용 사진 -> MMS, 발신번호 서류 인증에 쓰이는 문서 등 -> DOCUMENT, RCS 이미지 -> RCS
   * @param name 파일 이름
   * @param link 파일 링크, 친구톡의 경우 필수 값
   */
  uploadFile;
  constructor(apiKey, apiSecret) {
    if (!apiKey || !apiSecret)
      throw new ApiKeyError({
        message: "API Key\uC640 API Secret\uC740 \uD544\uC218\uC785\uB2C8\uB2E4."
      });
    let cashService = new CashService(apiKey, apiSecret), iamService = new IamService(apiKey, apiSecret), kakaoChannelService = new KakaoChannelService(apiKey, apiSecret), kakaoTemplateService = new KakaoTemplateService(apiKey, apiSecret), groupService = new GroupService(apiKey, apiSecret), messageService = new MessageService(apiKey, apiSecret), storageService = new StorageService(apiKey, apiSecret);
    this.getBalance = cashService.getBalance.bind(cashService), this.getBlacks = iamService.getBlacks.bind(iamService), this.getBlockGroups = iamService.getBlockGroups.bind(iamService), this.getBlockNumbers = iamService.getBlockNumbers.bind(iamService), this.getKakaoChannelCategories = kakaoChannelService.getKakaoChannelCategories.bind(kakaoChannelService), this.getKakaoChannels = kakaoChannelService.getKakaoChannels.bind(kakaoChannelService), this.getKakaoChannel = kakaoChannelService.getKakaoChannel.bind(kakaoChannelService), this.requestKakaoChannelToken = kakaoChannelService.requestKakaoChannelToken.bind(kakaoChannelService), this.createKakaoChannel = kakaoChannelService.createKakaoChannel.bind(kakaoChannelService), this.removeKakaoChannel = kakaoChannelService.removeKakaoChannel.bind(kakaoChannelService), this.getKakaoAlimtalkTemplateCategories = kakaoTemplateService.getKakaoAlimtalkTemplateCategories.bind(
      kakaoTemplateService
    ), this.createKakaoAlimtalkTemplate = kakaoTemplateService.createKakaoAlimtalkTemplate.bind(
      kakaoTemplateService
    ), this.getKakaoAlimtalkTemplates = kakaoTemplateService.getKakaoAlimtalkTemplates.bind(kakaoTemplateService), this.getKakaoAlimtalkTemplate = kakaoTemplateService.getKakaoAlimtalkTemplate.bind(kakaoTemplateService), this.cancelInspectionKakaoAlimtalkTemplate = kakaoTemplateService.cancelInspectionKakaoAlimtalkTemplate.bind(
      kakaoTemplateService
    ), this.updateKakaoAlimtalkTemplate = kakaoTemplateService.updateKakaoAlimtalkTemplate.bind(
      kakaoTemplateService
    ), this.updateKakaoAlimtalkTemplateName = kakaoTemplateService.updateKakaoAlimtalkTemplateName.bind(
      kakaoTemplateService
    ), this.removeKakaoAlimtalkTemplate = kakaoTemplateService.removeKakaoAlimtalkTemplate.bind(
      kakaoTemplateService
    ), this.createGroup = groupService.createGroup.bind(groupService), this.addMessagesToGroup = groupService.addMessagesToGroup.bind(groupService), this.sendGroup = groupService.sendGroup.bind(groupService), this.reserveGroup = groupService.reserveGroup.bind(groupService), this.removeReservationToGroup = groupService.removeReservationToGroup.bind(groupService), this.getGroups = groupService.getGroups.bind(groupService), this.getGroup = groupService.getGroup.bind(groupService), this.getGroupMessages = groupService.getGroupMessages.bind(groupService), this.removeGroupMessages = groupService.removeGroupMessages.bind(groupService), this.removeGroup = groupService.removeGroup.bind(groupService), this.send = messageService.send.bind(messageService), this.getMessages = messageService.getMessages.bind(messageService), this.getStatistics = messageService.getStatistics.bind(messageService), this.uploadFile = storageService.uploadFile.bind(storageService);
  }
};

exports.ApiKeyError = ApiKeyError;
exports.BMS_ACCEPTABLE_FIELDS = BMS_ACCEPTABLE_FIELDS;
exports.BMS_ADDITIONAL_CONTENT_MAX = BMS_ADDITIONAL_CONTENT_MAX;
exports.BMS_ALLOWED_LINK_TYPES = BMS_ALLOWED_LINK_TYPES;
exports.BMS_BUTTON_COUNT = BMS_BUTTON_COUNT;
exports.BMS_BUTTON_NAME_MAX = BMS_BUTTON_NAME_MAX;
exports.BMS_CAROUSEL_COMMERCE_ADDITIONAL_CONTENT_MAX = BMS_CAROUSEL_COMMERCE_ADDITIONAL_CONTENT_MAX;
exports.BMS_CAROUSEL_FEED_ITEM_CONTENT_MAX = BMS_CAROUSEL_FEED_ITEM_CONTENT_MAX;
exports.BMS_CAROUSEL_HEAD_CONTENT_MAX = BMS_CAROUSEL_HEAD_CONTENT_MAX;
exports.BMS_CAROUSEL_LIST_RANGE = BMS_CAROUSEL_LIST_RANGE;
exports.BMS_CONTENT_MAX = BMS_CONTENT_MAX;
exports.BMS_COUPON_DESCRIPTION_MAX = BMS_COUPON_DESCRIPTION_MAX;
exports.BMS_HEADER_MAX = BMS_HEADER_MAX;
exports.BMS_MAIN_WIDE_ITEM_TITLE_MAX = BMS_MAIN_WIDE_ITEM_TITLE_MAX;
exports.BMS_SUB_WIDE_ITEM_TITLE_MAX = BMS_SUB_WIDE_ITEM_TITLE_MAX;
exports.BadRequestError = BadRequestError;
exports.ClientError = ClientError;
exports.DefaultError = DefaultError;
exports.InvalidDateError = InvalidDateError;
exports.MessageNotReceivedError = MessageNotReceivedError;
exports.NetworkError = NetworkError;
exports.ResponseSchemaMismatchError = ResponseSchemaMismatchError;
exports.ServerError = ServerError;
exports.SolapiMessageService = SolapiMessageService;
exports.UnexpectedDefectError = UnexpectedDefectError;
exports.UnhandledExitError = UnhandledExitError;
exports.VariableValidationError = VariableValidationError;
exports.addMessageResponseSchema = addMessageResponseSchema;
exports.addMessageResultSchema = addMessageResultSchema;
exports.additionalBodySchema = additionalBodySchema;
exports.appSchema = appSchema;
exports.baseKakaoOptionSchema = baseKakaoOptionSchema;
exports.blackSchema = blackSchema;
exports.blockGroupSchema = blockGroupSchema;
exports.blockNumberSchema = blockNumberSchema;
exports.bmsAppButtonSchema = bmsAppButtonSchema;
exports.bmsBotKeywordButtonSchema = bmsBotKeywordButtonSchema;
exports.bmsBotTransferButtonSchema = bmsBotTransferButtonSchema;
exports.bmsBusinessFormButtonSchema = bmsBusinessFormButtonSchema;
exports.bmsButtonLinkTypeSchema = bmsButtonLinkTypeSchema;
exports.bmsButtonSchema = bmsButtonSchema;
exports.bmsCarouselCommerceItemSchema = bmsCarouselCommerceItemSchema;
exports.bmsCarouselCommerceSchema = bmsCarouselCommerceSchema;
exports.bmsCarouselFeedItemSchema = bmsCarouselFeedItemSchema;
exports.bmsCarouselFeedSchema = bmsCarouselFeedSchema;
exports.bmsCarouselHeadSchema = bmsCarouselHeadSchema;
exports.bmsCarouselTailSchema = bmsCarouselTailSchema;
exports.bmsChannelAddButtonSchema = bmsChannelAddButtonSchema;
exports.bmsChatBubbleTypeSchema = bmsChatBubbleTypeSchema;
exports.bmsCommerceSchema = bmsCommerceSchema;
exports.bmsConsultButtonSchema = bmsConsultButtonSchema;
exports.bmsCouponSchema = bmsCouponSchema;
exports.bmsCouponTitleSchema = bmsCouponTitleSchema;
exports.bmsLinkButtonSchema = bmsLinkButtonSchema;
exports.bmsMainWideItemSchema = bmsMainWideItemSchema;
exports.bmsMessageDeliveryButtonSchema = bmsMessageDeliveryButtonSchema;
exports.bmsSubWideItemSchema = bmsSubWideItemSchema;
exports.bmsVideoSchema = bmsVideoSchema;
exports.bmsWebButtonSchema = bmsWebButtonSchema;
exports.commonCashResponseSchema = commonCashResponseSchema;
exports.countForChargeSchema = countForChargeSchema;
exports.countSchema = countSchema;
exports.createGroupRequestSchema = createGroupRequestSchema;
exports.createKakaoAlimtalkTemplateRequestSchema = createKakaoAlimtalkTemplateRequestSchema;
exports.createKakaoChannelRequestSchema = createKakaoChannelRequestSchema;
exports.createKakaoChannelResponseSchema = createKakaoChannelResponseSchema;
exports.createKakaoChannelTokenRequestSchema = createKakaoChannelTokenRequestSchema;
exports.datePayloadSchema = datePayloadSchema;
exports.dateTypeSchema = dateTypeSchema;
exports.decodeKakaoAlimtalkTemplate = decodeKakaoAlimtalkTemplate;
exports.decodeKakaoChannel = decodeKakaoChannel;
exports.defaultAgentTypeSchema = defaultAgentTypeSchema;
exports.defaultMessageRequestSchema = defaultMessageRequestSchema;
exports.detailGroupMessageResponseSchema = detailGroupMessageResponseSchema;
exports.failedMessageSchema = failedMessageSchema;
exports.fileIdsSchema = fileIdsSchema;
exports.fileTypeSchema = fileTypeSchema;
exports.fileUploadRequestSchema = fileUploadRequestSchema;
exports.fileUploadResponseSchema = fileUploadResponseSchema;
exports.finalizeGetBlacksRequest = finalizeGetBlacksRequest;
exports.finalizeGetBlockGroupsRequest = finalizeGetBlockGroupsRequest;
exports.finalizeGetBlockNumbersRequest = finalizeGetBlockNumbersRequest;
exports.finalizeGetGroupsRequest = finalizeGetGroupsRequest;
exports.finalizeGetKakaoAlimtalkTemplatesRequest = finalizeGetKakaoAlimtalkTemplatesRequest;
exports.finalizeGetKakaoChannelsRequest = finalizeGetKakaoChannelsRequest;
exports.finalizeGetMessagesRequest = finalizeGetMessagesRequest;
exports.finalizeGetStatisticsRequest = finalizeGetStatisticsRequest;
exports.getBalanceResponseSchema = getBalanceResponseSchema;
exports.getBlacksRequestSchema = getBlacksRequestSchema;
exports.getBlacksResponseSchema = getBlacksResponseSchema;
exports.getBlockGroupsRequestSchema = getBlockGroupsRequestSchema;
exports.getBlockGroupsResponseSchema = getBlockGroupsResponseSchema;
exports.getBlockNumbersRequestSchema = getBlockNumbersRequestSchema;
exports.getBlockNumbersResponseSchema = getBlockNumbersResponseSchema;
exports.getGroupMessagesRequestSchema = getGroupMessagesRequestSchema;
exports.getGroupsRequestSchema = getGroupsRequestSchema;
exports.getGroupsResponseSchema = getGroupsResponseSchema;
exports.getKakaoAlimtalkTemplatesRequestSchema = getKakaoAlimtalkTemplatesRequestSchema;
exports.getKakaoAlimtalkTemplatesResponseSchema = getKakaoAlimtalkTemplatesResponseSchema;
exports.getKakaoChannelsRequestSchema = getKakaoChannelsRequestSchema;
exports.getKakaoChannelsResponseSchema = getKakaoChannelsResponseSchema;
exports.getKakaoTemplateResponseSchema = getKakaoTemplateResponseSchema;
exports.getMessagesRequestSchema = getMessagesRequestSchema;
exports.getMessagesResponseSchema = getMessagesResponseSchema;
exports.getStatisticsRequestSchema = getStatisticsRequestSchema;
exports.getStatisticsResponseSchema = getStatisticsResponseSchema;
exports.groupIdSchema = groupIdSchema;
exports.groupMessageAddRequestSchema = groupMessageAddRequestSchema;
exports.groupMessageResponseSchema = groupMessageResponseSchema;
exports.groupSchema = groupSchema;
exports.handleKeySchema = handleKeySchema;
exports.isErrorResponse = isErrorResponse;
exports.kakaoAlimtalkTemplateAssignTypeSchema = kakaoAlimtalkTemplateAssignTypeSchema;
exports.kakaoAlimtalkTemplateCommentTypeSchema = kakaoAlimtalkTemplateCommentTypeSchema;
exports.kakaoAlimtalkTemplateEmphasizeTypeSchema = kakaoAlimtalkTemplateEmphasizeTypeSchema;
exports.kakaoAlimtalkTemplateHighlightTypeSchema = kakaoAlimtalkTemplateHighlightTypeSchema;
exports.kakaoAlimtalkTemplateItemTypeSchema = kakaoAlimtalkTemplateItemTypeSchema;
exports.kakaoAlimtalkTemplateMessageTypeSchema = kakaoAlimtalkTemplateMessageTypeSchema;
exports.kakaoAlimtalkTemplateQuickReplyAppLinkSchema = kakaoAlimtalkTemplateQuickReplyAppLinkSchema;
exports.kakaoAlimtalkTemplateQuickReplyDefaultSchema = kakaoAlimtalkTemplateQuickReplyDefaultSchema;
exports.kakaoAlimtalkTemplateQuickReplySchema = kakaoAlimtalkTemplateQuickReplySchema;
exports.kakaoAlimtalkTemplateQuickReplyWebLinkSchema = kakaoAlimtalkTemplateQuickReplyWebLinkSchema;
exports.kakaoAlimtalkTemplateSchema = kakaoAlimtalkTemplateSchema;
exports.kakaoAlimtalkTemplateStatusSchema = kakaoAlimtalkTemplateStatusSchema;
exports.kakaoButtonSchema = kakaoButtonSchema;
exports.kakaoChannelCategorySchema = kakaoChannelCategorySchema;
exports.kakaoChannelSchema = kakaoChannelSchema;
exports.logSchema = logSchema;
exports.messageResponseItemSchema = messageResponseItemSchema;
exports.messageSchema = messageSchema;
exports.messageTypeRecordSchema = messageTypeRecordSchema;
exports.messageTypeSchema = messageTypeSchema;
exports.multipleMessageSendingRequestSchema = multipleMessageSendingRequestSchema;
exports.naverOptionSchema = naverOptionSchema;
exports.osPlatform = osPlatform;
exports.phoneNumberSchema = phoneNumberSchema;
exports.rcsButtonSchema = rcsButtonSchema;
exports.rcsOptionRequestSchema = rcsOptionRequestSchema;
exports.rcsOptionSchema = rcsOptionSchema;
exports.removeGroupMessagesResponseSchema = removeGroupMessagesResponseSchema;
exports.removeMessageIdsToGroupRequestSchema = removeMessageIdsToGroupRequestSchema;
exports.requestKakaoChannelTokenResponseSchema = requestKakaoChannelTokenResponseSchema;
exports.requestSendMessageSchema = requestSendMessageSchema;
exports.requestSendOneMessageSchema = requestSendOneMessageSchema;
exports.scheduledDateSendingRequestSchema = scheduledDateSendingRequestSchema;
exports.sdkVersion = sdkVersion;
exports.sendRequestConfigSchema = sendRequestConfigSchema;
exports.storedMessageSchema = storedMessageSchema;
exports.transformVariables = transformVariables;
exports.updateKakaoAlimtalkTemplateRequestSchema = updateKakaoAlimtalkTemplateRequestSchema;
exports.validateAcceptableFields = validateAcceptableFields;
exports.validateAllowedLinkTypes = validateAllowedLinkTypes;
exports.validateButtonCount = validateButtonCount;
exports.validateButtonNames = validateButtonNames;
exports.validateCarouselListCount = validateCarouselListCount;
exports.validateCouponDescription = validateCouponDescription;
exports.validateForbiddenVariables = validateForbiddenVariables;
exports.validateImageIdLength = validateImageIdLength;
exports.validateLinks = validateLinks;
exports.validateNewlineLimits = validateNewlineLimits;
exports.validateTextLengths = validateTextLengths;
exports.validateVariableNames = validateVariableNames;
exports.voiceOptionSchema = voiceOptionSchema;
