// solapi/dist/index.mjs
import { Data, Schema, ParseResult, Effect, pipe, Array as Array$1, Exit, Match, Schedule, Cause, Chunk } from 'effect';
import { createHmac, randomBytes } from 'crypto';
import { formatISO, parseISO } from 'date-fns';
import * as Effect10 from 'effect/Effect';
import { promises } from 'fs';
import { URL as URL$1 } from 'url';

// src/lib/effectErrorHandler.ts
var InvalidDateError = class extends Data.TaggedError("InvalidDateError") {
  toString() {
    return `InvalidDateError: ${this.message}${this.originalValue ? ` (value: ${this.originalValue})` : ""}`;
  }
}, ApiKeyError = class extends Data.TaggedError("ApiKeyError") {
  toString() {
    return `ApiKeyError: ${this.message}`;
  }
}, DefaultError = class extends Data.TaggedError("DefaultError") {
  get message() {
    return `${this.errorCode}: ${this.errorMessage}`;
  }
  toString() {
    return this.message;
  }
}, MessageNotReceivedError = class extends Data.TaggedError(
  "MessageNotReceivedError"
) {
  get message() {
    return `${this.totalCount}\uAC1C\uC758 \uBA54\uC2DC\uC9C0\uAC00 \uC811\uC218\uB418\uC9C0 \uBABB\uD588\uC2B5\uB2C8\uB2E4. \uC790\uC138\uD55C \uC5D0\uB7EC \uBA54\uC2DC\uC9C0\uB294 \uD574\uB2F9 \uC5D0\uB7EC \uB0B4 failedMessageList\uB97C \uD655\uC778\uD574\uC8FC\uC138\uC694.`;
  }
  toString() {
    return `MessageNotReceivedError: ${this.message}
Failed messages: ${this.failedMessageList.length}`;
  }
}, BadRequestError = class extends Data.TaggedError("BadRequestError") {
  toString() {
    return `BadRequestError: ${this.message}${this.validationErrors ? `
Validation errors: ${this.validationErrors.join(", ")}` : ""}`;
  }
}, NetworkError = class extends Data.TaggedError("NetworkError") {
  get message() {
    return `${this.method} ${this.url} \uC694\uCCAD \uC2E4\uD328 - ${this.cause}`;
  }
  toString() {
    return `NetworkError: ${this.message}`;
  }
}, ClientError = class extends Data.TaggedError("ClientError") {
  get message() {
    return `${this.errorCode}: ${this.errorMessage}`;
  }
  toString() {
    return this.message;
  }
}, UnexpectedDefectError = class extends Data.TaggedError(
  "UnexpectedDefectError"
) {
  toString() {
    return `UnexpectedDefectError: ${this.message}`;
  }
}, UnhandledExitError = class extends Data.TaggedError("UnhandledExitError") {
  toString() {
    return `UnhandledExitError: ${this.message}`;
  }
}, ResponseSchemaMismatchError = class extends Data.TaggedError(
  "ResponseSchemaMismatchError"
) {
  toString() {
    let header = `ResponseSchemaMismatchError: ${this.message}`, url = this.url ? `
URL: ${this.url}` : "", issues = this.validationErrors.length > 0 ? `
Issues:
- ${this.validationErrors.join(`
- `)}` : "", env = "production"?.trim().toLowerCase(), body = (env === "development" || env === "test") && this.responseBody ? `
Response: ${this.responseBody.substring(0, 500)}` : "";
    return `${header}${url}${issues}${body}`;
  }
}, ServerError = class extends Data.TaggedError("ServerError") {
  get message() {
    return `${this.errorCode} - ${this.errorMessage}`;
  }
  toString() {
    return `ServerError(${this.httpStatus}): ${this.message}`;
  }
}, isErrorResponse = (value) => value == null || typeof value != "object" || !("errorCode" in value) || !("errorMessage" in value) ? false : typeof value.errorCode == "string" && value.errorCode !== "" && typeof value.errorMessage == "string" && value.errorMessage !== "";

// src/lib/effectErrorHandler.ts
var isTaggedDefect = (value) => value !== null && typeof value == "object" && "_tag" in value && typeof value._tag == "string", extractDefectInfo = (defect) => {
  if (isTaggedDefect(defect)) {
    let tag = defect._tag, message = defect.message != null ? String(defect.message) : "";
    return {
      summary: `${tag}${message ? `: ${message}` : ""}`,
      details: `Tagged Error [${tag}]: ${JSON.stringify(defect, null, 2)}`
    };
  }
  if (defect !== null && typeof defect == "object") {
    let keys = Object.keys(defect);
    return {
      summary: keys.length > 0 ? `Object with keys: ${keys.slice(0, 3).join(", ")}${keys.length > 3 ? "..." : ""}` : "Empty object",
      details: JSON.stringify(defect, null, 2)
    };
  }
  return {
    summary: String(defect),
    details: `Value (${typeof defect}): ${String(defect)}`
  };
}, unwrapCause = (cause) => {
  let failure = Cause.failureOption(cause);
  if (failure._tag === "Some")
    return failure.value;
  let defects = Cause.defects(cause);
  if (defects.length > 0) {
    let firstDefect = Chunk.unsafeGet(defects, 0);
    if (firstDefect instanceof Error)
      return firstDefect;
    let defectInfo = extractDefectInfo(firstDefect), message2 = `Unexpected error: ${defectInfo.summary}` ;
    return new UnexpectedDefectError({ message: message2 });
  }
  let message = "Effect execution failed unexpectedly" ;
  return new UnhandledExitError({ message });
}, runSafePromise = (effect) => Effect.runPromiseExit(effect).then(
  Exit.match({
    onFailure: (cause) => Promise.reject(unwrapCause(cause)),
    onSuccess: (value) => Promise.resolve(value)
  })
);
var countSchema = Schema.Struct({
  total: Schema.Number,
  sentTotal: Schema.Number,
  sentFailed: Schema.Number,
  sentSuccess: Schema.Number,
  sentPending: Schema.Number,
  sentReplacement: Schema.Number,
  refund: Schema.Number,
  registeredFailed: Schema.Number,
  registeredSuccess: Schema.Number
}), countryChargeStatusSchema = Schema.Record({
  key: Schema.String,
  value: Schema.Number
}), countForChargeSchema = Schema.Struct({
  sms: countryChargeStatusSchema,
  lms: countryChargeStatusSchema,
  mms: countryChargeStatusSchema,
  ata: countryChargeStatusSchema,
  cta: countryChargeStatusSchema,
  cti: countryChargeStatusSchema,
  nsa: countryChargeStatusSchema,
  rcs_sms: countryChargeStatusSchema,
  rcs_lms: countryChargeStatusSchema,
  rcs_mms: countryChargeStatusSchema,
  rcs_tpl: countryChargeStatusSchema,
  rcs_itpl: Schema.optional(countryChargeStatusSchema),
  rcs_ltpl: Schema.optional(countryChargeStatusSchema),
  fax: Schema.optional(countryChargeStatusSchema),
  voice: Schema.optional(countryChargeStatusSchema),
  bms_text: Schema.optional(countryChargeStatusSchema),
  bms_image: Schema.optional(countryChargeStatusSchema),
  bms_wide: Schema.optional(countryChargeStatusSchema),
  bms_wide_item_list: Schema.optional(countryChargeStatusSchema),
  bms_carousel_feed: Schema.optional(countryChargeStatusSchema),
  bms_premium_video: Schema.optional(countryChargeStatusSchema),
  bms_commerce: Schema.optional(countryChargeStatusSchema),
  bms_carousel_commerce: Schema.optional(countryChargeStatusSchema),
  bms_free: Schema.optional(countryChargeStatusSchema)
}), commonCashResponseSchema = Schema.Struct({
  requested: Schema.Number,
  replacement: Schema.Number,
  refund: Schema.Number,
  sum: Schema.Number
}), messageTypeRecordSchema = Schema.Struct({
  sms: Schema.Number,
  lms: Schema.Number,
  mms: Schema.Number,
  ata: Schema.Number,
  cta: Schema.Number,
  cti: Schema.Number,
  nsa: Schema.Number,
  rcs_sms: Schema.Number,
  rcs_lms: Schema.Number,
  rcs_mms: Schema.Number,
  rcs_tpl: Schema.Number,
  rcs_itpl: Schema.optional(Schema.Number),
  rcs_ltpl: Schema.optional(Schema.Number),
  fax: Schema.optional(Schema.Number),
  voice: Schema.optional(Schema.Number),
  bms_text: Schema.optional(Schema.Number),
  bms_image: Schema.optional(Schema.Number),
  bms_wide: Schema.optional(Schema.Number),
  bms_wide_item_list: Schema.optional(Schema.Number),
  bms_carousel_feed: Schema.optional(Schema.Number),
  bms_premium_video: Schema.optional(Schema.Number),
  bms_commerce: Schema.optional(Schema.Number),
  bms_carousel_commerce: Schema.optional(Schema.Number),
  bms_free: Schema.optional(Schema.Number)
}), partialMessageTypeRecordSchema = Schema.partial(
  messageTypeRecordSchema
), appSchema = Schema.Struct({
  profit: messageTypeRecordSchema,
  appId: Schema.NullishOr(Schema.String)
}), logSchema = Schema.Array(
  Schema.Record({ key: Schema.String, value: Schema.Unknown })
), groupIdSchema = Schema.String, groupSchema = Schema.Struct({
  count: countSchema,
  balance: commonCashResponseSchema,
  point: commonCashResponseSchema,
  app: appSchema,
  sdkVersion: Schema.NullishOr(Schema.String),
  osPlatform: Schema.NullishOr(Schema.String),
  log: logSchema,
  status: Schema.String,
  scheduledDate: Schema.NullishOr(Schema.String),
  dateSent: Schema.NullishOr(Schema.String),
  dateCompleted: Schema.NullishOr(Schema.String),
  isRefunded: Schema.Boolean,
  groupId: groupIdSchema,
  accountId: Schema.String,
  countForCharge: countForChargeSchema,
  dateCreated: Schema.String,
  dateUpdated: Schema.String
}), handleKeySchema = Schema.String, blackSchema = Schema.Struct({
  handleKey: handleKeySchema,
  type: Schema.Literal("DENIAL"),
  senderNumber: Schema.String,
  recipientNumber: Schema.String,
  dateCreated: Schema.String,
  dateUpdated: Schema.String
}), blockGroupSchema = Schema.Struct({
  blockGroupId: Schema.String,
  accountId: Schema.String,
  status: Schema.Literal("INACTIVE", "ACTIVE"),
  name: Schema.NullishOr(Schema.String),
  useAll: Schema.Boolean,
  senderNumbers: Schema.Array(Schema.String),
  dateCreated: Schema.String,
  dateUpdated: Schema.String
}), blockNumberSchema = Schema.Struct({
  blockNumberId: Schema.String,
  accountId: Schema.String,
  memo: Schema.String,
  phoneNumber: Schema.String,
  blockGroupIds: Schema.Array(Schema.String),
  dateCreated: Schema.String,
  dateUpdated: Schema.String
});
var bmsButtonLinkTypeSchema = Schema.Literal(
  "AC",
  "WL",
  "AL",
  "BK",
  "MD",
  "BC",
  "BT",
  "BF"
), bmsWebButtonSchema = Schema.Struct({
  name: Schema.String,
  linkType: Schema.Literal("WL"),
  linkMobile: Schema.String,
  linkPc: Schema.optional(Schema.String),
  targetOut: Schema.optional(Schema.Boolean)
}), bmsAppButtonSchema = Schema.Struct({
  name: Schema.String,
  linkType: Schema.Literal("AL"),
  linkMobile: Schema.optional(Schema.String),
  linkAndroid: Schema.optional(Schema.String),
  linkIos: Schema.optional(Schema.String),
  targetOut: Schema.optional(Schema.Boolean)
}).pipe(
  Schema.filter((button) => button.linkMobile !== void 0 && button.linkMobile.trim() !== "" || button.linkAndroid !== void 0 && button.linkAndroid.trim() !== "" || button.linkIos !== void 0 && button.linkIos.trim() !== "" ? true : "AL \uD0C0\uC785 \uBC84\uD2BC\uC740 linkMobile, linkAndroid, linkIos \uC911 \uD558\uB098 \uC774\uC0C1 \uD544\uC218\uC785\uB2C8\uB2E4.")
), bmsChannelAddButtonSchema = Schema.Struct({
  name: Schema.optional(Schema.String),
  linkType: Schema.Literal("AC")
}), bmsBotKeywordButtonSchema = Schema.Struct({
  name: Schema.String,
  linkType: Schema.Literal("BK"),
  chatExtra: Schema.optional(Schema.String)
}), bmsMessageDeliveryButtonSchema = Schema.Struct({
  name: Schema.String,
  linkType: Schema.Literal("MD"),
  chatExtra: Schema.optional(Schema.String)
}), bmsConsultButtonSchema = Schema.Struct({
  name: Schema.String,
  linkType: Schema.Literal("BC"),
  chatExtra: Schema.optional(Schema.String)
}), bmsBotTransferButtonSchema = Schema.Struct({
  name: Schema.String,
  linkType: Schema.Literal("BT"),
  chatExtra: Schema.optional(Schema.String)
}), bmsBusinessFormButtonSchema = Schema.Struct({
  name: Schema.String,
  linkType: Schema.Literal("BF")
}), bmsButtonSchema = Schema.Union(
  bmsWebButtonSchema,
  bmsAppButtonSchema,
  bmsChannelAddButtonSchema,
  bmsBotKeywordButtonSchema,
  bmsMessageDeliveryButtonSchema,
  bmsConsultButtonSchema,
  bmsBotTransferButtonSchema,
  bmsBusinessFormButtonSchema
), bmsLinkButtonSchema = Schema.Union(
  bmsWebButtonSchema,
  bmsAppButtonSchema
);
var NumberOrNumericString = Schema.transformOrFail(
  Schema.Union(Schema.Number, Schema.String),
  Schema.Number,
  {
    strict: true,
    decode: (input, _, ast) => {
      if (typeof input == "number")
        return ParseResult.succeed(input);
      if (input.trim() === "")
        return ParseResult.fail(
          new ParseResult.Type(ast, input, "\uC720\uD6A8\uD55C \uC22B\uC790 \uD615\uC2DD\uC774 \uC544\uB2D9\uB2C8\uB2E4.")
        );
      let parsed = parseFloat(input);
      return Number.isNaN(parsed) ? ParseResult.fail(
        new ParseResult.Type(ast, input, "\uC720\uD6A8\uD55C \uC22B\uC790 \uD615\uC2DD\uC774 \uC544\uB2D9\uB2C8\uB2E4.")
      ) : ParseResult.succeed(parsed);
    },
    encode: (n) => ParseResult.succeed(n)
  }
), COMMERCE_PRICE_MAX = 99999999, DISCOUNT_RATE_MAX = 100, numberInRange = (fieldName, min, max) => NumberOrNumericString.pipe(
  Schema.filter((n) => n >= min && n <= max, {
    message: () => `${fieldName} \uAC12\uC774 \uC798\uBABB\uB418\uC5C8\uC2B5\uB2C8\uB2E4. ${min} \uC774\uC0C1 ${max} \uC774\uD558\uC758 \uC22B\uC790\uC5EC\uC57C \uD569\uB2C8\uB2E4.`
  })
), validateCommercePricingCombination = (commerce) => {
  let hasDiscountPrice = commerce.discountPrice !== void 0, hasDiscountRate = commerce.discountRate !== void 0, hasDiscountFixed = commerce.discountFixed !== void 0;
  return !hasDiscountPrice && !hasDiscountRate && !hasDiscountFixed || hasDiscountPrice && hasDiscountRate && !hasDiscountFixed || hasDiscountPrice && hasDiscountFixed && !hasDiscountRate ? true : hasDiscountRate && hasDiscountFixed ? "discountRate\uC640 discountFixed\uB294 \uB3D9\uC2DC\uC5D0 \uC0AC\uC6A9\uD560 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4. \uD560\uC778\uC728(discountRate) \uB610\uB294 \uC815\uC561\uD560\uC778(discountFixed) \uC911 \uD558\uB098\uB9CC \uC120\uD0DD\uD558\uC138\uC694." : !hasDiscountPrice && (hasDiscountRate || hasDiscountFixed) ? "discountRate \uB610\uB294 discountFixed\uB97C \uC0AC\uC6A9\uD558\uB824\uBA74 discountPrice(\uD560\uC778\uAC00)\uB3C4 \uD568\uAED8 \uC9C0\uC815\uD574\uC57C \uD569\uB2C8\uB2E4." : hasDiscountPrice && !hasDiscountRate && !hasDiscountFixed ? "discountPrice\uB97C \uC0AC\uC6A9\uD558\uB824\uBA74 discountRate(\uD560\uC778\uC728) \uB610\uB294 discountFixed(\uC815\uC561\uD560\uC778) \uC911 \uD558\uB098\uB97C \uD568\uAED8 \uC9C0\uC815\uD574\uC57C \uD569\uB2C8\uB2E4." : "\uC54C \uC218 \uC5C6\uB294 \uAC00\uACA9 \uC870\uD569\uC785\uB2C8\uB2E4. regularPrice\uB9CC \uC0AC\uC6A9\uD558\uAC70\uB098, regularPrice + discountPrice + discountRate/discountFixed \uC870\uD569\uC744 \uC0AC\uC6A9\uD558\uC138\uC694.";
}, bmsCommerceSchema = Schema.Struct({
  title: Schema.String,
  regularPrice: numberInRange("regularPrice", 0, COMMERCE_PRICE_MAX),
  discountPrice: Schema.optional(
    numberInRange("discountPrice", 0, COMMERCE_PRICE_MAX)
  ),
  discountRate: Schema.optional(
    numberInRange("discountRate", 0, DISCOUNT_RATE_MAX)
  ),
  discountFixed: Schema.optional(
    numberInRange("discountFixed", 0, COMMERCE_PRICE_MAX)
  )
}).pipe(Schema.filter(validateCommercePricingCombination));
var wonDiscountPattern = /^([1-9]\d{0,7})원 할인 쿠폰$/, percentDiscountPattern = /^([1-9]\d?|100)% 할인 쿠폰$/, freeCouponPattern = /^\S{1,7} 무료 쿠폰$/, upCouponPattern = /^\S{1,7} UP 쿠폰$/, isValidCouponTitle = (title) => {
  if (title === "\uBC30\uC1A1\uBE44 \uD560\uC778 \uCFE0\uD3F0") return true;
  let wonMatch = title.match(wonDiscountPattern);
  if (wonMatch) {
    let num = parseInt(wonMatch[1], 10);
    return num >= 1 && num <= 99999999;
  }
  return percentDiscountPattern.test(title) || freeCouponPattern.test(title) ? true : upCouponPattern.test(title);
}, bmsCouponTitleSchema = Schema.String.pipe(
  Schema.filter(isValidCouponTitle, {
    message: () => '\uCFE0\uD3F0 \uC81C\uBAA9\uC740 \uB2E4\uC74C \uD615\uC2DD \uC911 \uD558\uB098\uC5EC\uC57C \uD569\uB2C8\uB2E4: "N\uC6D0 \uD560\uC778 \uCFE0\uD3F0" (1~99999999), "N% \uD560\uC778 \uCFE0\uD3F0" (1~100), "\uBC30\uC1A1\uBE44 \uD560\uC778 \uCFE0\uD3F0", "OOO \uBB34\uB8CC \uCFE0\uD3F0" (7\uC790 \uC774\uB0B4), "OOO UP \uCFE0\uD3F0" (7\uC790 \uC774\uB0B4)'
  })
), bmsCouponSchema = Schema.Struct({
  title: bmsCouponTitleSchema,
  description: Schema.String,
  linkMobile: Schema.optional(Schema.String),
  linkPc: Schema.optional(Schema.String),
  linkAndroid: Schema.optional(Schema.String),
  linkIos: Schema.optional(Schema.String)
});

// src/models/base/kakao/bms/bmsCarousel.ts
var bmsCarouselHeadSchema = Schema.Struct({
  header: Schema.String,
  content: Schema.String,
  imageId: Schema.String,
  linkMobile: Schema.optional(Schema.String),
  linkPc: Schema.optional(Schema.String),
  linkAndroid: Schema.optional(Schema.String),
  linkIos: Schema.optional(Schema.String)
}).pipe(
  Schema.filter((head) => (head.linkPc !== void 0 || head.linkAndroid !== void 0 || head.linkIos !== void 0) && head.linkMobile === void 0 ? "linkPc, linkAndroid, linkIos \uC911 \uD558\uB098\uB77C\uB3C4 \uC788\uC73C\uBA74 linkMobile \uAC12\uC774 \uD544\uC218\uC785\uB2C8\uB2E4." : true)
), bmsCarouselTailSchema = Schema.Struct({
  linkMobile: Schema.String,
  linkPc: Schema.optional(Schema.String),
  linkAndroid: Schema.optional(Schema.String),
  linkIos: Schema.optional(Schema.String)
}), bmsCarouselFeedItemSchema = Schema.Struct({
  header: Schema.String,
  content: Schema.String,
  imageId: Schema.String,
  imageLink: Schema.optional(Schema.String),
  buttons: Schema.Array(bmsLinkButtonSchema),
  coupon: Schema.optional(bmsCouponSchema)
}), bmsCarouselCommerceItemSchema = Schema.Struct({
  commerce: bmsCommerceSchema,
  imageId: Schema.String,
  imageLink: Schema.optional(Schema.String),
  buttons: Schema.Array(bmsLinkButtonSchema),
  additionalContent: Schema.optional(Schema.String),
  coupon: Schema.optional(bmsCouponSchema)
}), bmsCarouselFeedSchema = Schema.Struct({
  list: Schema.Array(bmsCarouselFeedItemSchema),
  tail: Schema.optional(bmsCarouselTailSchema)
}), bmsCarouselCommerceSchema = Schema.Struct({
  head: Schema.optional(bmsCarouselHeadSchema),
  list: Schema.Array(bmsCarouselCommerceItemSchema),
  tail: Schema.optional(bmsCarouselTailSchema)
});
var bmsChatBubbleTypeSchema = Schema.Literal(
  "TEXT",
  "IMAGE",
  "WIDE",
  "WIDE_ITEM_LIST",
  "COMMERCE",
  "CAROUSEL_FEED",
  "CAROUSEL_COMMERCE",
  "PREMIUM_VIDEO"
);

// src/models/base/kakao/bms/bmsConstraints.ts
var BMS_COUPON_DESCRIPTION_MAX = {
  TEXT: 12,
  IMAGE: 12,
  WIDE: 18,
  WIDE_ITEM_LIST: 18,
  COMMERCE: 12,
  CAROUSEL_FEED: 12,
  CAROUSEL_COMMERCE: 12,
  PREMIUM_VIDEO: 12
}, BMS_BUTTON_COUNT = {
  TEXT: { max: 5, maxWithCoupon: 4 },
  IMAGE: { max: 5, maxWithCoupon: 4 },
  WIDE: { max: 2 },
  WIDE_ITEM_LIST: { max: 2 },
  COMMERCE: { max: 2, min: 1 },
  CAROUSEL_FEED: { max: 2, min: 1 },
  CAROUSEL_COMMERCE: { max: 2, min: 1 },
  PREMIUM_VIDEO: { max: 1 }
}, BMS_BUTTON_NAME_MAX = {
  TEXT: 14,
  IMAGE: 14,
  WIDE: 8,
  WIDE_ITEM_LIST: 8,
  COMMERCE: 8,
  CAROUSEL_FEED: 8,
  CAROUSEL_COMMERCE: 8,
  PREMIUM_VIDEO: 8
}, BMS_ALLOWED_LINK_TYPES = {
  COMMERCE: ["WL", "AL"],
  CAROUSEL_FEED: ["WL", "AL"],
  CAROUSEL_COMMERCE: ["WL", "AL"]
}, BMS_HEADER_MAX = 20, BMS_ADDITIONAL_CONTENT_MAX = 34, BMS_MAIN_WIDE_ITEM_TITLE_MAX = 25, BMS_SUB_WIDE_ITEM_TITLE_MAX = 30, BMS_CAROUSEL_HEAD_CONTENT_MAX = 50, BMS_CAROUSEL_FEED_ITEM_CONTENT_MAX = 180, BMS_CAROUSEL_COMMERCE_ADDITIONAL_CONTENT_MAX = 34, BMS_CONTENT_MAX = {
  TEXT: 1300,
  IMAGE: 1300,
  WIDE: 76,
  PREMIUM_VIDEO: 76
}, BMS_CAROUSEL_LIST_RANGE = {
  withHead: { min: 1, max: 5 },
  withoutHead: { min: 2, max: 6 }
}, hasCoupon = (bms) => bms.coupon != null, validateCouponDescription = (bms) => {
  let limit = BMS_COUPON_DESCRIPTION_MAX[bms.chatBubbleType];
  if (bms.coupon && bms.coupon.description.length > limit)
    return `\uCFE0\uD3F0 \uC124\uBA85\uC740 \uCD5C\uB300 ${limit}\uC790 \uC774\uD558\uB85C \uC785\uB825\uD574\uC8FC\uC138\uC694.`;
  for (let item of bms.carousel?.list ?? [])
    if (item.coupon && item.coupon.description.length > limit)
      return `\uCFE0\uD3F0 \uC124\uBA85\uC740 \uCD5C\uB300 ${limit}\uC790 \uC774\uD558\uB85C \uC785\uB825\uD574\uC8FC\uC138\uC694.`;
  return true;
}, validateSingleButtonGroup = (chatBubbleType, buttons, withCoupon) => {
  let range = BMS_BUTTON_COUNT[chatBubbleType], max = withCoupon && range.maxWithCoupon != null ? range.maxWithCoupon : range.max;
  return buttons.length > max ? withCoupon && range.maxWithCoupon != null ? `${chatBubbleType} \uD0C0\uC785\uC5D0\uC11C\uB294 \uCFE0\uD3F0 \uC0AC\uC6A9 \uC2DC \uCD5C\uB300 ${max}\uAC1C\uC758 \uBC84\uD2BC\uB9CC \uC0AC\uC6A9\uD560 \uC218 \uC788\uC2B5\uB2C8\uB2E4.` : `${chatBubbleType} \uD0C0\uC785\uC5D0\uC11C\uB294 \uCD5C\uB300 ${max}\uAC1C\uC758 \uBC84\uD2BC\uB9CC \uC0AC\uC6A9\uD560 \uC218 \uC788\uC2B5\uB2C8\uB2E4.` : range.min != null && buttons.length < range.min ? `${chatBubbleType} \uD0C0\uC785\uC5D0\uC11C\uB294 \uCD5C\uC18C ${range.min}\uAC1C\uC758 \uBC84\uD2BC\uC774 \uD544\uC694\uD569\uB2C8\uB2E4.` : true;
}, validateButtonCount = (bms) => {
  let type = bms.chatBubbleType;
  if (type === "CAROUSEL_FEED" || type === "CAROUSEL_COMMERCE") {
    for (let item of bms.carousel?.list ?? []) {
      let result = validateSingleButtonGroup(
        type,
        item.buttons ?? [],
        hasCoupon(bms) || item.coupon != null
      );
      if (result !== true) return result;
    }
    return true;
  }
  return bms.buttons ? validateSingleButtonGroup(type, bms.buttons, hasCoupon(bms)) : true;
}, validateButtonNames = (bms) => {
  let type = bms.chatBubbleType, limit = BMS_BUTTON_NAME_MAX[type], checkList = (buttons) => {
    for (let button of buttons)
      if (button.linkType !== "AC" && button.name != null && button.name.length > limit)
        return `${type} \uD0C0\uC785 button.name\uC740 \uCD5C\uB300 ${limit}\uC790 \uC774\uD558\uB85C \uC785\uB825\uD574\uC8FC\uC138\uC694.`;
    return true;
  };
  if (bms.buttons) {
    let result = checkList(bms.buttons);
    if (result !== true) return result;
  }
  for (let item of bms.carousel?.list ?? []) {
    let result = checkList(item.buttons ?? []);
    if (result !== true) return result;
  }
  return true;
}, validateAllowedLinkTypes = (bms) => {
  let type = bms.chatBubbleType, allowed = BMS_ALLOWED_LINK_TYPES[type];
  if (!allowed) return true;
  let checkList = (buttons) => {
    for (let button of buttons)
      if (!allowed.includes(button.linkType))
        return `${type} \uD0C0\uC785\uC5D0\uC11C\uB294 ${allowed.join(", ")} \uD0C0\uC785\uC758 \uBC84\uD2BC\uB9CC \uC0AC\uC6A9\uD560 \uC218 \uC788\uC2B5\uB2C8\uB2E4.`;
    return true;
  };
  if (bms.buttons) {
    let result = checkList(bms.buttons);
    if (result !== true) return result;
  }
  for (let item of bms.carousel?.list ?? []) {
    let result = checkList(item.buttons ?? []);
    if (result !== true) return result;
  }
  return true;
}, lengthError = (chatBubbleType, fieldName, max) => `${chatBubbleType} \uD0C0\uC785 ${fieldName}\uC740 \uCD5C\uB300 ${max}\uC790 \uC774\uD558\uB85C \uC785\uB825\uD574\uC8FC\uC138\uC694.`, validateTextLengths = (bms) => {
  let type = bms.chatBubbleType;
  if (bms.header != null && bms.header.length > BMS_HEADER_MAX)
    return lengthError(type, "header", BMS_HEADER_MAX);
  let contentLimit = BMS_CONTENT_MAX[type];
  if (contentLimit != null && bms.content != null && bms.content.length > contentLimit)
    return lengthError(type, "content", contentLimit);
  if (bms.additionalContent != null && bms.additionalContent.length > BMS_ADDITIONAL_CONTENT_MAX)
    return lengthError(type, "additionalContent", BMS_ADDITIONAL_CONTENT_MAX);
  if (bms.mainWideItem?.title != null && bms.mainWideItem.title.length > BMS_MAIN_WIDE_ITEM_TITLE_MAX)
    return lengthError(
      type,
      "mainWideItem.title",
      BMS_MAIN_WIDE_ITEM_TITLE_MAX
    );
  for (let sub of bms.subWideItemList ?? [])
    if (sub.title.length > BMS_SUB_WIDE_ITEM_TITLE_MAX)
      return lengthError(
        type,
        "subWideItem.title",
        BMS_SUB_WIDE_ITEM_TITLE_MAX
      );
  if (bms.carousel?.head) {
    let { header, content } = bms.carousel.head;
    if (header != null && header.length > BMS_HEADER_MAX)
      return lengthError(type, "carousel.head.header", BMS_HEADER_MAX);
    if (content != null && content.length > BMS_CAROUSEL_HEAD_CONTENT_MAX)
      return lengthError(
        type,
        "carousel.head.content",
        BMS_CAROUSEL_HEAD_CONTENT_MAX
      );
  }
  if (type === "CAROUSEL_FEED")
    for (let item of bms.carousel?.list ?? []) {
      if (item.header != null && item.header.length > BMS_HEADER_MAX)
        return lengthError(type, "carousel.list.header", BMS_HEADER_MAX);
      if (item.content != null && item.content.length > BMS_CAROUSEL_FEED_ITEM_CONTENT_MAX)
        return lengthError(
          type,
          "carousel.list.content",
          BMS_CAROUSEL_FEED_ITEM_CONTENT_MAX
        );
    }
  if (type === "CAROUSEL_COMMERCE") {
    for (let item of bms.carousel?.list ?? [])
      if (item.additionalContent != null && item.additionalContent.length > BMS_CAROUSEL_COMMERCE_ADDITIONAL_CONTENT_MAX)
        return lengthError(
          type,
          "carousel.list.additionalContent",
          BMS_CAROUSEL_COMMERCE_ADDITIONAL_CONTENT_MAX
        );
  }
  return true;
}, validateCarouselListCount = (bms) => {
  let type = bms.chatBubbleType;
  if (type !== "CAROUSEL_FEED" && type !== "CAROUSEL_COMMERCE") return true;
  let list = bms.carousel?.list ?? [], hasHead = bms.carousel?.head != null, range = hasHead ? BMS_CAROUSEL_LIST_RANGE.withHead : BMS_CAROUSEL_LIST_RANGE.withoutHead;
  return list.length < range.min || list.length > range.max ? `${hasHead ? "\uCE90\uB7EC\uC140 \uC778\uD2B8\uB85C\uAC00 \uC788\uB294 \uACBD\uC6B0 \uCE90\uB7EC\uC140 \uB9AC\uC2A4\uD2B8\uB294" : "\uCE90\uB7EC\uC140 \uB9AC\uC2A4\uD2B8\uB294"} \uCD5C\uC18C ${range.min}\uAC1C, \uCD5C\uB300 ${range.max}\uAC1C\uAE4C\uC9C0 \uAC00\uB2A5\uD569\uB2C8\uB2E4.` : true;
}, countNewlines = (text) => {
  let matches = text.match(/\r\n|\r|\n/g);
  return matches ? matches.length : 0;
}, newlineError = (chatBubbleType, fieldName, max) => `${chatBubbleType} \uD0C0\uC785 ${fieldName}\uC740 \uC904\uBC14\uAFC8 \uCD5C\uB300 ${max}\uAC1C \uAE4C\uC9C0 \uAC00\uB2A5\uD569\uB2C8\uB2E4.`, contentNewlineLimit = (type) => type === "TEXT" || type === "IMAGE" ? 99 : type === "WIDE" || type === "PREMIUM_VIDEO" ? 1 : null, validateNewlineLimits = (bms) => {
  let type = bms.chatBubbleType;
  if (bms.header != null && countNewlines(bms.header) > 0)
    return newlineError(type, "header", 0);
  let contentNL = contentNewlineLimit(type);
  if (contentNL != null && bms.content != null && countNewlines(bms.content) > contentNL)
    return newlineError(type, "content", contentNL);
  if (bms.mainWideItem?.title != null && countNewlines(bms.mainWideItem.title) > 1)
    return newlineError(type, "mainWideItem.title", 1);
  for (let sub of bms.subWideItemList ?? [])
    if (countNewlines(sub.title) > 1)
      return newlineError(type, "subWideItem.title", 1);
  if (bms.carousel?.head) {
    let { header, content } = bms.carousel.head;
    if (header != null && countNewlines(header) > 0)
      return newlineError(type, "carousel.head.header", 0);
    if (content != null && countNewlines(content) > 2)
      return newlineError(type, "carousel.head.content", 2);
  }
  for (let item of bms.carousel?.list ?? []) {
    if (item.header != null && countNewlines(item.header) > 0)
      return newlineError(type, "carousel.list.header", 0);
    if (type === "CAROUSEL_FEED" && item.content != null && countNewlines(item.content) > 2)
      return newlineError(type, "carousel.list.content", 2);
    if (type === "CAROUSEL_COMMERCE" && item.additionalContent != null && countNewlines(item.additionalContent) > 1)
      return newlineError(type, "carousel.list.additionalContent", 1);
  }
  return true;
}, HTTP_URL_PATTERN = /^https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{1,64}\.[a-z0-9]{1,64}\b([-a-zA-Z0-9@:%~_~+.~#?&//=]*)$/i, VARIABLE_PLACEHOLDER_PATTERN = /#\{[^{}]+\}/, VARIABLE_PLACEHOLDER_GLOBAL = /#\{[^{}]+\}/g, WHITESPACE_IN_LINK_PATTERN = /\s/, isValidLink = (link) => {
  if (WHITESPACE_IN_LINK_PATTERN.test(link)) return false;
  if (HTTP_URL_PATTERN.test(link)) return true;
  if (!VARIABLE_PLACEHOLDER_PATTERN.test(link)) return false;
  let normalized = link.replace(VARIABLE_PLACEHOLDER_GLOBAL, "x");
  return HTTP_URL_PATTERN.test(normalized);
}, checkLinkPair = (obj) => obj ? obj.linkPc !== void 0 && !isValidLink(obj.linkPc) ? "linkPc \uAC12\uC774 \uC798\uBABB\uB418\uC5C8\uC2B5\uB2C8\uB2E4. \uC62C\uBC14\uB978 \uD615\uC2DD\uC740 \uC6F9 \uB9C1\uD06C \uD615\uC2DD\uC774\uC5B4\uC57C \uD569\uB2C8\uB2E4." : obj.linkMobile !== void 0 && !isValidLink(obj.linkMobile) ? "linkMobile \uAC12\uC774 \uC798\uBABB\uB418\uC5C8\uC2B5\uB2C8\uB2E4. \uC62C\uBC14\uB978 \uD615\uC2DD\uC740 \uC6F9 \uB9C1\uD06C \uD615\uC2DD\uC774\uC5B4\uC57C \uD569\uB2C8\uB2E4." : true : true, IMAGE_LINK_ERROR = "imageLink \uAC12\uC774 \uC798\uBABB\uB418\uC5C8\uC2B5\uB2C8\uB2E4. http:// \uB610\uB294 https:// \uB85C \uC2DC\uC791\uD558\uB294 \uC815\uC0C1\uC801\uC778 \uC8FC\uC18C\uB97C \uC62C\uB824\uC8FC\uC138\uC694.", checkImageLink = (imageLink) => imageLink === void 0 || isValidLink(imageLink) ? true : IMAGE_LINK_ERROR, validateLinks = (bms) => {
  let imageLinkResult = checkImageLink(bms.imageLink);
  if (imageLinkResult !== true) return imageLinkResult;
  let couponResult = checkLinkPair(bms.coupon);
  if (couponResult !== true) return couponResult;
  for (let button of bms.buttons ?? []) {
    let r = checkLinkPair(button);
    if (r !== true) return r;
  }
  let mainResult = checkLinkPair(bms.mainWideItem);
  if (mainResult !== true) return mainResult;
  for (let sub of bms.subWideItemList ?? []) {
    let r = checkLinkPair(sub);
    if (r !== true) return r;
  }
  let headResult = checkLinkPair(bms.carousel?.head);
  if (headResult !== true) return headResult;
  let tailResult = checkLinkPair(bms.carousel?.tail);
  if (tailResult !== true) return tailResult;
  for (let item of bms.carousel?.list ?? []) {
    let imageResult = checkImageLink(item.imageLink);
    if (imageResult !== true) return imageResult;
    for (let button of item.buttons ?? []) {
      let r = checkLinkPair(button);
      if (r !== true) return r;
    }
    let itemCouponResult = checkLinkPair(item.coupon);
    if (itemCouponResult !== true) return itemCouponResult;
  }
  return true;
}, FORBIDDEN_VARIABLE_PATTERN = /#\{((?!(#{|})).)+\}/, validateForbiddenVariables = (bms) => {
  let checkButtons = (buttons) => {
    for (let button of buttons)
      if (button.name && FORBIDDEN_VARIABLE_PATTERN.test(button.name))
        return "button.name\uC5D0\uB294 \uBCC0\uC218\uB97C \uC0AC\uC6A9\uD560 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4.";
    return true;
  };
  if (bms.buttons) {
    let r = checkButtons(bms.buttons);
    if (r !== true) return r;
  }
  for (let item of bms.carousel?.list ?? []) {
    let r = checkButtons(item.buttons ?? []);
    if (r !== true) return r;
  }
  let tail = bms.carousel?.tail;
  if (tail) {
    let linkKeys = [
      "linkMobile",
      "linkPc",
      "linkAndroid",
      "linkIos"
    ];
    for (let key of linkKeys) {
      let value = tail[key];
      if (value && FORBIDDEN_VARIABLE_PATTERN.test(value))
        return `${key}\uC5D0\uB294 \uBCC0\uC218\uB97C \uC0AC\uC6A9\uD560 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4.`;
    }
  }
  return true;
}, BMS_ACCEPTABLE_FIELDS = {
  TEXT: ["adult", "content", "buttons", "coupon"],
  IMAGE: ["adult", "content", "imageId", "imageLink", "buttons", "coupon"],
  WIDE: ["adult", "content", "imageId", "buttons", "coupon"],
  WIDE_ITEM_LIST: [
    "adult",
    "header",
    "mainWideItem",
    "subWideItemList",
    "buttons",
    "coupon"
  ],
  COMMERCE: [
    "adult",
    "additionalContent",
    "imageId",
    "commerce",
    "buttons",
    "coupon"
  ],
  CAROUSEL_FEED: ["adult", "carousel"],
  CAROUSEL_COMMERCE: ["adult", "additionalContent", "carousel"],
  PREMIUM_VIDEO: ["adult", "header", "content", "video", "buttons", "coupon"]
}, RESERVED_FIELDS = ["targeting", "chatBubbleType"], validateAcceptableFields = (bms) => {
  let type = bms.chatBubbleType, acceptable = BMS_ACCEPTABLE_FIELDS[type], allowed = /* @__PURE__ */ new Set([...acceptable, ...RESERVED_FIELDS]);
  for (let key of Object.keys(bms))
    if (getField(bms, key) != null && !allowed.has(key))
      return `${type}\uD0C0\uC785 \uC5D0\uC11C\uB294 ${acceptable.join(", ")} \uAC12\uB9CC \uC0AC\uC6A9\uC774 \uAC00\uB2A5\uD569\uB2C8\uB2E4.`;
  return true;
}, getField = (obj, key) => obj[key], BMS_IMAGE_ID_MAX = 32, validateImageIdLength = (bms) => bms.chatBubbleType !== "IMAGE" || !bms.imageId ? true : bms.imageId.replace(/\s/g, "").length > BMS_IMAGE_ID_MAX ? `IMAGE \uD0C0\uC785 imageId\uC740 \uCD5C\uB300 ${BMS_IMAGE_ID_MAX}\uC790 \uC774\uD558\uB85C \uC785\uB825\uD574\uC8FC\uC138\uC694.` : true;
var KAKAO_TV_URL_PREFIX = "https://tv.kakao.com/", isKakaoTvUrl = (url) => url.startsWith(KAKAO_TV_URL_PREFIX), bmsVideoSchema = Schema.Struct({
  videoUrl: Schema.String.pipe(
    Schema.filter(isKakaoTvUrl, {
      message: () => `videoUrl\uC740 '${KAKAO_TV_URL_PREFIX}'\uC73C\uB85C \uC2DC\uC791\uD558\uB294 \uCE74\uCE74\uC624TV \uB3D9\uC601\uC0C1 \uB9C1\uD06C\uC5EC\uC57C \uD569\uB2C8\uB2E4.`
    })
  ),
  imageId: Schema.optional(Schema.String),
  imageLink: Schema.optional(Schema.String)
});
var bmsMainWideItemSchema = Schema.Struct({
  title: Schema.optional(Schema.String),
  imageId: Schema.String,
  linkMobile: Schema.String,
  linkPc: Schema.optional(Schema.String),
  linkAndroid: Schema.optional(Schema.String),
  linkIos: Schema.optional(Schema.String)
}), bmsSubWideItemSchema = Schema.Struct({
  title: Schema.String,
  imageId: Schema.String,
  linkMobile: Schema.String,
  linkPc: Schema.optional(Schema.String),
  linkAndroid: Schema.optional(Schema.String),
  linkIos: Schema.optional(Schema.String)
});
var kakaoWebButtonSchema = Schema.Struct({
  buttonName: Schema.String,
  buttonType: Schema.Literal("WL"),
  linkMo: Schema.String,
  linkPc: Schema.optional(Schema.NullOr(Schema.String))
}), kakaoAppButtonSchema = Schema.Struct({
  buttonName: Schema.String,
  buttonType: Schema.Literal("AL"),
  linkAnd: Schema.String,
  linkIos: Schema.String
}), kakaoDefaultButtonSchema = Schema.Struct({
  buttonName: Schema.String,
  buttonType: Schema.Literal("BK", "MD", "DS", "BC", "BT", "AC")
}), kakaoButtonSchema = Schema.Union(
  kakaoWebButtonSchema,
  kakaoAppButtonSchema,
  kakaoDefaultButtonSchema
);

// src/models/base/kakao/kakaoOption.ts
var VariableValidationError = class extends Data.TaggedError(
  "VariableValidationError"
) {
  get message() {
    return `\uBCC0\uC218\uBA85 ${this.invalidVariables.map((v) => `\`${v}\``).join(", ")}\uC5D0 \uC810(.)\uC744 \uD3EC\uD568\uD560 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4. \uC5B8\uB354\uC2A4\uCF54\uC5B4(_)\uB098 \uB2E4\uB978 \uBB38\uC790\uB97C \uC0AC\uC6A9\uD574\uC8FC\uC138\uC694.`;
  }
  toString() {
    return `VariableValidationError: ${this.message}`;
  }
}, BMS_REQUIRED_FIELDS = {
  TEXT: [],
  IMAGE: ["imageId"],
  WIDE: ["imageId"],
  WIDE_ITEM_LIST: ["header", "mainWideItem", "subWideItemList"],
  COMMERCE: ["imageId", "commerce", "buttons"],
  CAROUSEL_FEED: ["carousel"],
  CAROUSEL_COMMERCE: ["carousel"],
  PREMIUM_VIDEO: ["video"]
}, bmsCarouselSchema = Schema.Union(
  bmsCarouselFeedSchema,
  bmsCarouselCommerceSchema
), baseBmsSchema = Schema.Struct({
  // 필수 필드
  targeting: Schema.Literal("I", "M", "N"),
  chatBubbleType: bmsChatBubbleTypeSchema,
  // 선택 필드
  adult: Schema.optional(Schema.Boolean),
  header: Schema.optional(Schema.String),
  imageId: Schema.optional(Schema.String),
  imageLink: Schema.optional(Schema.String),
  additionalContent: Schema.optional(Schema.String),
  content: Schema.optional(Schema.String),
  // 복합 타입 필드
  carousel: Schema.optional(bmsCarouselSchema),
  mainWideItem: Schema.optional(bmsMainWideItemSchema),
  subWideItemList: Schema.optional(Schema.Array(bmsSubWideItemSchema)),
  buttons: Schema.optional(Schema.Array(bmsButtonSchema)),
  coupon: Schema.optional(bmsCouponSchema),
  commerce: Schema.optional(bmsCommerceSchema),
  video: Schema.optional(bmsVideoSchema)
}), WIDE_ITEM_LIST_MIN_SUB_ITEMS = 3, validateBmsRequiredFields = (bms) => {
  let chatBubbleType = bms.chatBubbleType, missingFields = (BMS_REQUIRED_FIELDS[chatBubbleType] ?? []).filter(
    (field) => bms[field] === void 0 || bms[field] === null
  );
  if (missingFields.length > 0)
    return `BMS ${chatBubbleType} \uD0C0\uC785\uC5D0 \uD544\uC218 \uD544\uB4DC\uAC00 \uB204\uB77D\uB418\uC5C8\uC2B5\uB2C8\uB2E4: ${missingFields.join(", ")}`;
  if (chatBubbleType === "WIDE_ITEM_LIST") {
    let subWideItemList = bms.subWideItemList;
    if (!subWideItemList || subWideItemList.length < WIDE_ITEM_LIST_MIN_SUB_ITEMS)
      return `WIDE_ITEM_LIST \uD0C0\uC785\uC758 subWideItemList\uB294 \uCD5C\uC18C ${WIDE_ITEM_LIST_MIN_SUB_ITEMS}\uAC1C \uC774\uC0C1\uC774\uC5B4\uC57C \uD569\uB2C8\uB2E4. \uD604\uC7AC: ${subWideItemList?.length ?? 0}\uAC1C`;
  }
  return true;
}, validateBmsConstraints = (bms) => {
  let input = bms, validators = [
    validateAcceptableFields,
    validateCarouselListCount,
    validateCouponDescription,
    validateAllowedLinkTypes,
    validateButtonCount,
    validateButtonNames,
    validateImageIdLength,
    validateTextLengths,
    validateNewlineLimits,
    validateForbiddenVariables,
    validateLinks
  ], requiredResult = validateBmsRequiredFields(bms);
  if (requiredResult !== true) return requiredResult;
  for (let validator of validators) {
    let result = validator(input);
    if (result !== true) return result;
  }
  return true;
}, kakaoOptionBmsSchema = baseBmsSchema.pipe(
  Schema.filter(validateBmsConstraints)
), VARIABLE_KEY_PATTERN = /^#\{.+}$/, DOT_PATTERN = /\./, extractVariableName = (key) => VARIABLE_KEY_PATTERN.test(key) ? key.slice(2, -1) : key, formatVariableKey = (key) => VARIABLE_KEY_PATTERN.test(key) ? key : `#{${key}}`, validateVariableNames = (variables) => pipe(
  Object.keys(variables),
  Array$1.map(extractVariableName),
  Array$1.filter((variableName) => DOT_PATTERN.test(variableName)),
  (invalidVariables) => invalidVariables.length > 0 ? Effect.fail(new VariableValidationError({ invalidVariables })) : Effect.succeed(variables)
), transformVariables = (variables) => pipe(
  validateVariableNames(variables),
  Effect.map(
    (validVariables) => pipe(
      Object.entries(validVariables),
      Array$1.map(
        ([key, value]) => [formatVariableKey(key), value]
      ),
      (entries) => Object.fromEntries(entries)
    )
  )
), baseKakaoOptionSchema = Schema.Struct({
  pfId: Schema.String,
  templateId: Schema.optional(Schema.String),
  variables: Schema.optional(
    Schema.Record({ key: Schema.String, value: Schema.String }).pipe(
      Schema.transformOrFail(
        Schema.Record({ key: Schema.String, value: Schema.String }),
        {
          decode: (fromU, _, ast) => transformVariables(fromU).pipe(
            Effect.mapError(
              (err) => new ParseResult.Type(ast, fromU, err.message)
            )
          ),
          encode: (toI) => ParseResult.succeed(toI)
        }
      )
    )
  ),
  disableSms: Schema.optional(Schema.Boolean),
  adFlag: Schema.optional(Schema.Boolean),
  imageId: Schema.optional(Schema.String),
  buttons: Schema.optional(Schema.Array(kakaoButtonSchema)),
  bms: Schema.optional(kakaoOptionBmsSchema)
});
var naverOptionButtonSchema = Schema.Struct({
  buttonName: Schema.String,
  buttonType: Schema.String,
  linkMo: Schema.optional(Schema.String),
  linkPc: Schema.optional(Schema.String),
  linkAnd: Schema.optional(Schema.String),
  linkIos: Schema.optional(Schema.String)
}), naverOptionSchema = Schema.Struct({
  talkId: Schema.String,
  templateId: Schema.String,
  disableSms: Schema.optional(Schema.Boolean),
  variables: Schema.optional(
    Schema.Record({ key: Schema.String, value: Schema.String })
  ),
  buttons: Schema.optional(Schema.Array(naverOptionButtonSchema))
});
Schema.Literal(
  "WL",
  "ML",
  "MQ",
  "MR",
  "CA",
  "CL",
  "DL",
  "MS"
); var rcsWebButtonSchema = Schema.Struct({
  buttonName: Schema.String,
  buttonType: Schema.Literal("WL"),
  link: Schema.String
}), rcsMapButtonSchema = Schema.Struct({
  buttonName: Schema.String,
  buttonType: Schema.Literal("ML"),
  latitude: Schema.String,
  longitude: Schema.String
}), rcsDefaultButtonSchema = Schema.Struct({
  buttonName: Schema.String,
  buttonType: Schema.Literal("MQ", "MR", "CA", "CL", "DL", "MS"),
  link: Schema.String
}), rcsButtonSchema = Schema.Union(
  rcsWebButtonSchema,
  rcsMapButtonSchema,
  rcsDefaultButtonSchema
);

// src/models/base/rcs/rcsOption.ts
var additionalBodySchema = Schema.Struct({
  title: Schema.String,
  description: Schema.String,
  imaggeId: Schema.optional(Schema.String),
  buttons: Schema.optional(Schema.Array(rcsButtonSchema))
}), rcsOptionRequestSchema = Schema.Struct({
  brandId: Schema.String,
  templateId: Schema.optional(Schema.String),
  copyAllowed: Schema.optional(Schema.Boolean),
  variables: Schema.optional(
    Schema.Record({ key: Schema.String, value: Schema.String })
  ),
  mmsType: Schema.optional(
    Schema.Literal("M3", "S3", "M4", "S4", "M5", "S5", "M6", "S6")
  ),
  commercialType: Schema.optional(Schema.Boolean),
  disableSms: Schema.optional(Schema.Boolean),
  additionalBody: Schema.optional(additionalBodySchema),
  buttons: Schema.optional(Schema.Array(rcsButtonSchema))
}), rcsOptionSchema = rcsOptionRequestSchema;
var voiceOptionSchema = Schema.Struct({
  voiceType: Schema.Literal("FEMALE", "MALE"),
  headerMessage: Schema.optional(Schema.String),
  tailMessage: Schema.optional(Schema.String),
  replyRange: Schema.optional(Schema.Literal(1, 2, 3, 4, 5, 6, 7, 8, 9)),
  counselorNumber: Schema.optional(Schema.String)
});
var messageTypeSchema = Schema.Literal(
  "SMS",
  "LMS",
  "MMS",
  "ATA",
  "CTA",
  "CTI",
  "NSA",
  "RCS_SMS",
  "RCS_LMS",
  "RCS_MMS",
  "RCS_TPL",
  "RCS_ITPL",
  "RCS_LTPL",
  "FAX",
  "VOICE",
  "BMS_TEXT",
  "BMS_IMAGE",
  "BMS_WIDE",
  "BMS_WIDE_ITEM_LIST",
  "BMS_CAROUSEL_FEED",
  "BMS_PREMIUM_VIDEO",
  "BMS_COMMERCE",
  "BMS_CAROUSEL_COMMERCE",
  "BMS_FREE"
), messageSchema = Schema.Struct({
  to: Schema.Union(Schema.String, Schema.Array(Schema.String)),
  from: Schema.optional(Schema.String),
  text: Schema.optional(Schema.String),
  imageId: Schema.optional(Schema.String),
  type: Schema.optional(messageTypeSchema),
  subject: Schema.optional(Schema.String),
  autoTypeDetect: Schema.optional(Schema.Boolean),
  kakaoOptions: Schema.optional(baseKakaoOptionSchema),
  rcsOptions: Schema.optional(rcsOptionSchema),
  country: Schema.optional(Schema.String),
  replacements: Schema.optional(Schema.Array(Schema.Struct({}))),
  customFields: Schema.optional(
    Schema.Record({ key: Schema.String, value: Schema.String })
  ),
  naverOptions: Schema.optional(naverOptionSchema),
  faxOptions: Schema.optional(
    Schema.Struct({ fileIds: Schema.Array(Schema.String) })
  ),
  voiceOptions: Schema.optional(voiceOptionSchema)
});

// src/models/base/messages/storedMessage.ts
var booleanOrZeroOne = Schema.transformOrFail(
  Schema.Union(Schema.Boolean, Schema.Number),
  Schema.Boolean,
  {
    decode: (value, _opts, ast) => typeof value == "boolean" ? ParseResult.succeed(value) : value === 0 ? ParseResult.succeed(false) : value === 1 ? ParseResult.succeed(true) : ParseResult.fail(
      new ParseResult.Type(
        ast,
        value,
        `Expected boolean, 0, or 1 but received ${String(value)}`
      )
    ),
    encode: (value) => ParseResult.succeed(value),
    strict: true
  }
), storedMessageSchema = Schema.Struct({
  messageId: Schema.optional(Schema.String),
  type: Schema.NullishOr(messageTypeSchema),
  to: Schema.optional(Schema.Union(Schema.String, Schema.Array(Schema.String))),
  from: Schema.NullishOr(Schema.String),
  text: Schema.NullishOr(Schema.String),
  imageId: Schema.NullishOr(Schema.String),
  subject: Schema.NullishOr(Schema.String),
  country: Schema.NullishOr(Schema.String),
  accountId: Schema.optional(Schema.String),
  groupId: Schema.optional(Schema.String),
  status: Schema.NullishOr(Schema.String),
  statusCode: Schema.NullishOr(Schema.String),
  reason: Schema.NullishOr(Schema.String),
  networkName: Schema.NullishOr(Schema.String),
  networkCode: Schema.NullishOr(Schema.String),
  customFields: Schema.optional(
    Schema.NullishOr(Schema.Record({ key: Schema.String, value: Schema.String }))
  ),
  autoTypeDetect: Schema.optional(booleanOrZeroOne),
  replacement: Schema.optional(booleanOrZeroOne),
  resendCount: Schema.optional(Schema.Number),
  dateCreated: Schema.optional(Schema.String),
  dateUpdated: Schema.optional(Schema.String),
  dateProcessed: Schema.NullishOr(Schema.String),
  dateReceived: Schema.NullishOr(Schema.String),
  dateReported: Schema.NullishOr(Schema.String),
  // 옵션 객체는 서버 정규화 포맷(저장 형태)으로 발송 요청용 스키마와 필드가 다르다.
  // 상세 타이핑을 확정하려면 각 옵션별 별도 조회 스키마 정의가 필요하지만 본 PR 범위를
  // 벗어나므로, 최소한 "object"임을 보장해 원시 값이 섞이는 drift를 감지할 수 있게 한다.
  kakaoOptions: Schema.optional(
    Schema.NullishOr(
      Schema.Record({ key: Schema.String, value: Schema.Unknown })
    )
  ),
  rcsOptions: Schema.optional(
    Schema.NullishOr(
      Schema.Record({ key: Schema.String, value: Schema.Unknown })
    )
  ),
  naverOptions: Schema.optional(
    Schema.NullishOr(
      Schema.Record({ key: Schema.String, value: Schema.Unknown })
    )
  ),
  faxOptions: Schema.optional(
    Schema.NullishOr(
      Schema.Record({ key: Schema.String, value: Schema.Unknown })
    )
  ),
  voiceOptions: Schema.optional(
    Schema.NullishOr(
      Schema.Record({ key: Schema.String, value: Schema.Unknown })
    )
  ),
  replacements: Schema.optional(Schema.NullishOr(Schema.Array(Schema.Unknown))),
  log: Schema.optional(Schema.NullishOr(Schema.Array(Schema.Unknown))),
  queues: Schema.optional(Schema.NullishOr(Schema.Array(Schema.Unknown))),
  currentQueue: Schema.optional(Schema.NullishOr(Schema.Unknown)),
  clusterKey: Schema.NullishOr(Schema.String),
  unavailableSenderNumber: Schema.optional(Schema.NullishOr(booleanOrZeroOne)),
  faxPageCount: Schema.optional(Schema.NullishOr(Schema.Number)),
  voiceDuration: Schema.optional(Schema.NullishOr(Schema.Number)),
  voiceReplied: Schema.optional(Schema.NullishOr(booleanOrZeroOne)),
  _id: Schema.optional(Schema.String)
});

// src/models/responses/messageResponses.ts
var groupMessageResponseSchema = Schema.Struct({
  count: countSchema,
  countForCharge: countForChargeSchema,
  balance: commonCashResponseSchema,
  point: commonCashResponseSchema,
  app: appSchema,
  log: logSchema,
  status: Schema.String,
  allowDuplicates: Schema.Boolean,
  isRefunded: Schema.Boolean,
  accountId: Schema.String,
  masterAccountId: Schema.NullOr(Schema.String),
  apiVersion: Schema.String,
  groupId: Schema.String,
  price: Schema.Record({ key: Schema.String, value: Schema.Unknown }),
  dateCreated: Schema.String,
  dateUpdated: Schema.String,
  scheduledDate: Schema.NullishOr(Schema.String),
  dateSent: Schema.NullishOr(Schema.String),
  dateCompleted: Schema.NullishOr(Schema.String)
}), addMessageResultSchema = Schema.Struct({
  to: Schema.String,
  from: Schema.String,
  type: Schema.String,
  country: Schema.String,
  messageId: Schema.String,
  statusCode: Schema.String,
  statusMessage: Schema.String,
  accountId: Schema.String,
  customFields: Schema.optional(
    Schema.Record({ key: Schema.String, value: Schema.String })
  )
}), addMessageResponseSchema = Schema.Struct({
  errorCount: Schema.String,
  resultList: Schema.Array(addMessageResultSchema)
}), getMessagesResponseSchema = Schema.Struct({
  startKey: Schema.optional(Schema.NullOr(Schema.String)),
  nextKey: Schema.optional(Schema.NullOr(Schema.String)),
  limit: Schema.Number,
  messageList: Schema.Record({ key: Schema.String, value: storedMessageSchema })
}), removeGroupMessagesResponseSchema = Schema.Struct({
  groupId: groupIdSchema,
  errorCount: Schema.Number,
  resultList: Schema.Array(
    Schema.Struct({
      messageId: Schema.String,
      resultCode: Schema.String
    })
  )
}), getGroupsResponseSchema = Schema.Struct({
  startKey: Schema.NullishOr(Schema.String),
  limit: Schema.Number,
  nextKey: Schema.NullishOr(Schema.String),
  groupList: Schema.Record({ key: groupIdSchema, value: groupSchema })
}), statisticsPeriodResultSchema = Schema.Struct({
  total: Schema.Number,
  sms: Schema.Number,
  lms: Schema.Number,
  mms: Schema.Number,
  ata: Schema.Number,
  cta: Schema.Number,
  cti: Schema.Number,
  nsa: Schema.Number,
  rcs_sms: Schema.Number,
  rcs_lms: Schema.Number,
  rcs_mms: Schema.Number,
  rcs_tpl: Schema.Number,
  rcs_itpl: Schema.optional(Schema.Number),
  rcs_ltpl: Schema.optional(Schema.Number),
  fax: Schema.optional(Schema.Number),
  voice: Schema.optional(Schema.Number),
  bms_text: Schema.optional(Schema.Number),
  bms_image: Schema.optional(Schema.Number),
  bms_wide: Schema.optional(Schema.Number),
  bms_wide_item_list: Schema.optional(Schema.Number),
  bms_carousel_feed: Schema.optional(Schema.Number),
  bms_premium_video: Schema.optional(Schema.Number),
  bms_commerce: Schema.optional(Schema.Number),
  bms_carousel_commerce: Schema.optional(Schema.Number),
  bms_free: Schema.optional(Schema.Number)
}), refundSchema = Schema.Struct({
  balance: Schema.Number,
  point: Schema.Number,
  deposit: Schema.optional(Schema.Number)
}), dayPeriodSchema = Schema.Struct({
  _id: Schema.String,
  month: Schema.String,
  date: Schema.optional(Schema.String),
  balance: Schema.Number,
  point: Schema.Number,
  deposit: Schema.optional(Schema.Number),
  statusCode: Schema.Record({
    key: Schema.String,
    value: partialMessageTypeRecordSchema
  }),
  refund: refundSchema,
  total: statisticsPeriodResultSchema,
  successed: statisticsPeriodResultSchema,
  failed: statisticsPeriodResultSchema
}), monthPeriodRefundSchema = Schema.Struct({
  balance: Schema.Number,
  balanceAvg: Schema.Number,
  point: Schema.Number,
  pointAvg: Schema.Number,
  deposit: Schema.optional(Schema.Number),
  depositAvg: Schema.optional(Schema.Number)
}), monthPeriodSchema = Schema.Struct({
  date: Schema.String,
  balance: Schema.Number,
  balanceAvg: Schema.Number,
  point: Schema.Number,
  pointAvg: Schema.Number,
  deposit: Schema.optional(Schema.Number),
  depositAvg: Schema.optional(Schema.Number),
  dayPeriod: Schema.Array(dayPeriodSchema),
  refund: Schema.optional(monthPeriodRefundSchema),
  total: statisticsPeriodResultSchema,
  successed: statisticsPeriodResultSchema,
  failed: statisticsPeriodResultSchema
}), getStatisticsResponseSchema = Schema.Struct({
  balance: Schema.Number,
  point: Schema.Number,
  deposit: Schema.optional(Schema.Number),
  monthlyBalanceAvg: Schema.Number,
  monthlyPointAvg: Schema.Number,
  monthlyDepositAvg: Schema.optional(Schema.Number),
  monthPeriod: Schema.Array(monthPeriodSchema),
  total: statisticsPeriodResultSchema,
  successed: statisticsPeriodResultSchema,
  failed: statisticsPeriodResultSchema,
  dailyBalanceAvg: Schema.optional(Schema.Number),
  dailyPointAvg: Schema.optional(Schema.Number),
  dailyTotalCountAvg: Schema.optional(Schema.Number),
  dailyFailedCountAvg: Schema.optional(Schema.Number),
  dailySuccessedCountAvg: Schema.optional(Schema.Number)
}), lowBalanceAlertSchema = Schema.Struct({
  notificationBalance: Schema.String,
  currentBalance: Schema.String,
  balances: Schema.Array(Schema.Number),
  channels: Schema.Array(Schema.String),
  enabled: Schema.Boolean
}), getBalanceResponseSchema = Schema.Struct({
  lowBalanceAlert: Schema.optional(lowBalanceAlertSchema),
  point: Schema.Number,
  minimumCash: Schema.optional(Schema.Number),
  rechargeTo: Schema.optional(Schema.Number),
  rechargeTryCount: Schema.optional(Schema.Number),
  autoRecharge: Schema.optional(Schema.Number),
  accountId: Schema.optional(Schema.String),
  balance: Schema.Number,
  deposit: Schema.optional(Schema.Number),
  balanceOnly: Schema.optional(Schema.Number)
}), fileUploadResponseSchema = Schema.Struct({
  fileId: Schema.String,
  type: Schema.String,
  link: Schema.NullishOr(Schema.String)
}), requestKakaoChannelTokenResponseSchema = Schema.Struct({
  success: Schema.Boolean
}), createKakaoChannelResponseSchema = Schema.Struct({
  accountId: Schema.String,
  phoneNumber: Schema.String,
  searchId: Schema.String,
  dateCreated: Schema.String,
  dateUpdated: Schema.String,
  channelId: Schema.String
});
function genCustomText(alphabet, size) {
  let bytes = randomBytes(size), genRandomText = "";
  for (let i = 0; i < size; i++)
    genRandomText += alphabet[bytes[i] % alphabet.length];
  return genRandomText;
}
function getAuthInfo(authenticationParameter, authType = 0 /* API_KEY */) {
  let { apiKey, apiSecret } = authenticationParameter;
  return !apiKey || !apiSecret ? Effect.fail(
    new ApiKeyError({ message: "API Key\uC640 API Secret\uC740 \uD544\uC218\uC785\uB2C8\uB2E4." })
  ) : Effect.try({
    try: () => {
      let salt = genCustomText(
        "1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ",
        32
      ), date = formatISO(/* @__PURE__ */ new Date()), hmacData = date + salt, genHmac = createHmac("sha256", apiSecret);
      genHmac.update(hmacData);
      let signature = genHmac.digest("hex");
      return `HMAC-SHA256 apiKey=${apiKey}, date=${date}, salt=${salt}, signature=${signature}`;
    },
    catch: (e) => new DefaultError({
      errorCode: "AuthError",
      errorMessage: e instanceof Error ? e.message : String(e)
    })
  });
}

// src/lib/defaultFetcher.ts
var ERROR_MESSAGE_PREVIEW_LENGTH = 200, RETRYABLE_ERROR_KEYWORDS = [
  "aborted",
  "refused",
  "reset",
  "econn"
], RetryableError = class extends Data.TaggedError("RetryableError") {
}, toMessage = (e) => e instanceof Error ? e.message : String(e), isRetryableNetworkError = (error) => {
  let cause = error.cause, causeCode = cause && typeof cause == "object" && "code" in cause ? String(cause.code) : "", message = `${error.message} ${causeCode}`.toLowerCase();
  return RETRYABLE_ERROR_KEYWORDS.some((keyword) => message.includes(keyword));
}, makeParseError = (res, message) => new DefaultError({
  errorCode: "ParseError",
  errorMessage: message,
  context: { responseStatus: res.status, responseUrl: res.url }
}), handleOkResponse = (res) => pipe(
  Effect.tryPromise({
    try: () => res.text(),
    catch: (e) => makeParseError(res, toMessage(e))
  }),
  Effect.flatMap((responseText) => responseText ? Effect.try({
    try: () => JSON.parse(responseText),
    catch: (e) => makeParseError(res, toMessage(e))
  }) : res.status === 204 ? Effect.succeed({}) : Effect.fail(
    makeParseError(res, "API returned empty response body")
  ))
), handleClientErrorResponse = (res) => pipe(
  Effect.tryPromise({
    try: () => res.text(),
    catch: (e) => makeParseError(res, toMessage(e))
  }),
  Effect.flatMap((text) => {
    let genericError = new ClientError({
      errorCode: `HTTP_${res.status}`,
      errorMessage: text.substring(0, ERROR_MESSAGE_PREVIEW_LENGTH) || "Client error occurred",
      httpStatus: res.status,
      url: res.url
    });
    return Effect.flatMap(
      Effect.try({
        try: () => JSON.parse(text),
        catch: (e) => e instanceof SyntaxError ? genericError : new ClientError({
          errorCode: "ResponseParseError",
          errorMessage: toMessage(e),
          httpStatus: res.status,
          url: res.url
        })
      }),
      (json) => Effect.fail(
        isErrorResponse(json) ? new ClientError({
          errorCode: json.errorCode,
          errorMessage: json.errorMessage,
          httpStatus: res.status,
          url: res.url
        }) : genericError
      )
    );
  })
);
function parseServerErrorBody(text, genericError, makeError) {
  return Effect.flatMap(
    Effect.try({
      try: () => JSON.parse(text),
      catch: (e) => e instanceof SyntaxError ? genericError : makeError("ResponseParseError", toMessage(e))
    }),
    (json) => Effect.fail(
      isErrorResponse(json) ? makeError(json.errorCode, json.errorMessage) : genericError
    )
  );
}
var handleServerErrorResponse = (res) => pipe(
  Effect.tryPromise({
    try: () => res.text(),
    catch: (e) => new DefaultError({
      errorCode: "ResponseReadError",
      errorMessage: toMessage(e),
      context: { responseStatus: res.status, responseUrl: res.url }
    })
  }),
  Effect.flatMap((text) => {
    let makeError = (errorCode, errorMessage) => new ServerError({
      errorCode,
      errorMessage,
      httpStatus: res.status,
      url: res.url,
      responseBody: void 0
    }), genericError = makeError(
      `HTTP_${res.status}`,
      text.substring(0, ERROR_MESSAGE_PREVIEW_LENGTH) || "Server error occurred"
    );
    return parseServerErrorBody(text, genericError, makeError);
  })
);
function defaultFetcherEffect(authParameter, request, data) {
  let effect = Effect.gen(function* () {
    let authorizationHeaderData = yield* getAuthInfo(authParameter), body = yield* Effect.try({
      try: () => data ? JSON.stringify(data) : void 0,
      catch: (e) => new DefaultError({
        errorCode: "JSONStringifyError",
        errorMessage: toMessage(e),
        context: { data }
      })
    }), response = yield* Effect.tryPromise({
      try: () => fetch(request.url, {
        headers: {
          Authorization: authorizationHeaderData,
          "Content-Type": "application/json"
        },
        body,
        method: request.method
      }),
      catch: (error) => error instanceof Error ? isRetryableNetworkError(error) ? new RetryableError({ error }) : new NetworkError({
        url: request.url,
        method: request.method,
        cause: error.message,
        isRetryable: false
      }) : new NetworkError({
        url: request.url,
        method: request.method,
        cause: String(error),
        isRetryable: false
      })
    });
    return yield* pipe(
      Match.value(response),
      Match.when(
        (res) => res.status === 503,
        () => Effect.fail(new RetryableError({ error: "Service Unavailable" }))
      ),
      Match.when(
        (res) => res.status >= 400 && res.status < 500,
        handleClientErrorResponse
      ),
      Match.when((res) => !res.ok, handleServerErrorResponse),
      Match.orElse(handleOkResponse)
    );
  }), retryCount = 3, policy = pipe(
    Schedule.recurs(retryCount),
    Schedule.whileInput(
      (e) => e instanceof RetryableError
    )
  );
  return pipe(
    effect,
    Effect.retry(policy),
    Effect.catchTag(
      "RetryableError",
      () => Effect.fail(
        new DefaultError({
          errorCode: "RequestFailedAfterRetryError",
          errorMessage: `Request failed after retry(count: ${retryCount})`,
          context: {
            url: request.url,
            method: request.method,
            retryCount
          }
        })
      )
    )
  );
}
function formatWithTransfer(value) {
  return formatISO(stringDateTransfer(value));
}
function stringDateTransfer(value) {
  if (typeof value == "string") {
    let originalString = value;
    value = parseISO(originalString);
    let invalidDateText = "Invalid Date";
    if (value.toString() === invalidDateText)
      throw new InvalidDateError({
        message: invalidDateText,
        originalValue: originalString
      });
  }
  return value;
}

// src/lib/schemaUtils.ts
var decodeWithBadRequest = (schema, data) => Effect10.mapError(
  Schema.decodeUnknown(schema)(data),
  (error) => new BadRequestError({
    message: ParseResult.TreeFormatter.formatErrorSync(error),
    validationErrors: ParseResult.ArrayFormatter.formatErrorSync(error).map(
      (issue) => `${issue.path.length > 0 ? issue.path.join(".") : "(root)"}: ${issue.message}`
    )
  })
), safeDateTransfer = (value) => value != null ? Effect10.try({
  try: () => stringDateTransfer(value),
  catch: (error) => error instanceof InvalidDateError ? error : new InvalidDateError({
    message: error instanceof Error ? error.message : String(error)
  })
}) : Effect10.succeed(void 0), safeFormatWithTransfer = (value) => Effect10.try({
  try: () => formatWithTransfer(value),
  catch: (error) => error instanceof InvalidDateError ? error : new InvalidDateError({
    message: error instanceof Error ? error.message : String(error)
  })
}), safeFinalize = (fn) => Effect10.try({
  try: fn,
  catch: (error) => error instanceof InvalidDateError ? error : new BadRequestError({
    message: error instanceof Error ? error.message : String(error)
  })
}), stringifyResponseBody = (data) => {
  if (data !== void 0) {
    if (typeof data == "string") return data;
    try {
      return JSON.stringify(data);
    } catch (err) {
      return `[unserializable: ${err instanceof Error ? err.message : String(err)}] ${Object.prototype.toString.call(data)}`;
    }
  }
}, redactUrlForProduction = (url) => {
  if (!url) return url;
  try {
    let parsed = new URL(url), hadQuery = parsed.search.length > 0;
    return parsed.search = hadQuery ? "?[redacted]" : "", parsed.hash = "", parsed.username = "", parsed.password = "", parsed.toString();
  } catch {
    let cut = url.search(/[?#;]/);
    return cut === -1 ? url : `${url.slice(0, cut)}?[redacted]`;
  }
}, shouldRedactSensitive = () => {
  let env = "production"?.trim().toLowerCase();
  return env !== "development" && env !== "test";
}, decodeServerResponse = (schema, data, context) => (
  // onExcessProperty: 'preserve' — 서버가 추가로 내려준 미선언 필드를 strip 하지 않는다.
  // 부분 스키마로 검증하는 조회 엔드포인트에서 필드 조용히 사라지는 silent data loss를 방지.
  Effect10.mapError(
    Schema.decodeUnknown(schema, { onExcessProperty: "preserve" })(data),
    (err) => {
      let redact = shouldRedactSensitive(), issues = ParseResult.ArrayFormatter.formatErrorSync(err);
      return new ResponseSchemaMismatchError({
        message: redact ? `Response schema mismatch on ${issues.length} field(s)` : ParseResult.TreeFormatter.formatErrorSync(err),
        validationErrors: issues.map((issue) => {
          let path = issue.path.length > 0 ? issue.path.join(".") : "(root)";
          return redact ? `${path}: [${issue._tag}]` : `${path}: ${issue.message}`;
        }),
        url: redact ? redactUrlForProduction(context?.url) : context?.url,
        responseBody: redact ? void 0 : stringifyResponseBody(data)
      });
    }
  )
);

// src/lib/stringifyQuery.ts
function stringifyQuery(obj, options = {}) {
  if (!obj || typeof obj != "object" || Object.keys(obj).length === 0)
    return "";
  let processValue = (key, value) => {
    if (Array.isArray(value))
      return options.indices === false ? value.map(
        (item) => `${encodeURIComponent(key)}=${encodeURIComponent(String(item))}`
      ) : value.map(
        (item, index) => `${encodeURIComponent(key)}[${index}]=${encodeURIComponent(String(item))}`
      );
    if (value == null)
      return [];
    if (typeof value == "object") {
      let nested = [];
      for (let [subKey, subValue] of Object.entries(value))
        nested.push(...processValue(`${key}[${subKey}]`, subValue));
      return nested;
    }
    return [`${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`];
  }, pairs = [];
  for (let [key, value] of Object.entries(obj))
    pairs.push(...processValue(key, value));
  let queryString = pairs.join("&");
  return queryString ? options.addQueryPrefix === false ? queryString : `?${queryString}` : "";
}
var DefaultService = class {
  baseUrl = "https://api.solapi.com";
  authInfo;
  constructor(apiKey, apiSecret) {
    this.authInfo = {
      apiKey,
      apiSecret
    };
  }
  requestEffect(parameter) {
    let { httpMethod, url, body, responseSchema } = parameter, requestConfig = {
      method: httpMethod,
      url: `${this.baseUrl}/${url}`
    };
    return responseSchema ? Effect10.flatMap(
      defaultFetcherEffect(this.authInfo, requestConfig, body),
      (data) => decodeServerResponse(responseSchema, data, { url: requestConfig.url })
    ) : defaultFetcherEffect(this.authInfo, requestConfig, body);
  }
  async request(parameter) {
    return runSafePromise(this.requestEffect(parameter));
  }
  getWithQuery(config) {
    let reqEffect = this.requestEffect.bind(this);
    return Effect10.gen(function* () {
      let validated = config.data ? yield* decodeWithBadRequest(config.schema, config.data) : void 0, payload = yield* safeFinalize(() => config.finalize(validated)), parameter = stringifyQuery(payload, {
        indices: false,
        addQueryPrefix: true
      });
      return yield* reqEffect({
        httpMethod: "GET",
        url: `${config.url}${parameter}`,
        responseSchema: config.responseSchema
      });
    });
  }
};

// src/services/cash/cashService.ts
var CashService = class extends DefaultService {
  /**
   * 잔액조회
   * @returns GetBalanceResponse
   */
  async getBalance() {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "GET",
        url: "cash/v1/balance",
        responseSchema: getBalanceResponseSchema
      })
    );
  }
};
var getBlacksRequestSchema = Schema.Struct({
  senderNumber: Schema.optional(Schema.String),
  startKey: Schema.optional(Schema.String),
  limit: Schema.optional(Schema.Number),
  startDate: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf)),
  endDate: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf))
});
function finalizeGetBlacksRequest(data) {
  if (!data) return { type: "DENIAL" };
  let payload = { type: "DENIAL" };
  return payload.senderNumber = data.senderNumber, payload.startKey = data.startKey, payload.limit = data.limit, data.startDate != null && (payload.dateCreated = Object.assign(payload.dateCreated ?? {}, {
    gte: formatWithTransfer(data.startDate)
  })), data.endDate != null && (payload.dateCreated = Object.assign(payload.dateCreated ?? {}, {
    lte: formatWithTransfer(data.endDate)
  })), payload;
}
var likeLiteralSchema = Schema.Struct({ like: Schema.String }), datePayloadSchema = Schema.Struct({
  eq: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf)),
  gte: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf)),
  lte: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf)),
  gt: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf)),
  lt: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf))
});

// src/models/requests/iam/getBlockGroupsRequest.ts
var getBlockGroupsRequestSchema = Schema.Struct({
  blockGroupId: Schema.optional(Schema.String),
  useAll: Schema.optional(Schema.Boolean),
  senderNumber: Schema.optional(Schema.String),
  name: Schema.optional(Schema.Union(Schema.String, likeLiteralSchema)),
  status: Schema.optional(Schema.Literal("ACTIVE", "INACTIVE")),
  startKey: Schema.optional(Schema.String),
  limit: Schema.optional(Schema.Number)
});
function finalizeGetBlockGroupsRequest(data) {
  if (!data) return {};
  let payload = {
    blockGroupId: data.blockGroupId,
    useAll: data.useAll,
    senderNumber: data.senderNumber,
    status: data.status,
    startKey: data.startKey,
    limit: data.limit
  };
  return data.name != null && (payload.name = typeof data.name == "string" ? { like: data.name } : data.name), payload;
}
var getBlockNumbersRequestSchema = Schema.Struct({
  blockNumberId: Schema.optional(Schema.String),
  phoneNumber: Schema.optional(Schema.String),
  blockGroupId: Schema.optional(Schema.String),
  memo: Schema.optional(Schema.Union(Schema.String, likeLiteralSchema)),
  startKey: Schema.optional(Schema.String),
  limit: Schema.optional(Schema.Number)
});
function finalizeGetBlockNumbersRequest(data) {
  if (!data) return {};
  let payload = {
    blockNumberId: data.blockNumberId,
    phoneNumber: data.phoneNumber,
    blockGroupId: data.blockGroupId,
    startKey: data.startKey,
    limit: data.limit
  };
  return data.memo != null && (payload.memo = typeof data.memo == "string" ? { like: data.memo } : data.memo), payload;
}
var getBlacksResponseSchema = Schema.Struct({
  startKey: Schema.NullishOr(Schema.String),
  limit: Schema.Number,
  nextKey: Schema.NullishOr(Schema.String),
  blackList: Schema.Array(blackSchema)
});
var getBlockGroupsResponseSchema = Schema.Struct({
  startKey: Schema.NullishOr(Schema.String),
  limit: Schema.Number,
  nextKey: Schema.NullishOr(Schema.String),
  blockGroups: Schema.Array(blockGroupSchema)
});
var getBlockNumbersResponseSchema = Schema.Struct({
  startKey: Schema.NullishOr(Schema.String),
  limit: Schema.Number,
  nextKey: Schema.NullishOr(Schema.String),
  blockNumbers: Schema.Array(blockNumberSchema)
});

// src/services/iam/iamService.ts
var IamService = class extends DefaultService {
  /**
   * 080 수신 거부 조회
   * @param data 080 수신 거부 상세 조회용 request 데이터
   * @returns GetBlacksResponse
   */
  async getBlacks(data) {
    return runSafePromise(
      this.getWithQuery({
        schema: getBlacksRequestSchema,
        finalize: finalizeGetBlacksRequest,
        url: "iam/v1/black",
        data,
        responseSchema: getBlacksResponseSchema
      })
    );
  }
  /**
   * 수신 거부 그룹 조회
   * @param data 수신 거부 그룹 조회용 request 데이터
   * @returns GetBlockGroupsResponse
   */
  async getBlockGroups(data) {
    return runSafePromise(
      this.getWithQuery({
        schema: getBlockGroupsRequestSchema,
        finalize: finalizeGetBlockGroupsRequest,
        url: "iam/v1/block/groups",
        data,
        responseSchema: getBlockGroupsResponseSchema
      })
    );
  }
  /**
   * 수신 거부 번호 조회
   * @param data 수신 거부 번호 상세 조회용 request 데이터
   * @returns GetBlockNumbersResponse
   */
  async getBlockNumbers(data) {
    return runSafePromise(
      this.getWithQuery({
        schema: getBlockNumbersRequestSchema,
        finalize: finalizeGetBlockNumbersRequest,
        url: "iam/v1/block/numbers",
        data,
        responseSchema: getBlockNumbersResponseSchema
      })
    );
  }
};
var kakaoChannelCategorySchema = Schema.Struct({
  code: Schema.String,
  name: Schema.String
}), kakaoChannelSchema = Schema.Struct({
  channelId: Schema.String,
  searchId: Schema.String,
  accountId: Schema.String,
  phoneNumber: Schema.optional(Schema.String),
  sharedAccountIds: Schema.Array(Schema.String),
  dateCreated: Schema.optional(
    Schema.Union(Schema.String, Schema.DateFromSelf)
  ),
  dateUpdated: Schema.optional(
    Schema.Union(Schema.String, Schema.DateFromSelf)
  )
});
function decodeKakaoChannel(data) {
  return Effect10.gen(function* () {
    let dateCreated = yield* safeDateTransfer(data.dateCreated), dateUpdated = yield* safeDateTransfer(data.dateUpdated);
    return {
      channelId: data.channelId,
      searchId: data.searchId,
      accountId: data.accountId,
      phoneNumber: data.phoneNumber,
      sharedAccountIds: data.sharedAccountIds,
      dateCreated,
      dateUpdated
    };
  });
}
var getKakaoChannelsRequestSchema = Schema.Struct({
  channelId: Schema.optional(Schema.String),
  searchId: Schema.optional(Schema.String),
  phoneNumber: Schema.optional(Schema.String),
  categoryCode: Schema.optional(Schema.String),
  startKey: Schema.optional(Schema.String),
  limit: Schema.optional(Schema.Number),
  isMine: Schema.optional(Schema.Boolean),
  startDate: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf)),
  endDate: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf))
});
function finalizeGetKakaoChannelsRequest(data) {
  if (!data) return {};
  let payload = {
    channelId: data.channelId,
    searchId: data.searchId,
    phoneNumber: data.phoneNumber,
    categoryCode: data.categoryCode,
    startKey: data.startKey,
    limit: data.limit,
    isMine: data.isMine
  };
  return data.startDate != null && (payload.dateCreated = Object.assign(payload.dateCreated ?? {}, {
    gte: formatWithTransfer(data.startDate)
  })), data.endDate != null && (payload.dateCreated = Object.assign(payload.dateCreated ?? {}, {
    lte: formatWithTransfer(data.endDate)
  })), payload;
}
var getKakaoChannelsResponseSchema = Schema.Struct({
  limit: Schema.Number,
  startKey: Schema.NullishOr(Schema.String),
  nextKey: Schema.NullishOr(Schema.String),
  channelList: Schema.Array(kakaoChannelSchema)
});
var kakaoChannelCategoryListSchema = Schema.Array(kakaoChannelCategorySchema), KakaoChannelService = class extends DefaultService {
  async getKakaoChannelCategories() {
    return runSafePromise(
      Effect10.map(
        this.requestEffect({
          httpMethod: "GET",
          url: "kakao/v2/channels/categories",
          responseSchema: kakaoChannelCategoryListSchema
        }),
        (list) => [...list]
      )
    );
  }
  async getKakaoChannels(data) {
    let reqEffect = this.requestEffect.bind(this);
    return runSafePromise(
      Effect10.gen(function* () {
        let validated = data ? yield* decodeWithBadRequest(getKakaoChannelsRequestSchema, data) : void 0, payload = yield* safeFinalize(
          () => finalizeGetKakaoChannelsRequest(validated)
        ), parameter = stringifyQuery(payload, {
          indices: false,
          addQueryPrefix: true
        }), response = yield* reqEffect({
          httpMethod: "GET",
          url: `kakao/v2/channels${parameter}`,
          responseSchema: getKakaoChannelsResponseSchema
        });
        return {
          limit: response.limit,
          nextKey: response.nextKey,
          startKey: response.startKey,
          channelList: yield* Effect10.all(
            response.channelList.map(decodeKakaoChannel)
          )
        };
      })
    );
  }
  async getKakaoChannel(channelId) {
    return runSafePromise(
      Effect10.flatMap(
        this.requestEffect({
          httpMethod: "GET",
          url: `kakao/v2/channels/${channelId}`,
          responseSchema: kakaoChannelSchema
        }),
        decodeKakaoChannel
      )
    );
  }
  async requestKakaoChannelToken(data) {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "POST",
        url: "kakao/v2/channels/token",
        body: data
      })
    );
  }
  async createKakaoChannel(data) {
    return runSafePromise(
      this.requestEffect(
        {
          httpMethod: "POST",
          url: "kakao/v2/channels",
          body: data
        }
      )
    );
  }
  async removeKakaoChannel(channelId) {
    return runSafePromise(
      Effect10.flatMap(
        this.requestEffect({
          httpMethod: "DELETE",
          url: `kakao/v2/channels/${channelId}`
        }),
        decodeKakaoChannel
      )
    );
  }
};
var kakaoAlimtalkTemplateQuickReplyWebLinkSchema = Schema.Struct({
  name: Schema.String,
  linkType: Schema.Literal("WL"),
  linkMo: Schema.String,
  linkPc: Schema.optional(Schema.String)
}), kakaoAlimtalkTemplateQuickReplyAppLinkSchema = Schema.Struct({
  name: Schema.String,
  linkType: Schema.Literal("AL"),
  linkAnd: Schema.String,
  linkIos: Schema.String
}), kakaoAlimtalkTemplateQuickReplyDefaultSchema = Schema.Struct({
  name: Schema.String,
  linkType: Schema.Literal("BK", "BT", "BC")
}), kakaoAlimtalkTemplateQuickReplySchema = Schema.Union(
  kakaoAlimtalkTemplateQuickReplyWebLinkSchema,
  kakaoAlimtalkTemplateQuickReplyAppLinkSchema,
  kakaoAlimtalkTemplateQuickReplyDefaultSchema
);

// src/models/base/kakao/kakaoAlimtalkTemplate.ts
var kakaoAlimtalkTemplateMessageTypeSchema = Schema.Literal(
  "BA",
  "EX",
  "AD",
  "MI"
), kakaoAlimtalkTemplateEmphasizeTypeSchema = Schema.Literal(
  "NONE",
  "TEXT",
  "IMAGE",
  "ITEM_LIST"
), kakaoAlimtalkTemplateAssignTypeSchema = Schema.Literal(
  "CHANNEL",
  "GROUP"
), kakaoAlimtalkTemplateStatusSchema = Schema.Literal(
  "PENDING",
  "INSPECTING",
  "APPROVED",
  "REJECTED"
), kakaoAlimtalkTemplateCommentTypeSchema = Schema.Struct({
  isAdmin: Schema.Boolean,
  memberId: Schema.String,
  content: Schema.NullOr(Schema.String),
  dateCreated: Schema.String
}), kakaoAlimtalkTemplateHighlightTypeSchema = Schema.Struct({
  title: Schema.optional(Schema.NullOr(Schema.String)),
  description: Schema.optional(Schema.NullOr(Schema.String)),
  imageId: Schema.optional(Schema.NullOr(Schema.String))
}), kakaoAlimtalkTemplateItemTypeSchema = Schema.Struct({
  list: Schema.Array(
    Schema.Struct({
      title: Schema.String,
      description: Schema.String
    })
  ),
  summary: Schema.Struct({
    title: Schema.optional(Schema.NullOr(Schema.String)),
    description: Schema.optional(Schema.NullOr(Schema.String))
  })
}), kakaoAlimtalkTemplateSchema = Schema.Struct({
  name: Schema.String,
  channelId: Schema.optional(Schema.NullOr(Schema.String)),
  channelGroupId: Schema.optional(Schema.NullOr(Schema.String)),
  content: Schema.optional(Schema.String),
  isHidden: Schema.optional(Schema.Boolean),
  messageType: kakaoAlimtalkTemplateMessageTypeSchema,
  emphasizeType: kakaoAlimtalkTemplateEmphasizeTypeSchema,
  extra: Schema.optional(Schema.NullOr(Schema.String)),
  ad: Schema.optional(Schema.NullOr(Schema.String)),
  emphasizeTitle: Schema.optional(Schema.NullOr(Schema.String)),
  emphasizeSubtitle: Schema.optional(Schema.NullOr(Schema.String)),
  securityFlag: Schema.Boolean,
  imageId: Schema.optional(Schema.NullOr(Schema.String)),
  assignType: Schema.optional(kakaoAlimtalkTemplateAssignTypeSchema),
  buttons: Schema.optional(Schema.Array(kakaoButtonSchema)),
  comments: Schema.optional(
    Schema.Array(kakaoAlimtalkTemplateCommentTypeSchema)
  ),
  commentable: Schema.optional(Schema.Boolean),
  quickReplies: Schema.optional(
    Schema.Array(kakaoAlimtalkTemplateQuickReplySchema)
  ),
  header: Schema.optional(Schema.NullOr(Schema.String)),
  highlight: Schema.optional(
    Schema.NullOr(kakaoAlimtalkTemplateHighlightTypeSchema)
  ),
  item: Schema.optional(Schema.NullOr(kakaoAlimtalkTemplateItemTypeSchema)),
  templateId: Schema.String,
  code: Schema.optional(Schema.NullOr(Schema.String)),
  status: kakaoAlimtalkTemplateStatusSchema,
  variables: Schema.optional(
    Schema.Array(
      Schema.Struct({
        name: Schema.String
      })
    )
  ),
  dateCreated: Schema.optional(
    Schema.Union(Schema.String, Schema.DateFromSelf)
  ),
  dateUpdated: Schema.optional(
    Schema.Union(Schema.String, Schema.DateFromSelf)
  )
});
function decodeKakaoAlimtalkTemplate(data) {
  return Effect10.gen(function* () {
    let dateCreated = yield* safeDateTransfer(data.dateCreated), dateUpdated = yield* safeDateTransfer(data.dateUpdated);
    return {
      ...data,
      dateCreated,
      dateUpdated
    };
  });
}
var alimtalkTemplatesNameTypeSchema = Schema.Union(
  Schema.String,
  Schema.Struct({ like: Schema.String }),
  Schema.Struct({
    eq: Schema.optional(Schema.String),
    ne: Schema.optional(Schema.String)
  })
), getKakaoAlimtalkTemplatesRequestSchema = Schema.Struct({
  name: Schema.optional(alimtalkTemplatesNameTypeSchema),
  channelId: Schema.optional(Schema.String),
  templateId: Schema.optional(Schema.String),
  isHidden: Schema.optional(Schema.Boolean),
  status: Schema.optional(kakaoAlimtalkTemplateStatusSchema),
  startKey: Schema.optional(Schema.String),
  limit: Schema.optional(Schema.Number),
  startDate: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf)),
  endDate: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf))
});
function finalizeGetKakaoAlimtalkTemplatesRequest(data) {
  if (!data) return {};
  let payload = {
    channelId: data.channelId,
    isHidden: data.isHidden,
    templateId: data.templateId,
    startKey: data.startKey,
    status: data.status,
    limit: data.limit
  };
  return data.name != null && (payload.name = typeof data.name == "string" ? { like: data.name } : data.name), data.startDate != null && (payload.dateCreated = Object.assign(payload.dateCreated ?? {}, {
    gte: formatWithTransfer(data.startDate)
  })), data.endDate != null && (payload.dateCreated = Object.assign(payload.dateCreated ?? {}, {
    lte: formatWithTransfer(data.endDate)
  })), payload;
}
var getKakaoTemplateResponseSchema = kakaoAlimtalkTemplateSchema.pipe(
  Schema.omit("assignType", "commentable", "dateCreated", "dateUpdated"),
  Schema.extend(
    Schema.Struct({
      assignType: kakaoAlimtalkTemplateAssignTypeSchema,
      accountId: Schema.NullishOr(Schema.String),
      commentable: Schema.Boolean,
      dateCreated: Schema.String,
      dateUpdated: Schema.String
    })
  )
);

// src/models/responses/kakao/getKakaoAlimtalkTemplatesResponse.ts
var getKakaoAlimtalkTemplatesResponseSchema = Schema.Struct({
  limit: Schema.Number,
  templateList: Schema.Array(getKakaoTemplateResponseSchema),
  startKey: Schema.NullishOr(Schema.String),
  nextKey: Schema.NullishOr(Schema.String)
});
var kakaoAlimtalkTemplateCategoryListSchema = Schema.Array(
  kakaoChannelCategorySchema
), KakaoTemplateService = class extends DefaultService {
  /**
   * 카카오 템플릿 카테고리 조회
   */
  async getKakaoAlimtalkTemplateCategories() {
    return runSafePromise(
      Effect10.map(
        this.requestEffect({
          httpMethod: "GET",
          url: "kakao/v2/templates/categories",
          responseSchema: kakaoAlimtalkTemplateCategoryListSchema
        }),
        (list) => [...list]
      )
    );
  }
  /**
   * @description 카카오 알림톡 템플릿 생성
   */
  async createKakaoAlimtalkTemplate(data) {
    return runSafePromise(
      Effect10.flatMap(
        this.requestEffect({
          httpMethod: "POST",
          url: "kakao/v2/templates",
          body: data
        }),
        decodeKakaoAlimtalkTemplate
      )
    );
  }
  /**
   * 카카오 템플릿 목록 조회
   */
  async getKakaoAlimtalkTemplates(data) {
    let reqEffect = this.requestEffect.bind(this);
    return runSafePromise(
      Effect10.gen(function* () {
        let validated = data ? yield* decodeWithBadRequest(
          getKakaoAlimtalkTemplatesRequestSchema,
          data
        ) : void 0, payload = yield* safeFinalize(
          () => finalizeGetKakaoAlimtalkTemplatesRequest(validated)
        ), parameter = stringifyQuery(payload, {
          indices: false,
          addQueryPrefix: true
        }), response = yield* reqEffect({
          httpMethod: "GET",
          url: `kakao/v2/templates${parameter}`,
          responseSchema: getKakaoAlimtalkTemplatesResponseSchema
        }), templateList = yield* Effect10.all(
          response.templateList.map(
            (item) => Effect10.flatMap(
              decodeWithBadRequest(kakaoAlimtalkTemplateSchema, item),
              decodeKakaoAlimtalkTemplate
            )
          )
        );
        return {
          limit: response.limit,
          nextKey: response.nextKey,
          startKey: response.startKey,
          templateList
        };
      })
    );
  }
  /**
   * 카카오 템플릿 상세 조회
   */
  async getKakaoAlimtalkTemplate(templateId) {
    return runSafePromise(
      Effect10.flatMap(
        this.requestEffect({
          httpMethod: "GET",
          url: `kakao/v2/templates/${templateId}`,
          responseSchema: getKakaoTemplateResponseSchema
        }),
        decodeKakaoAlimtalkTemplate
      )
    );
  }
  /**
   * 카카오 알림톡 템플릿 검수 취소 요청
   */
  async cancelInspectionKakaoAlimtalkTemplate(templateId) {
    return runSafePromise(
      Effect10.flatMap(
        this.requestEffect({
          httpMethod: "PUT",
          url: `kakao/v2/templates/${templateId}/inspection/cancel`
        }),
        decodeKakaoAlimtalkTemplate
      )
    );
  }
  /**
   * 카카오 알림톡 템플릿 수정(검수 X)
   */
  async updateKakaoAlimtalkTemplate(templateId, data) {
    return runSafePromise(
      Effect10.flatMap(
        this.requestEffect({
          httpMethod: "PUT",
          url: `kakao/v2/templates/${templateId}`,
          body: data
        }),
        decodeKakaoAlimtalkTemplate
      )
    );
  }
  /**
   * 카카오 알림톡 템플릿 이름 수정(검수 상태 상관없이 변경가능)
   */
  async updateKakaoAlimtalkTemplateName(templateId, name) {
    return runSafePromise(
      Effect10.flatMap(
        this.requestEffect({
          httpMethod: "PUT",
          url: `kakao/v2/templates/${templateId}/name`,
          body: { name }
        }),
        decodeKakaoAlimtalkTemplate
      )
    );
  }
  /**
   * 카카오 알림톡 템플릿 삭제(대기, 반려 상태일 때만 삭제가능)
   */
  async removeKakaoAlimtalkTemplate(templateId) {
    return runSafePromise(
      Effect10.flatMap(
        this.requestEffect({
          httpMethod: "DELETE",
          url: `kakao/v2/templates/${templateId}`
        }),
        decodeKakaoAlimtalkTemplate
      )
    );
  }
};
var getGroupsRequestSchema = Schema.Struct({
  groupId: Schema.optional(Schema.String),
  startKey: Schema.optional(Schema.String),
  limit: Schema.optional(Schema.Number),
  startDate: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf)),
  endDate: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf))
});
function finalizeGetGroupsRequest(data) {
  if (!data) return {};
  let payload = {
    startKey: data.startKey,
    limit: data.limit
  };
  return data.groupId && (payload.criteria = "groupId", payload.cond = "eq", payload.value = data.groupId), data.startDate != null && (payload.startDate = formatWithTransfer(data.startDate)), data.endDate != null && (payload.endDate = formatWithTransfer(data.endDate)), payload;
}

// package.json
var package_default = {
  version: "6.0.1"};

// src/models/requests/messages/requestConfig.ts
var osPlatform = `${process.platform} | ${process.version}`, sdkVersion = `nodejs/${package_default.version}`, defaultAgentTypeSchema = Schema.Struct({
  sdkVersion: Schema.optional(Schema.String).pipe(
    Schema.withDecodingDefault(() => sdkVersion),
    Schema.withConstructorDefault(() => sdkVersion)
  ),
  osPlatform: Schema.optional(Schema.String).pipe(
    Schema.withDecodingDefault(() => osPlatform),
    Schema.withConstructorDefault(() => osPlatform)
  ),
  appId: Schema.optional(Schema.String)
}), sendRequestConfigSchema = Schema.Struct({
  scheduledDate: Schema.optional(
    Schema.Union(Schema.DateFromSelf, Schema.DateFromString).pipe(
      Schema.transformOrFail(Schema.String, {
        decode: (fromA, _, ast) => safeFormatWithTransfer(fromA).pipe(
          Effect.mapError(
            (err) => new ParseResult.Type(ast, fromA, err.message)
          )
        ),
        encode: (toI) => ParseResult.succeed(new Date(toI))
      })
    )
  ),
  allowDuplicates: Schema.optional(Schema.Boolean),
  appId: Schema.optional(Schema.String),
  showMessageList: Schema.optional(Schema.Boolean)
}), defaultMessageRequestSchema = Schema.Struct({
  allowDuplicates: Schema.optional(Schema.Boolean),
  agent: Schema.optional(defaultAgentTypeSchema)
});
var removeHyphens = (s) => s.replace(/-/g, ""), phoneNumberSchema = Schema.String.pipe(
  Schema.transform(Schema.String, {
    decode: removeHyphens,
    encode: (s) => s
  }),
  Schema.filter((s) => s.trim().length > 0, {
    message: () => "\uC804\uD654\uBC88\uD638\uB294 \uBE48 \uBB38\uC790\uC5F4\uC77C \uC218 \uC5C6\uC2B5\uB2C8\uB2E4."
  }),
  Schema.filter((s) => /^[0-9]+$/.test(s), {
    message: () => "\uC804\uD654\uBC88\uD638\uB294 \uC22B\uC790 \uBC0F \uD2B9\uC218\uBB38\uC790 - \uC678 \uBB38\uC790\uB97C \uD3EC\uD568\uD560 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4."
  })
), nonEmptyArrayFilter = (schema) => Schema.Array(schema).pipe(
  Schema.filter((arr) => arr.length > 0, {
    message: () => "\uB370\uC774\uD130\uAC00 \uBC18\uB4DC\uC2DC 1\uAC74 \uC774\uC0C1 \uAE30\uC785\uB418\uC5B4 \uC788\uC5B4\uC57C \uD569\uB2C8\uB2E4."
  })
), requestSendOneMessageSchema = messageSchema.pipe(
  Schema.omit("to", "from"),
  Schema.extend(
    Schema.Struct({
      to: Schema.Union(phoneNumberSchema, Schema.Array(phoneNumberSchema)),
      from: Schema.optional(phoneNumberSchema)
    })
  )
), requestSendMessageSchema = Schema.Union(
  requestSendOneMessageSchema,
  nonEmptyArrayFilter(requestSendOneMessageSchema)
), defaultAgentValue = Schema.decodeSync(defaultAgentTypeSchema)({}), agentWithDefaultSchema = Schema.optional(defaultAgentTypeSchema).pipe(
  Schema.withDecodingDefault(() => defaultAgentValue),
  Schema.withConstructorDefault(() => defaultAgentValue)
), multipleMessageSendingRequestSchema = Schema.Struct({
  allowDuplicates: Schema.optional(Schema.Boolean),
  agent: agentWithDefaultSchema,
  messages: nonEmptyArrayFilter(requestSendOneMessageSchema),
  scheduledDate: Schema.optional(
    Schema.Union(Schema.DateFromSelf, Schema.DateFromString)
  ),
  showMessageList: Schema.optional(Schema.Boolean)
});
var GroupService = class extends DefaultService {
  /**
   * 그룹 생성
   * @param allowDuplicates 생성할 그룹이 중복 수신번호를 허용하는지 여부를 확인합니다.
   * @param appId 생성할 그룹에 특정 appId를 넣을 수 있습니다.
   * @param customFields 생성할 그룹에 사용자 정의 데이터를 Record 형태로 삽입할 수 있습니다.
   */
  async createGroup(allowDuplicates = false, appId, customFields) {
    return runSafePromise(
      Effect10.map(
        this.requestEffect({
          httpMethod: "POST",
          url: "messages/v4/groups",
          body: {
            sdkVersion,
            osPlatform,
            allowDuplicates,
            appId,
            customFields
          }
        }),
        (response) => response.groupId
      )
    );
  }
  /**
   * 그룹 메시지 추가
   * 한번 요청으로 최대 10,000건의 메시지를 추가할 수 있습니다.
   * 추가 가능한 최대 메시지 건 수는 1,000,000건 입니다.
   * @param groupId 생성 된 Group ID
   * @param messages 여러 메시지(문자, 알림톡 등)
   */
  async addMessagesToGroup(groupId, messages) {
    let reqEffect = this.requestEffect.bind(this);
    return runSafePromise(
      Effect10.flatMap(
        decodeWithBadRequest(requestSendMessageSchema, messages),
        (validatedMessages) => reqEffect({
          httpMethod: "PUT",
          url: `messages/v4/groups/${groupId}/messages`,
          body: {
            messages: Array.isArray(validatedMessages) ? validatedMessages : [validatedMessages]
          }
        })
      )
    );
  }
  /**
   * 그룹 메시지 전송 요청
   * @param groupId 생성 된 Group ID
   */
  async sendGroup(groupId) {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "POST",
        url: `messages/v4/groups/${groupId}/send`
      })
    );
  }
  /**
   * 그룹 예약 발송 설정
   * @param groupId 생성 된 Group ID
   * @param scheduledDate 예약발송 할 날짜
   */
  async reserveGroup(groupId, scheduledDate) {
    let reqEffect = this.requestEffect.bind(this);
    return runSafePromise(
      Effect10.flatMap(
        safeFormatWithTransfer(scheduledDate),
        (formattedScheduledDate) => reqEffect({
          httpMethod: "POST",
          url: `messages/v4/groups/${groupId}/schedule`,
          body: { scheduledDate: formattedScheduledDate }
        })
      )
    );
  }
  /**
   * 그룹 내 예약 발송 취소(메시지 실패 전체 처리 됨)
   * @param groupId 생성 된 Group ID
   */
  async removeReservationToGroup(groupId) {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "DELETE",
        url: `messages/v4/groups/${groupId}/schedule`
      })
    );
  }
  /**
   * 그룹 목록 정보 조회
   * @param data 그룹 정보 상세 조회용 request 데이터
   */
  async getGroups(data) {
    return runSafePromise(
      this.getWithQuery({
        schema: getGroupsRequestSchema,
        finalize: finalizeGetGroupsRequest,
        url: "messages/v4/groups",
        data,
        responseSchema: getGroupsResponseSchema
      })
    );
  }
  /**
   * 단일 그룹정보 조회
   * @param groupId 그룹 ID
   */
  async getGroup(groupId) {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "GET",
        url: `messages/v4/groups/${groupId}`,
        responseSchema: groupMessageResponseSchema
      })
    );
  }
  /**
   * 그룹 내 메시지 목록 조회
   * @param groupId 생성 된 Group ID
   * @param data startkey, limit 등 쿼리 조건 파라미터
   */
  async getGroupMessages(groupId, data) {
    let parameter = stringifyQuery(data, {
      indices: false,
      addQueryPrefix: true
    });
    return runSafePromise(
      this.requestEffect({
        httpMethod: "GET",
        url: `messages/v4/groups/${groupId}/messages${parameter}`,
        responseSchema: getMessagesResponseSchema
      })
    );
  }
  /**
   * 그룹 내 특정 메시지 삭제
   * @param groupId 생성 된 Group Id
   * @param messageIds 생성 된 메시지 ID 목록
   */
  async removeGroupMessages(groupId, messageIds) {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "DELETE",
        url: `messages/v4/groups/${groupId}/messages`,
        body: { messageIds }
      })
    );
  }
  /**
   * 그룹 삭제
   * @param groupId
   */
  async removeGroup(groupId) {
    return runSafePromise(
      this.requestEffect({
        httpMethod: "DELETE",
        url: `messages/v4/groups/${groupId}`
      })
    );
  }
};
var dateTypeSchema = Schema.Literal("CREATED", "UPDATED"), baseGetMessagesRequestSchema = Schema.Struct({
  startKey: Schema.optional(Schema.String),
  limit: Schema.optional(Schema.Number),
  messageId: Schema.optional(Schema.String),
  messageIds: Schema.optional(Schema.Array(Schema.String)),
  groupId: Schema.optional(Schema.String),
  to: Schema.optional(Schema.String),
  from: Schema.optional(Schema.String),
  type: Schema.optional(messageTypeSchema),
  statusCode: Schema.optional(Schema.String),
  dateType: Schema.optional(dateTypeSchema),
  startDate: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf)),
  endDate: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf))
}), getMessagesRequestSchema = baseGetMessagesRequestSchema.pipe(
  Schema.filter((data) => {
    let hasDate = data.startDate != null || data.endDate != null;
    return data.dateType != null && !hasDate ? "dateType\uC740 startDate \uB610\uB294 endDate\uC640 \uD568\uAED8 \uC0AC\uC6A9\uD574\uC57C \uD569\uB2C8\uB2E4." : true;
  })
);
function finalizeGetMessagesRequest(data) {
  if (!data) return {};
  let payload = {
    startKey: data.startKey,
    limit: data.limit,
    dateType: data.startDate != null || data.endDate != null ? data.dateType ?? "CREATED" : data.dateType,
    messageId: data.messageId,
    messageIds: data.messageIds,
    groupId: data.groupId,
    to: data.to,
    from: data.from,
    type: data.type,
    statusCode: data.statusCode
  };
  return data.startDate != null && (payload.startDate = formatWithTransfer(data.startDate)), data.endDate != null && (payload.endDate = formatWithTransfer(data.endDate)), payload;
}
var getStatisticsRequestSchema = Schema.Struct({
  masterAccountId: Schema.optional(Schema.String),
  startDate: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf)),
  endDate: Schema.optional(Schema.Union(Schema.String, Schema.DateFromSelf))
});
function finalizeGetStatisticsRequest(data) {
  if (!data) return {};
  let payload = {
    masterAccountId: data.masterAccountId
  };
  return data.startDate != null && (payload.startDate = formatWithTransfer(data.startDate)), data.endDate != null && (payload.endDate = formatWithTransfer(data.endDate)), payload;
}
var MessageService = class extends DefaultService {
  /**
   * 메시지 발송 기능, sendMany 함수보다 개선된 오류 표시 기능등을 제공합니다.
   * 한번의 요청으로 최대 10,000건까지 발송할 수 있습니다.
   * @param messages 발송 요청할 메시지 파라미터(문자, 알림톡 등)
   * @param requestConfigParameter request시 필요한 파라미터 오브젝트
   * @throws MessageNotReceivedError 모든 메시지 접수건이 실패건으로 진행되는 경우 반환되는 에러
   * @throws BadRequestError 잘못된 파라미터를 기입했거나, 데이터가 아예 없는 경우 반환되는 에러
   */
  async send(messages, requestConfigParameter) {
    let reqEffect = this.requestEffect.bind(this);
    return runSafePromise(
      Effect10.gen(function* () {
        let messageSchema2 = yield* decodeWithBadRequest(
          requestSendMessageSchema,
          messages
        ), messageParameters = Array.isArray(messageSchema2) ? messageSchema2 : [messageSchema2];
        if (messageParameters.length === 0)
          return yield* new BadRequestError({
            message: "\uB370\uC774\uD130\uAC00 \uBC18\uB4DC\uC2DC 1\uAC74 \uC774\uC0C1 \uAE30\uC785\uB418\uC5B4 \uC788\uC5B4\uC57C \uD569\uB2C8\uB2E4."
          });
        let decodedConfig = yield* decodeWithBadRequest(
          sendRequestConfigSchema,
          requestConfigParameter ?? {}
        ), parameterObject = {
          messages: messageParameters,
          allowDuplicates: decodedConfig.allowDuplicates,
          ...decodedConfig.appId ? { agent: { appId: decodedConfig.appId } } : {},
          scheduledDate: decodedConfig.scheduledDate,
          showMessageList: decodedConfig.showMessageList
        }, parameter = yield* decodeWithBadRequest(
          multipleMessageSendingRequestSchema,
          parameterObject
        ), response = yield* reqEffect({
          httpMethod: "POST",
          url: "messages/v4/send-many/detail",
          body: parameter
        }), { count } = response.groupInfo;
        return response.failedMessageList.length > 0 && count.total === count.registeredFailed ? yield* new MessageNotReceivedError({
          failedMessageList: response.failedMessageList,
          totalCount: response.failedMessageList.length
        }) : response;
      })
    );
  }
  /**
   * 메시지 목록 조회
   * @param data 목록 조회 상세조건 파라미터
   */
  async getMessages(data) {
    return runSafePromise(
      this.getWithQuery({
        schema: getMessagesRequestSchema,
        finalize: finalizeGetMessagesRequest,
        url: "messages/v4/list",
        data,
        responseSchema: getMessagesResponseSchema
      })
    );
  }
  /**
   * 통계 조회
   * @param data 통계 상세 조건 파라미터
   * @returns GetStatisticsResponse 통계 결과
   */
  async getStatistics(data) {
    return runSafePromise(
      this.getWithQuery({
        schema: getStatisticsRequestSchema,
        finalize: finalizeGetStatisticsRequest,
        url: "messages/v4/statistics",
        data,
        responseSchema: getStatisticsResponseSchema
      })
    );
  }
};
var isHttpUrl = (value) => {
  try {
    let url = new URL$1(value);
    return url.protocol === "http:" || url.protocol === "https:";
  } catch {
    return false;
  }
}, fromUrl = (url) => Effect10.flatMap(
  Effect10.tryPromise({
    try: () => fetch(url),
    catch: (error) => new DefaultError({
      errorCode: "FileUrlFetchError",
      errorMessage: `\uB124\uD2B8\uC6CC\uD06C \uC624\uB958\uB85C URL(${url})\uC744(\uB97C) \uAC00\uC838\uC624\uC9C0 \uBABB\uD588\uC2B5\uB2C8\uB2E4.`,
      context: { url, cause: String(error) }
    })
  }),
  (response) => response.ok ? Effect10.tryPromise({
    try: () => response.arrayBuffer(),
    catch: (error) => new DefaultError({
      errorCode: "FileReadError",
      errorMessage: "\uC751\uB2F5 body \uCC98\uB9AC \uC911 \uC624\uB958\uAC00 \uBC1C\uC0DD\uD588\uC2B5\uB2C8\uB2E4.",
      context: { url, cause: String(error) }
    })
  }) : Effect10.fail(
    new DefaultError({
      errorCode: "FileUrlFetchError",
      errorMessage: `URL(${url}) \uC694\uCCAD \uC2E4\uD328 \u2013 \uC0C1\uD0DC \uCF54\uB4DC: ${response.status} ${response.statusText}`,
      context: { url, status: response.status }
    })
  )
).pipe(
  Effect10.map((arrayBuffer) => Buffer.from(arrayBuffer).toString("base64"))
), fromPath = (path) => Effect10.tryPromise({
  try: () => promises.readFile(path),
  catch: (error) => new DefaultError({
    errorCode: "FileReadError",
    errorMessage: `\uD30C\uC77C\uC744 \uC77D\uC744 \uC218 \uC5C6\uC2B5\uB2C8\uB2E4: ${path}`,
    context: { path, cause: String(error) }
  })
}).pipe(Effect10.map((buffer) => buffer.toString("base64")));
function fileToBase64Effect(path) {
  return isHttpUrl(path) ? fromUrl(path) : fromPath(path);
}
var StorageService = class extends DefaultService {
  /**
   * 파일(이미지) 업로드
   * 카카오 친구톡 이미지는 500kb, MMS는 200kb, 발신번호 서류 인증용 파일은 2mb의 제한이 있음
   * @param filePath 해당 파일의 경로 또는 접근 가능한 이미지 URL
   * @param fileType 저장할 파일의 유형, 예) 카카오 친구톡 용 이미지 -> KAKAO, MMS용 사진 -> MMS, 발신번호 서류 인증에 쓰이는 문서 등 -> DOCUMENT, RCS 이미지 -> RCS
   * @param name 파일 이름
   * @param link 파일 링크, 친구톡의 경우 필수 값
   */
  async uploadFile(filePath, fileType, name, link) {
    let reqEffect = this.requestEffect.bind(this);
    return runSafePromise(
      Effect10.flatMap(
        fileToBase64Effect(filePath),
        (encodedFile) => reqEffect({
          httpMethod: "POST",
          url: "storage/v1/files",
          body: { file: encodedFile, type: fileType, name, link }
        })
      )
    );
  }
};
var baseKakaoAlimtalkTemplateRequestSchema = Schema.Struct({
  name: Schema.String,
  content: Schema.String,
  categoryCode: Schema.String,
  buttons: Schema.optional(Schema.Array(kakaoButtonSchema)),
  quickReplies: Schema.optional(
    Schema.Array(kakaoAlimtalkTemplateQuickReplySchema)
  ),
  messageType: Schema.optional(kakaoAlimtalkTemplateMessageTypeSchema),
  emphasizeType: Schema.optional(kakaoAlimtalkTemplateEmphasizeTypeSchema),
  header: Schema.optional(Schema.String),
  highlight: Schema.optional(kakaoAlimtalkTemplateHighlightTypeSchema),
  item: Schema.optional(kakaoAlimtalkTemplateItemTypeSchema),
  extra: Schema.optional(Schema.String),
  emphasizeTitle: Schema.optional(Schema.String),
  emphasizeSubTitle: Schema.optional(Schema.String),
  securityFlag: Schema.optional(Schema.Boolean),
  imageId: Schema.optional(Schema.String)
}), createKakaoChannelAlimtalkTemplateRequestSchema = Schema.extend(
  baseKakaoAlimtalkTemplateRequestSchema,
  Schema.Struct({ channelId: Schema.String })
), createKakaoChannelGroupAlimtalkTemplateRequestSchema = Schema.extend(
  baseKakaoAlimtalkTemplateRequestSchema,
  Schema.Struct({ channelGroupId: Schema.String })
), createKakaoAlimtalkTemplateRequestSchema = Schema.Union(
  createKakaoChannelAlimtalkTemplateRequestSchema,
  createKakaoChannelGroupAlimtalkTemplateRequestSchema
);
var createKakaoChannelTokenRequestSchema = Schema.Struct({
  searchId: Schema.String,
  phoneNumber: Schema.String
}), createKakaoChannelRequestSchema = Schema.Struct({
  searchId: Schema.String,
  phoneNumber: Schema.String,
  categoryCode: Schema.String,
  token: Schema.String
});
var updateKakaoAlimtalkTemplateRequestSchema = Schema.Struct({
  name: Schema.optional(Schema.String),
  content: Schema.optional(Schema.String),
  categoryCode: Schema.optional(Schema.String),
  buttons: Schema.optional(Schema.Array(kakaoButtonSchema)),
  quickReplies: Schema.optional(
    Schema.Array(kakaoAlimtalkTemplateQuickReplySchema)
  ),
  messageType: Schema.optional(kakaoAlimtalkTemplateMessageTypeSchema),
  emphasizeType: Schema.optional(kakaoAlimtalkTemplateEmphasizeTypeSchema),
  header: Schema.optional(Schema.String),
  highlight: Schema.optional(kakaoAlimtalkTemplateHighlightTypeSchema),
  item: Schema.optional(kakaoAlimtalkTemplateItemTypeSchema),
  extra: Schema.optional(Schema.String),
  emphasizeTitle: Schema.optional(Schema.String),
  emphasizeSubTitle: Schema.optional(Schema.String),
  securityFlag: Schema.optional(Schema.Boolean),
  imageId: Schema.optional(Schema.String)
});
var groupMessageAddRequestSchema = Schema.Struct({
  messages: Schema.Array(messageSchema)
}), scheduledDateSendingRequestSchema = Schema.Struct({
  scheduledDate: Schema.String
}), removeMessageIdsToGroupRequestSchema = Schema.Struct({
  messageIds: Schema.Array(Schema.String)
}), getGroupMessagesRequestSchema = Schema.Struct({
  startKey: Schema.optional(Schema.String),
  limit: Schema.optional(Schema.Number)
}), fileIdsSchema = Schema.Struct({
  fileIds: Schema.Array(Schema.String)
}), fileTypeSchema = Schema.Literal(
  "KAKAO",
  "MMS",
  "DOCUMENT",
  "RCS",
  "FAX",
  "BMS",
  "BMS_WIDE",
  "BMS_WIDE_MAIN_ITEM_LIST",
  "BMS_WIDE_SUB_ITEM_LIST",
  "BMS_CAROUSEL_FEED_LIST",
  "BMS_CAROUSEL_COMMERCE_LIST"
), fileUploadRequestSchema = Schema.Struct({
  file: Schema.String,
  type: fileTypeSchema,
  name: Schema.optional(Schema.String),
  link: Schema.optional(Schema.String)
}), createGroupRequestSchema = Schema.extend(
  defaultAgentTypeSchema,
  Schema.Struct({
    allowDuplicates: Schema.Boolean,
    customFields: Schema.optional(
      Schema.Record({ key: Schema.String, value: Schema.String })
    )
  })
);
var failedMessageSchema = Schema.Struct({
  to: Schema.String,
  from: Schema.String,
  type: Schema.String,
  statusMessage: Schema.String,
  country: Schema.String,
  messageId: Schema.String,
  statusCode: Schema.String,
  accountId: Schema.String,
  customFields: Schema.optional(
    Schema.Record({ key: Schema.String, value: Schema.String })
  )
}), messageResponseItemSchema = Schema.Struct({
  messageId: Schema.String,
  statusCode: Schema.String,
  customFields: Schema.optional(
    Schema.Record({ key: Schema.String, value: Schema.String })
  ),
  statusMessage: Schema.String
}), detailGroupMessageResponseSchema = Schema.Struct({
  /**
   * 메시지 발송 접수에 실패한 메시지 요청 목록들
   */
  failedMessageList: Schema.Array(failedMessageSchema),
  /**
   * 발송 정보(성공, 실패 등) 응답 데이터
   */
  groupInfo: groupMessageResponseSchema,
  /**
   * Send 메소드 호출 당시 showMessageList 값이 true로 되어있을 때 표시되는 메시지 목록
   */
  messageList: Schema.optional(Schema.Array(messageResponseItemSchema))
});

// src/index.ts
var SolapiMessageService = class {
  /**
   * 잔액조회
   * @returns GetBalanceResponse
   */
  getBalance;
  /**
   * 080 수신 거부 조회
   * @param data 080 수신 거부 상세 조회용 request 데이터
   * @returns GetBlacksResponse
   */
  getBlacks;
  /**
   * 수신 거부 그룹 조회
   * @param data 수신 거부 그룹 조회용 request 데이터
   * @returns GetBlockGroupsResponse
   */
  getBlockGroups;
  /**
   * 수신 거부 번호 조회
   * @param data 수신 거부 번호 상세 조회용 request 데이터
   * @returns GetBlockNumbersResponse
   */
  getBlockNumbers;
  /**
   * 카카오 채널 카테고리 조회
   */
  getKakaoChannelCategories;
  /**
   * 카카오 채널 목록 조회
   * @param data 카카오 채널 목록을 더 자세하게 조회할 때 필요한 파라미터
   */
  getKakaoChannels;
  /**
   * @description 카카오 채널 조회
   * @param channelId 카카오 채널 ID(구 pfId)
   */
  getKakaoChannel;
  /**
   * @description 카카오 채널 연동을 위한 인증 토큰 요청
   */
  requestKakaoChannelToken;
  /**
   * @description 카카오 채널 연동 메소드
   * getKakaoChannelCategories, requestKakaoChannelToken 메소드를 선행적으로 호출해야 합니다!
   */
  createKakaoChannel;
  /**
   * @description 카카오 채널 삭제, 채널이 삭제 될 경우 해당 채널의 템플릿이 모두 삭제됩니다!
   * @param channelId 카카오 채널 ID
   */
  removeKakaoChannel;
  /**
   * 카카오 템플릿 카테고리 조회
   */
  getKakaoAlimtalkTemplateCategories;
  /**
   * @description 카카오 알림톡 템플릿 생성
   * 반드시 getKakaoAlimtalkTemplateCategories를 먼저 호출하여 카테고리 값을 확인해야 합니다!
   * @param data 알림톡 템플릿 생성을 위한 파라미터
   */
  createKakaoAlimtalkTemplate;
  /**
   * 카카오 템플릿 목록 조회
   * @param data 카카오 템플릿 목록을 더 자세하게 조회할 때 필요한 파라미터
   */
  getKakaoAlimtalkTemplates;
  /**
   * 카카오 템플릿 상세 조회
   * @param templateId 카카오 알림톡 템플릿 ID
   */
  getKakaoAlimtalkTemplate;
  /**
   * 카카오 알림톡 템플릿 검수 취소 요청
   * @param templateId 카카오 알림톡 템플릿 ID
   */
  cancelInspectionKakaoAlimtalkTemplate;
  /**
   * 카카오 알림톡 템플릿 수정(검수 X)
   * @param templateId 카카오 알림톡 템플릿 ID
   * @param data 카카오 알림톡 템플릿 수정을 위한 파라미터
   */
  updateKakaoAlimtalkTemplate;
  /**
   * 카카오 알림톡 템플릿 이름 수정(검수 상태 상관없이 변경가능)
   * @param templateId 카카오 알림톡 템플릿 ID
   * @param name 카카오 알림톡 템플릿 이름 변경을 위한 파라미터
   */
  updateKakaoAlimtalkTemplateName;
  /**
   * 카카오 알림톡 템플릿 삭제(대기, 반려 상태일 때만 삭제가능)
   * @param templateId 카카오 알림톡 템플릿 ID
   */
  removeKakaoAlimtalkTemplate;
  /**
   * 그룹 생성
   * @param allowDuplicates 생성할 그룹이 중복 수신번호를 허용하는지 여부를 확인합니다.
   * @param appId 생성할 그룹에 특정 appId를 넣을 수 있습니다.
   * @param customFields 생성할 그룹에 사용자 정의 데이터를 Record 형태로 삽입할 수 있습니다.
   */
  createGroup;
  /**
   * 그룹 메시지 추가
   * 한번 요청으로 최대 10,000건의 메시지를 추가할 수 있습니다.
   * 추가 가능한 최대 메시지 건 수는 1,000,000건 입니다.
   * @param groupId 생성 된 Group ID
   * @param messages 여러 메시지(문자, 알림톡 등)
   */
  addMessagesToGroup;
  /**
   * 그룹 메시지 전송 요청
   * @param groupId 생성 된 Group ID
   */
  sendGroup;
  /**
   * 그룹 예약 발송 설정
   * @param groupId 생성 된 Group ID
   * @param scheduledDate 예약발송 할 날짜
   */
  reserveGroup;
  /**
   * 그룹 내 예약 발송 취소(메시지 실패 전체 처리 됨)
   * @param groupId 생성 된 Group ID
   */
  removeReservationToGroup;
  /**
   * 그룹 목록 정보 조회
   * @param data 그룹 정보 상세 조회용 request 데이터
   */
  getGroups;
  /**
   * 단일 그룹정보 조회
   * @param groupId 그룹 ID
   */
  getGroup;
  /**
   * 그룹 내 메시지 목록 조회
   * @param groupId 생성 된 Group ID
   * @param data startkey, limit 등 쿼리 조건 파라미터
   */
  getGroupMessages;
  /**
   * 그룹 내 특정 메시지 삭제
   * @param groupId 생성 된 Group Id
   * @param messageIds 생성 된 메시지 ID 목록
   */
  removeGroupMessages;
  /**
   * 그룹 삭제
   * @param groupId
   */
  removeGroup;
  /**
   * 메시지 발송 기능, sendMany 함수보다 개선된 오류 표시 기능등을 제공합니다.
   * 한번의 요청으로 최대 10,000건까지 발송할 수 있습니다.
   * @param messages 발송 요청할 메시지 파라미터(문자, 알림톡 등)
   * @param requestConfigParameter request시 필요한 파라미터 오브젝트
   * @throws MessageNotReceivedError 모든 메시지 접수건이 실패건으로 진행되는 경우 반환되는 에러
   * @throws BadRequestError 잘못된 파라미터를 기입했거나, 데이터가 아예 없는 경우 반환되는 에러
   */
  send;
  /**
   * 메시지 목록 조회
   * @param data 목록 조회 상세조건 파라미터
   */
  getMessages;
  /**
   * 통계 조회
   * @param data 통계 상세 조건 파라미터
   * @returns GetStatisticsResponse 통계 결과
   */
  getStatistics;
  /**
   * 파일(이미지) 업로드
   * 카카오 친구톡 이미지는 500kb, MMS는 200kb, 발신번호 서류 인증용 파일은 2mb의 제한이 있음
   * @param filePath 해당 파일의 경로 또는 접근 가능한 이미지 URL
   * @param fileType 저장할 파일의 유형, 예) 카카오 친구톡 용 이미지 -> KAKAO, MMS용 사진 -> MMS, 발신번호 서류 인증에 쓰이는 문서 등 -> DOCUMENT, RCS 이미지 -> RCS
   * @param name 파일 이름
   * @param link 파일 링크, 친구톡의 경우 필수 값
   */
  uploadFile;
  constructor(apiKey, apiSecret) {
    if (!apiKey || !apiSecret)
      throw new ApiKeyError({
        message: "API Key\uC640 API Secret\uC740 \uD544\uC218\uC785\uB2C8\uB2E4."
      });
    let cashService = new CashService(apiKey, apiSecret), iamService = new IamService(apiKey, apiSecret), kakaoChannelService = new KakaoChannelService(apiKey, apiSecret), kakaoTemplateService = new KakaoTemplateService(apiKey, apiSecret), groupService = new GroupService(apiKey, apiSecret), messageService = new MessageService(apiKey, apiSecret), storageService = new StorageService(apiKey, apiSecret);
    this.getBalance = cashService.getBalance.bind(cashService), this.getBlacks = iamService.getBlacks.bind(iamService), this.getBlockGroups = iamService.getBlockGroups.bind(iamService), this.getBlockNumbers = iamService.getBlockNumbers.bind(iamService), this.getKakaoChannelCategories = kakaoChannelService.getKakaoChannelCategories.bind(kakaoChannelService), this.getKakaoChannels = kakaoChannelService.getKakaoChannels.bind(kakaoChannelService), this.getKakaoChannel = kakaoChannelService.getKakaoChannel.bind(kakaoChannelService), this.requestKakaoChannelToken = kakaoChannelService.requestKakaoChannelToken.bind(kakaoChannelService), this.createKakaoChannel = kakaoChannelService.createKakaoChannel.bind(kakaoChannelService), this.removeKakaoChannel = kakaoChannelService.removeKakaoChannel.bind(kakaoChannelService), this.getKakaoAlimtalkTemplateCategories = kakaoTemplateService.getKakaoAlimtalkTemplateCategories.bind(
      kakaoTemplateService
    ), this.createKakaoAlimtalkTemplate = kakaoTemplateService.createKakaoAlimtalkTemplate.bind(
      kakaoTemplateService
    ), this.getKakaoAlimtalkTemplates = kakaoTemplateService.getKakaoAlimtalkTemplates.bind(kakaoTemplateService), this.getKakaoAlimtalkTemplate = kakaoTemplateService.getKakaoAlimtalkTemplate.bind(kakaoTemplateService), this.cancelInspectionKakaoAlimtalkTemplate = kakaoTemplateService.cancelInspectionKakaoAlimtalkTemplate.bind(
      kakaoTemplateService
    ), this.updateKakaoAlimtalkTemplate = kakaoTemplateService.updateKakaoAlimtalkTemplate.bind(
      kakaoTemplateService
    ), this.updateKakaoAlimtalkTemplateName = kakaoTemplateService.updateKakaoAlimtalkTemplateName.bind(
      kakaoTemplateService
    ), this.removeKakaoAlimtalkTemplate = kakaoTemplateService.removeKakaoAlimtalkTemplate.bind(
      kakaoTemplateService
    ), this.createGroup = groupService.createGroup.bind(groupService), this.addMessagesToGroup = groupService.addMessagesToGroup.bind(groupService), this.sendGroup = groupService.sendGroup.bind(groupService), this.reserveGroup = groupService.reserveGroup.bind(groupService), this.removeReservationToGroup = groupService.removeReservationToGroup.bind(groupService), this.getGroups = groupService.getGroups.bind(groupService), this.getGroup = groupService.getGroup.bind(groupService), this.getGroupMessages = groupService.getGroupMessages.bind(groupService), this.removeGroupMessages = groupService.removeGroupMessages.bind(groupService), this.removeGroup = groupService.removeGroup.bind(groupService), this.send = messageService.send.bind(messageService), this.getMessages = messageService.getMessages.bind(messageService), this.getStatistics = messageService.getStatistics.bind(messageService), this.uploadFile = storageService.uploadFile.bind(storageService);
  }
};

export { ApiKeyError, BMS_ACCEPTABLE_FIELDS, BMS_ADDITIONAL_CONTENT_MAX, BMS_ALLOWED_LINK_TYPES, BMS_BUTTON_COUNT, BMS_BUTTON_NAME_MAX, BMS_CAROUSEL_COMMERCE_ADDITIONAL_CONTENT_MAX, BMS_CAROUSEL_FEED_ITEM_CONTENT_MAX, BMS_CAROUSEL_HEAD_CONTENT_MAX, BMS_CAROUSEL_LIST_RANGE, BMS_CONTENT_MAX, BMS_COUPON_DESCRIPTION_MAX, BMS_HEADER_MAX, BMS_MAIN_WIDE_ITEM_TITLE_MAX, BMS_SUB_WIDE_ITEM_TITLE_MAX, BadRequestError, ClientError, DefaultError, InvalidDateError, MessageNotReceivedError, NetworkError, ResponseSchemaMismatchError, ServerError, SolapiMessageService, UnexpectedDefectError, UnhandledExitError, VariableValidationError, addMessageResponseSchema, addMessageResultSchema, additionalBodySchema, appSchema, baseKakaoOptionSchema, blackSchema, blockGroupSchema, blockNumberSchema, bmsAppButtonSchema, bmsBotKeywordButtonSchema, bmsBotTransferButtonSchema, bmsBusinessFormButtonSchema, bmsButtonLinkTypeSchema, bmsButtonSchema, bmsCarouselCommerceItemSchema, bmsCarouselCommerceSchema, bmsCarouselFeedItemSchema, bmsCarouselFeedSchema, bmsCarouselHeadSchema, bmsCarouselTailSchema, bmsChannelAddButtonSchema, bmsChatBubbleTypeSchema, bmsCommerceSchema, bmsConsultButtonSchema, bmsCouponSchema, bmsCouponTitleSchema, bmsLinkButtonSchema, bmsMainWideItemSchema, bmsMessageDeliveryButtonSchema, bmsSubWideItemSchema, bmsVideoSchema, bmsWebButtonSchema, commonCashResponseSchema, countForChargeSchema, countSchema, createGroupRequestSchema, createKakaoAlimtalkTemplateRequestSchema, createKakaoChannelRequestSchema, createKakaoChannelResponseSchema, createKakaoChannelTokenRequestSchema, datePayloadSchema, dateTypeSchema, decodeKakaoAlimtalkTemplate, decodeKakaoChannel, defaultAgentTypeSchema, defaultMessageRequestSchema, detailGroupMessageResponseSchema, failedMessageSchema, fileIdsSchema, fileTypeSchema, fileUploadRequestSchema, fileUploadResponseSchema, finalizeGetBlacksRequest, finalizeGetBlockGroupsRequest, finalizeGetBlockNumbersRequest, finalizeGetGroupsRequest, finalizeGetKakaoAlimtalkTemplatesRequest, finalizeGetKakaoChannelsRequest, finalizeGetMessagesRequest, finalizeGetStatisticsRequest, getBalanceResponseSchema, getBlacksRequestSchema, getBlacksResponseSchema, getBlockGroupsRequestSchema, getBlockGroupsResponseSchema, getBlockNumbersRequestSchema, getBlockNumbersResponseSchema, getGroupMessagesRequestSchema, getGroupsRequestSchema, getGroupsResponseSchema, getKakaoAlimtalkTemplatesRequestSchema, getKakaoAlimtalkTemplatesResponseSchema, getKakaoChannelsRequestSchema, getKakaoChannelsResponseSchema, getKakaoTemplateResponseSchema, getMessagesRequestSchema, getMessagesResponseSchema, getStatisticsRequestSchema, getStatisticsResponseSchema, groupIdSchema, groupMessageAddRequestSchema, groupMessageResponseSchema, groupSchema, handleKeySchema, isErrorResponse, kakaoAlimtalkTemplateAssignTypeSchema, kakaoAlimtalkTemplateCommentTypeSchema, kakaoAlimtalkTemplateEmphasizeTypeSchema, kakaoAlimtalkTemplateHighlightTypeSchema, kakaoAlimtalkTemplateItemTypeSchema, kakaoAlimtalkTemplateMessageTypeSchema, kakaoAlimtalkTemplateQuickReplyAppLinkSchema, kakaoAlimtalkTemplateQuickReplyDefaultSchema, kakaoAlimtalkTemplateQuickReplySchema, kakaoAlimtalkTemplateQuickReplyWebLinkSchema, kakaoAlimtalkTemplateSchema, kakaoAlimtalkTemplateStatusSchema, kakaoButtonSchema, kakaoChannelCategorySchema, kakaoChannelSchema, logSchema, messageResponseItemSchema, messageSchema, messageTypeRecordSchema, messageTypeSchema, multipleMessageSendingRequestSchema, naverOptionSchema, osPlatform, phoneNumberSchema, rcsButtonSchema, rcsOptionRequestSchema, rcsOptionSchema, removeGroupMessagesResponseSchema, removeMessageIdsToGroupRequestSchema, requestKakaoChannelTokenResponseSchema, requestSendMessageSchema, requestSendOneMessageSchema, scheduledDateSendingRequestSchema, sdkVersion, sendRequestConfigSchema, storedMessageSchema, transformVariables, updateKakaoAlimtalkTemplateRequestSchema, validateAcceptableFields, validateAllowedLinkTypes, validateButtonCount, validateButtonNames, validateCarouselListCount, validateCouponDescription, validateForbiddenVariables, validateImageIdLength, validateLinks, validateNewlineLimits, validateTextLengths, validateVariableNames, voiceOptionSchema };
